// @anvia/core@0.7.1/dist/transcription/index.js
// src/transcription/index.ts
var TranscriptionRequestBuilder = class {
  constructor(model) {
    this.model = model;
  }
  model;
  request = {
    data: new Uint8Array(),
    filename: "file"
  };
  data(data) {
    this.request = { ...this.request, data: toUint8Array(data) };
    return this;
  }
  filename(filename) {
    this.request = { ...this.request, filename };
    return this;
  }
  language(language) {
    this.request = { ...this.request, language };
    return this;
  }
  prompt(prompt) {
    this.request = { ...this.request, prompt };
    return this;
  }
  temperature(temperature) {
    this.request = { ...this.request, temperature };
    return this;
  }
  additionalParams(additionalParams) {
    this.request = { ...this.request, additionalParams };
    return this;
  }
  build() {
    if (this.request.data.byteLength === 0) {
      throw new Error("Transcription data cannot be empty.");
    }
    return { ...this.request, data: toUint8Array(this.request.data) };
  }
  send() {
    return this.model.transcription(this.build());
  }
};
function transcriptionRequest(model) {
  return new TranscriptionRequestBuilder(model);
}
function toUint8Array(bytes) {
  if (bytes instanceof Uint8Array) {
    return new Uint8Array(
      bytes.buffer.slice(bytes.byteOffset, bytes.byteOffset + bytes.byteLength)
    );
  }
  return new Uint8Array(bytes);
}
export {
  TranscriptionRequestBuilder,
  transcriptionRequest
};
