// @anvia/anthropic@0.3.10/dist/index.js
var __defProp = Object.defineProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};

// src/anthropic/index.ts
var anthropic_exports = {};
__export(anthropic_exports, {
  AnthropicClient: () => AnthropicClient,
  AnthropicCompletionModel: () => AnthropicCompletionModel
});

// src/anthropic/client.ts
import Anthropic from "@anthropic-ai/sdk";
import {
  ModelListingError
} from "@anvia/core/model-listing";

// src/anthropic/completion.ts
import {
  AssistantContent,
  assertCompletionRequestSupported,
  Usage
} from "@anvia/core/completion";

// src/request-messages.ts
import {
  Message,
  normalizeDocuments
} from "@anvia/core/completion";
function orderedRequestMessages(request, options = {}) {
  const messages = [];
  if (options.includeInstructionsAsSystem === true && request.instructions !== void 0) {
    messages.push(Message.system(request.instructions));
  }
  messages.push(...request.chatHistory.filter((message) => message.role === "system"));
  const documents = normalizeDocuments(request.documents);
  if (documents !== void 0) {
    messages.push(documents);
  }
  messages.push(...request.chatHistory.filter((message) => message.role !== "system"));
  return messages;
}

// src/utils.ts
function isPlainObject(value) {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}
function numberFrom(value) {
  return typeof value === "number" && Number.isFinite(value) ? value : 0;
}
function stringFrom(value) {
  return typeof value === "string" ? value : void 0;
}

// src/anthropic/completion.ts
var DEFAULT_MAX_TOKENS = 1024;
var AnthropicCompletionModel = class {
  constructor(client, defaultModel = "claude-sonnet-4-20250514") {
    this.client = client;
    this.defaultModel = defaultModel;
  }
  client;
  defaultModel;
  provider = "anthropic";
  capabilities = {
    streaming: true,
    tools: true,
    toolChoice: true,
    imageInput: true,
    documentInput: true,
    outputSchema: false,
    reasoning: true
  };
  traceRequest(request, options = {}) {
    const params = toAnthropicMessagesParams(this.defaultModel, request);
    if (options.stream === true) {
      params.stream = true;
    }
    return providerRequestSummary(params, request, options);
  }
  async completion(request) {
    assertCompletionRequestSupported(this, request);
    const params = toAnthropicMessagesParams(this.defaultModel, request);
    const response = await this.client.messages.create(params);
    return fromAnthropicMessage(response);
  }
  async *streamCompletion(request) {
    assertCompletionRequestSupported(this, request, { streaming: true });
    const params = { ...toAnthropicMessagesParams(this.defaultModel, request), stream: true };
    const stream = await this.client.messages.create(params);
    const toolIdsByIndex = /* @__PURE__ */ new Map();
    const blocksWithInitialToolInput = /* @__PURE__ */ new Set();
    const streamUsage = Usage.empty();
    const streamUsageFields = emptyAnthropicStreamUsageFields();
    let hasStreamUsage = false;
    let streamMessageId;
    for await (const event of stream) {
      if (isPlainObject(event) && event.type === "content_block_start") {
        const index = numberFrom(event.index);
        const block = isPlainObject(event.content_block) ? event.content_block : {};
        const id = stringFrom(block.id);
        if (id !== void 0) {
          toolIdsByIndex.set(index, id);
        }
        if (toolInputArgumentsDelta(block.input) !== void 0) {
          blocksWithInitialToolInput.add(index);
        }
      }
      if (isPlainObject(event)) {
        if (event.type === "message_start" && isPlainObject(event.message)) {
          streamMessageId = stringFrom(event.message.id) ?? streamMessageId;
        }
        hasStreamUsage = applyAnthropicStreamUsage(event, streamUsage, streamUsageFields) || hasStreamUsage;
      }
      for (const mapped of fromAnthropicStreamEvent(event)) {
        const streamMapped = mapped.type === "final" && hasStreamUsage ? {
          ...mapped,
          response: {
            ...mapped.response,
            usage: mergeUsage(mapped.response.usage, streamUsage, streamUsageFields)
          }
        } : mapped;
        if (streamMapped.type === "tool_call_delta" && streamMapped.argumentsDelta !== void 0 && isPlainObject(event) && event.type === "content_block_delta" && blocksWithInitialToolInput.has(numberFrom(event.index))) {
          continue;
        }
        if (streamMapped.type === "tool_call_delta" && streamMapped.id.startsWith("tool_")) {
          const index = Number(streamMapped.id.slice("tool_".length));
          yield { ...streamMapped, id: toolIdsByIndex.get(index) ?? streamMapped.id };
        } else {
          yield streamMapped;
        }
      }
      if (isPlainObject(event) && event.type === "message_stop" && !isPlainObject(event.message)) {
        const response = {
          choice: [],
          usage: copyUsage(streamUsage),
          rawResponse: event
        };
        if (streamMessageId !== void 0) {
          response.messageId = streamMessageId;
        }
        yield { type: "final", response };
      }
    }
  }
};
function emptyAnthropicStreamUsageFields() {
  return {
    inputTokens: false,
    outputTokens: false,
    cachedInputTokens: false,
    cacheCreationInputTokens: false
  };
}
function applyAnthropicStreamUsage(event, usage, fields) {
  let changed = false;
  if (event.type === "message_start" && isPlainObject(event.message)) {
    const source = isPlainObject(event.message.usage) ? event.message.usage : void 0;
    if (source !== void 0) {
      if ("input_tokens" in source) {
        usage.inputTokens = numberFrom(source.input_tokens);
        fields.inputTokens = true;
        changed = true;
      }
      if ("cache_read_input_tokens" in source) {
        usage.cachedInputTokens = numberFrom(source.cache_read_input_tokens);
        fields.cachedInputTokens = true;
        changed = true;
      }
      if ("cache_creation_input_tokens" in source) {
        usage.cacheCreationInputTokens = numberFrom(source.cache_creation_input_tokens);
        fields.cacheCreationInputTokens = true;
        changed = true;
      }
    }
  }
  if (event.type === "message_delta") {
    const source = isPlainObject(event.usage) ? event.usage : void 0;
    if (source !== void 0 && "output_tokens" in source) {
      usage.outputTokens = Math.max(usage.outputTokens, numberFrom(source.output_tokens));
      fields.outputTokens = true;
      changed = true;
    }
  }
  if (changed) {
    usage.totalTokens = usage.inputTokens + usage.outputTokens;
  }
  return changed;
}
function mergeUsage(base, stream, fields) {
  const inputTokens = fields.inputTokens ? stream.inputTokens : base.inputTokens;
  const outputTokens = fields.outputTokens ? stream.outputTokens : base.outputTokens;
  return {
    inputTokens,
    outputTokens,
    totalTokens: inputTokens + outputTokens,
    cachedInputTokens: fields.cachedInputTokens ? stream.cachedInputTokens : base.cachedInputTokens,
    cacheCreationInputTokens: fields.cacheCreationInputTokens ? stream.cacheCreationInputTokens : base.cacheCreationInputTokens
  };
}
function copyUsage(usage) {
  return { ...usage };
}
function toAnthropicMessagesParams(defaultModel, request) {
  const messages = requestMessages(request);
  const system = systemFromMessages(request, messages);
  const params = {
    model: request.model ?? defaultModel,
    max_tokens: request.maxTokens ?? DEFAULT_MAX_TOKENS,
    messages: messages.flatMap(messageToAnthropicMessages)
  };
  if (system !== void 0) {
    params.system = system;
  }
  if (request.tools.length > 0) {
    params.tools = request.tools.map(toolDefinitionToAnthropic);
  }
  if (request.temperature !== void 0) {
    params.temperature = request.temperature;
  }
  if (request.toolChoice !== void 0) {
    params.tool_choice = toolChoiceToAnthropic(request.toolChoice);
  }
  if (request.additionalParams !== void 0 && isPlainObject(request.additionalParams)) {
    Object.assign(params, request.additionalParams);
  }
  return params;
}
function providerRequestSummary(params, request, options) {
  return compactJsonObject({
    provider: "anthropic",
    api: "messages",
    stream: options.stream === true,
    model: stringFrom(params.model),
    parameterKeys: Object.keys(params).sort(),
    messageCount: Array.isArray(params.messages) ? params.messages.length : void 0,
    toolCount: request.tools.length,
    toolNames: request.tools.map((tool) => tool.name),
    hasSystem: typeof params.system === "string" && params.system.length > 0,
    temperature: request.temperature,
    maxTokens: request.maxTokens ?? numberFrom(params.max_tokens),
    toolChoice: toolChoiceSummary(request.toolChoice),
    additionalParamKeys: isPlainObject(request.additionalParams) ? Object.keys(request.additionalParams).sort() : void 0
  });
}
function toolChoiceSummary(toolChoice) {
  if (toolChoice === void 0 || typeof toolChoice === "string") {
    return toolChoice;
  }
  return { type: toolChoice.type, name: toolChoice.name };
}
function compactJsonObject(values) {
  return Object.fromEntries(
    Object.entries(values).flatMap(([key, value]) => {
      if (value === void 0) {
        return [];
      }
      return [[key, toJsonValue(value)]];
    })
  );
}
function requestMessages(request) {
  return orderedRequestMessages(request);
}
function systemFromMessages(request, messages) {
  const systemMessages = messages.flatMap(
    (message) => message.role === "system" ? [message.content] : []
  );
  if (request.instructions !== void 0) {
    systemMessages.unshift(request.instructions);
  }
  return systemMessages.length === 0 ? void 0 : systemMessages.join("\n\n");
}
function fromAnthropicMessage(response) {
  const raw = response;
  const content = Array.isArray(raw.content) ? raw.content : [];
  const choice = [];
  for (const block of content) {
    if (!isPlainObject(block)) {
      continue;
    }
    if (block.type === "text" && typeof block.text === "string") {
      choice.push(AssistantContent.text(block.text));
    }
    if (block.type === "thinking" && typeof block.thinking === "string") {
      const text = typeof block.signature === "string" ? { type: "text", text: block.thinking, signature: block.signature } : { type: "text", text: block.thinking };
      choice.push(AssistantContent.reasoningFromContent([text]));
    }
    if (block.type === "redacted_thinking" && typeof block.data === "string") {
      choice.push(AssistantContent.reasoningRedacted(block.data));
    }
    if (block.type === "tool_use") {
      const id = typeof block.id === "string" ? block.id : crypto.randomUUID();
      const name = typeof block.name === "string" ? block.name : "";
      choice.push(AssistantContent.toolCall(id, name, toJsonValue(block.input)));
    }
  }
  const usageSource = isPlainObject(raw.usage) ? raw.usage : {};
  const result = {
    choice,
    usage: {
      ...Usage.empty(),
      inputTokens: numberFrom(usageSource.input_tokens),
      outputTokens: numberFrom(usageSource.output_tokens),
      totalTokens: numberFrom(usageSource.input_tokens) + numberFrom(usageSource.output_tokens),
      cachedInputTokens: numberFrom(usageSource.cache_read_input_tokens),
      cacheCreationInputTokens: numberFrom(usageSource.cache_creation_input_tokens)
    },
    rawResponse: response
  };
  if (typeof raw.id === "string") {
    result.messageId = raw.id;
  }
  return result;
}
function fromAnthropicStreamEvent(event) {
  if (!isPlainObject(event) || typeof event.type !== "string") {
    return [];
  }
  if (event.type === "message_start" && isPlainObject(event.message)) {
    const id = stringFrom(event.message.id);
    return id === void 0 ? [] : [{ type: "message_id", id }];
  }
  if (event.type === "content_block_start" && isPlainObject(event.content_block)) {
    const block = event.content_block;
    if (block.type === "tool_use") {
      return [
        toolCallDelta(stringFrom(block.id) ?? `tool_${numberFrom(event.index)}`, {
          name: stringFrom(block.name),
          argumentsDelta: toolInputArgumentsDelta(block.input)
        })
      ];
    }
    if (block.type === "redacted_thinking" && typeof block.data === "string") {
      return [
        {
          type: "reasoning_delta",
          delta: block.data,
          id: `thinking_${numberFrom(event.index)}`,
          contentType: "redacted"
        }
      ];
    }
    return [];
  }
  if (event.type === "content_block_delta" && isPlainObject(event.delta)) {
    const delta = event.delta;
    if (delta.type === "text_delta" && typeof delta.text === "string") {
      return [{ type: "text_delta", delta: delta.text }];
    }
    if (delta.type === "thinking_delta" && typeof delta.thinking === "string") {
      return [
        {
          type: "reasoning_delta",
          delta: delta.thinking,
          id: `thinking_${numberFrom(event.index)}`,
          contentType: "text"
        }
      ];
    }
    if (delta.type === "signature_delta" && typeof delta.signature === "string") {
      return [
        {
          type: "reasoning_delta",
          delta: "",
          id: `thinking_${numberFrom(event.index)}`,
          contentType: "text",
          signature: delta.signature
        }
      ];
    }
    if (delta.type === "input_json_delta" && typeof delta.partial_json === "string") {
      return [
        toolCallDelta(`tool_${numberFrom(event.index)}`, { argumentsDelta: delta.partial_json })
      ];
    }
  }
  if (event.type === "message_stop" && isPlainObject(event.message)) {
    return [{ type: "final", response: fromAnthropicMessage(event.message) }];
  }
  if (event.type === "error") {
    return [{ type: "error", error: event.error ?? event }];
  }
  return [];
}
function messageToAnthropicMessages(message) {
  if (message.role === "system") {
    return [];
  }
  if (message.role === "user") {
    return [
      {
        role: "user",
        content: message.content.map(userContentToAnthropicBlock)
      }
    ];
  }
  if (message.role === "tool") {
    return [
      {
        role: "user",
        content: message.content.map(toolContentToAnthropicBlock)
      }
    ];
  }
  return [
    {
      role: "assistant",
      content: message.content.flatMap((content) => {
        if (content.type === "text") {
          return [{ type: "text", text: content.text }];
        }
        if (content.type === "tool_call") {
          return [
            {
              type: "tool_use",
              id: content.callId ?? content.id,
              name: content.function.name,
              input: content.function.arguments ?? {}
            }
          ];
        }
        if (content.type === "reasoning" && content.content !== void 0) {
          return content.content.flatMap(reasoningContentToAnthropicBlocks);
        }
        if (content.type === "image") {
          throw new Error("Anthropic Messages does not support image content in assistant history");
        }
        return [];
      })
    }
  ];
}
function toolContentToAnthropicBlock(content) {
  return {
    type: "tool_result",
    tool_use_id: content.callId ?? content.id,
    content: toolResultContentToAnthropicContent(content.content)
  };
}
function toolResultContentToAnthropicContent(content) {
  if (content.every((item) => item.type === "text")) {
    return content.map((item) => item.text).join("\n");
  }
  return content.map((item) => {
    if (item.type === "text") {
      return { type: "text", text: item.text };
    }
    return {
      type: "image",
      source: {
        type: "base64",
        media_type: item.mediaType ?? "image/png",
        data: item.data
      }
    };
  });
}
function reasoningContentToAnthropicBlocks(content) {
  if (content.type === "text" || content.type === "summary") {
    const block = {
      type: "thinking",
      thinking: content.text
    };
    if (content.type === "text" && content.signature !== void 0) {
      block.signature = content.signature;
    }
    return [block];
  }
  if (content.type === "redacted") {
    return [{ type: "redacted_thinking", data: content.data }];
  }
  return [];
}
function userContentToAnthropicBlock(content) {
  if (content.type === "text") {
    return { type: "text", text: content.text };
  }
  if (content.type === "image") {
    return imageToAnthropicBlock(content);
  }
  if (content.type === "document") {
    return documentToAnthropicBlock(content);
  }
  throw new Error("Tool results must be mapped before user content blocks");
}
function imageToAnthropicBlock(image) {
  if (image.source.type === "url") {
    return {
      type: "image",
      source: {
        type: "url",
        url: image.source.url
      }
    };
  }
  return {
    type: "image",
    source: {
      type: "base64",
      media_type: image.source.mediaType,
      data: image.source.data
    }
  };
}
function documentToAnthropicBlock(document) {
  if (document.source.type === "text") {
    return { type: "text", text: document.source.text };
  }
  if (document.source.mediaType !== "application/pdf") {
    throw new Error("Anthropic Messages only supports PDF document attachments");
  }
  if (document.source.type === "url") {
    return {
      type: "document",
      source: {
        type: "url",
        url: document.source.url
      }
    };
  }
  return {
    type: "document",
    source: {
      type: "base64",
      media_type: document.source.mediaType,
      data: document.source.data
    }
  };
}
function toolDefinitionToAnthropic(tool) {
  return {
    name: tool.name,
    description: tool.description,
    input_schema: tool.parameters
  };
}
function toolChoiceToAnthropic(toolChoice) {
  if (toolChoice === "auto") {
    return { type: "auto" };
  }
  if (toolChoice === "required") {
    return { type: "any" };
  }
  if (toolChoice === "none") {
    return { type: "none" };
  }
  return {
    type: "tool",
    name: toolChoice.name
  };
}
function toJsonValue(value) {
  if (value === null || typeof value === "string" || typeof value === "number" || typeof value === "boolean" || Array.isArray(value) || isPlainObject(value)) {
    return value;
  }
  return String(value);
}
function toolInputArgumentsDelta(input) {
  if (input === void 0) {
    return void 0;
  }
  if (typeof input === "string") {
    const trimmed = input.trim();
    if (trimmed.length === 0 || trimmed === "{}") {
      return void 0;
    }
    try {
      JSON.parse(trimmed);
      return trimmed;
    } catch {
      return JSON.stringify(input);
    }
  }
  if (isPlainObject(input) && Object.keys(input).length === 0) {
    return void 0;
  }
  return JSON.stringify(toJsonValue(input));
}
function toolCallDelta(id, values) {
  const event = { type: "tool_call_delta", id };
  if (values.name !== void 0) event.name = values.name;
  if (values.argumentsDelta !== void 0) event.argumentsDelta = values.argumentsDelta;
  return event;
}

// src/anthropic/client.ts
var AnthropicClient = class {
  client;
  constructor(options = {}) {
    this.client = options.client ?? new Anthropic({
      apiKey: requireApiKey(options.apiKey),
      baseURL: options.baseUrl
    });
  }
  completionModel(model = "claude-sonnet-4-20250514") {
    return new AnthropicCompletionModel(this.client, model);
  }
  async listModels() {
    try {
      const response = await this.client.models.list();
      const data = (await collectModelsFromResponse(response)).map(toListedModel).filter(isListedModel);
      return { data };
    } catch (error) {
      throw toModelListingError("Anthropic", error);
    }
  }
};
function requireApiKey(apiKey) {
  if (apiKey === void 0 || apiKey.length === 0) {
    throw new Error(
      "Missing Anthropic credentials. Pass apiKey when constructing AnthropicClient."
    );
  }
  return apiKey;
}
async function collectModelsFromResponse(response) {
  if (isAsyncIterable(response)) {
    const models = [];
    for await (const model of response) {
      models.push(model);
    }
    return models;
  }
  if (Array.isArray(response)) {
    return response;
  }
  if (isObject(response) && Array.isArray(response.data)) {
    return response.data;
  }
  return [];
}
function toListedModel(model) {
  if (!isObject(model) || typeof model.id !== "string") {
    return void 0;
  }
  const createdAt = typeof model.created_at === "string" ? secondsFromDateString(model.created_at) : typeof model.created_at === "number" ? model.created_at : void 0;
  return {
    id: model.id,
    ...typeof model.display_name === "string" ? { name: model.display_name } : {},
    ...typeof model.name === "string" ? { name: model.name } : {},
    ...typeof model.description === "string" ? { description: model.description } : {},
    ...typeof model.type === "string" ? { type: model.type } : {},
    ...createdAt === void 0 ? {} : { createdAt },
    ...typeof model.owned_by === "string" ? { ownedBy: model.owned_by } : {},
    ...typeof model.max_input_tokens === "number" ? { contextLength: model.max_input_tokens } : {},
    ...typeof model.context_length === "number" ? { contextLength: model.context_length } : {}
  };
}
function secondsFromDateString(value) {
  const time = Date.parse(value);
  return Number.isNaN(time) ? void 0 : Math.floor(time / 1e3);
}
function isListedModel(model) {
  return model !== void 0;
}
function toModelListingError(provider, error) {
  if (error instanceof ModelListingError) {
    return error;
  }
  const statusCode = getStatusCode(error);
  return new ModelListingError(`${provider} model listing failed: ${getErrorMessage(error)}`, {
    provider,
    ...statusCode === void 0 ? {} : { statusCode },
    cause: error
  });
}
function getStatusCode(error) {
  if (!isObject(error)) {
    return void 0;
  }
  if (typeof error.status === "number") {
    return error.status;
  }
  if (typeof error.statusCode === "number") {
    return error.statusCode;
  }
  return void 0;
}
function getErrorMessage(error) {
  return error instanceof Error ? error.message : String(error);
}
function isObject(value) {
  return typeof value === "object" && value !== null;
}
function isAsyncIterable(value) {
  return isObject(value) && Symbol.asyncIterator in value;
}
export {
  AnthropicClient,
  AnthropicCompletionModel,
  anthropic_exports as anthropic
};
