// @anvia/core@0.8.0/dist/mcp/index.js
import {
  isRecord
} from "../chunk-MMHG7WAM.js";

// src/mcp/result.ts
function createCallToolParams(name, args) {
  if (args === null || args === void 0) {
    return { name };
  }
  if (!isRecord(args)) {
    throw new Error("MCP tool arguments must be a JSON object");
  }
  return { name, arguments: args };
}
function mapMcpToolResult(result) {
  if ("toolResult" in result) {
    return serializeMcpValue(result.toolResult);
  }
  if (result.isError === true) {
    throw new Error(mcpErrorMessage(result.content));
  }
  return result.content.map(mapMcpContent).join("");
}
function mcpErrorMessage(content) {
  const text = content.map((item) => item.type === "text" ? item.text : void 0).filter((item) => item !== void 0).join("\n");
  return text === "" ? "MCP tool returned an error" : text;
}
function mapMcpContent(content) {
  if (content.type === "text") {
    return content.text;
  }
  if (content.type === "image") {
    return `data:${content.mimeType};base64,${content.data}`;
  }
  if (content.type === "resource") {
    const mimeType = content.resource.mimeType === void 0 ? "" : `data:${content.resource.mimeType};`;
    if ("text" in content.resource) {
      return `${mimeType}${content.resource.uri}:${content.resource.text}`;
    }
    return `${mimeType}${content.resource.uri}:${content.resource.blob}`;
  }
  throw new Error(`Unsupported MCP tool result content: ${serializeMcpValue(content)}`);
}
function serializeMcpValue(value) {
  if (typeof value === "string") {
    return value;
  }
  const serialized = JSON.stringify(value);
  return serialized === void 0 ? String(value) : serialized;
}

// src/mcp/tool.ts
var MCP_TOOL_METADATA_KEY = /* @__PURE__ */ Symbol.for("anvia.mcp.tool.metadata");
function createMcpTool(definition, client, serverName) {
  const tool = {
    name: definition.name,
    definition() {
      return {
        name: definition.name,
        description: definition.description ?? "",
        parameters: definition.inputSchema
      };
    },
    async call(args) {
      const result = await client.callTool(createCallToolParams(definition.name, args));
      return mapMcpToolResult(result);
    }
  };
  if (serverName !== void 0) {
    Object.defineProperty(tool, MCP_TOOL_METADATA_KEY, {
      value: { serverName },
      enumerable: false
    });
  }
  return tool;
}

// src/mcp/connect.ts
async function connectMcp(connection) {
  const client = await connection.connect();
  let tools;
  try {
    ({ tools } = await client.listTools());
  } catch (error) {
    try {
      await client.close();
    } catch {
    }
    throw error;
  }
  return {
    name: connection.name,
    tools: tools.map((tool) => createMcpTool(tool, client, connection.name)),
    close: () => client.close()
  };
}

// src/mcp/connections.ts
import { Client } from "@modelcontextprotocol/sdk/client/index.js";
import { SSEClientTransport } from "@modelcontextprotocol/sdk/client/sse.js";
import { StdioClientTransport } from "@modelcontextprotocol/sdk/client/stdio.js";
import { StreamableHTTPClientTransport } from "@modelcontextprotocol/sdk/client/streamableHttp.js";
var mcp = {
  stdio(options) {
    return {
      name: options.name,
      async connect() {
        const { name: _name, ...server } = options;
        const client = createSdkClient();
        await client.connect(asSdkTransport(new StdioClientTransport(server)));
        return client;
      }
    };
  },
  http(options) {
    return {
      name: options.name,
      async connect() {
        const client = createSdkClient();
        await client.connect(
          asSdkTransport(
            new StreamableHTTPClientTransport(new URL(options.url), options.transport)
          )
        );
        return client;
      }
    };
  },
  sse(options) {
    return {
      name: options.name,
      async connect() {
        const client = createSdkClient();
        await client.connect(
          asSdkTransport(new SSEClientTransport(new URL(options.url), options.transport))
        );
        return client;
      }
    };
  }
};
function createSdkClient() {
  return new Client({
    name: "@anvia/core",
    version: "0.1.0"
  });
}
function asSdkTransport(transport) {
  return transport;
}
export {
  connectMcp,
  mcp
};
