// @anvia/core@0.12.8/dist/chunk-JS5MGVNM.js
import {
  mapWithConcurrency
} from "./chunk-OIMLU4SF.js";

// src/embeddings/distance.ts
function dotProduct(left, right) {
  assertSameDimensions(left, right);
  return left.reduce((sum, value, index) => sum + value * right[index], 0);
}
function cosineSimilarity(left, right) {
  assertSameDimensions(left, right);
  const leftMagnitude = magnitude(left);
  const rightMagnitude = magnitude(right);
  if (leftMagnitude === 0 || rightMagnitude === 0) {
    return 0;
  }
  return dotProduct(left, right) / (leftMagnitude * rightMagnitude);
}
function angularDistance(left, right) {
  const similarity = Math.max(-1, Math.min(1, cosineSimilarity(left, right)));
  return Math.acos(similarity) / Math.PI;
}
function euclideanDistance(left, right) {
  assertSameDimensions(left, right);
  return Math.sqrt(
    left.reduce((sum, value, index) => sum + (value - right[index]) ** 2, 0)
  );
}
function manhattanDistance(left, right) {
  assertSameDimensions(left, right);
  return left.reduce((sum, value, index) => sum + Math.abs(value - right[index]), 0);
}
function chebyshevDistance(left, right) {
  assertSameDimensions(left, right);
  return left.reduce(
    (max, value, index) => Math.max(max, Math.abs(value - right[index])),
    0
  );
}
function magnitude(vector) {
  return Math.sqrt(vector.reduce((sum, value) => sum + value ** 2, 0));
}
function assertSameDimensions(left, right) {
  if (left.length !== right.length) {
    throw new Error(`Vector dimension mismatch: ${left.length} !== ${right.length}`);
  }
}

// src/embeddings/embed.ts
async function embedText(model, text) {
  const embeddings = await embedTexts(model, [text]);
  const embedding = embeddings[0];
  if (embedding === void 0) {
    throw new Error("Embedding model returned no embeddings");
  }
  return embedding;
}
async function embedTexts(model, texts) {
  if (texts.length === 0) {
    return [];
  }
  const maxBatchSize = Math.max(1, Math.trunc(model.maxBatchSize ?? texts.length));
  const batches = [];
  for (let index = 0; index < texts.length; index += maxBatchSize) {
    batches.push(texts.slice(index, index + maxBatchSize));
  }
  const results = await mapWithConcurrency(batches, 1, (batch) => model.embedTexts(batch));
  const embeddings = results.flat();
  if (embeddings.length !== texts.length) {
    throw new Error(
      `Embedding model returned ${embeddings.length} embeddings for ${texts.length} texts`
    );
  }
  return embeddings;
}
async function embedDocuments(model, documents, options) {
  const prepared = documents.map((document, index) => {
    const content = options.content(document, index);
    const texts = Array.isArray(content) ? content : [content];
    return {
      id: options.id?.(document, index) ?? `doc${index}`,
      document,
      metadata: options.metadata?.(document, index),
      texts
    };
  });
  const flatTexts = prepared.flatMap(
    (item, documentIndex) => item.texts.map((text) => ({ documentIndex, text }))
  );
  const embeddings = await mapWithConcurrency(
    chunk(flatTexts, Math.max(1, Math.trunc(model.maxBatchSize ?? (flatTexts.length || 1)))),
    Math.max(1, Math.trunc(options.concurrency ?? 1)),
    async (batch) => {
      const batchEmbeddings = await model.embedTexts(batch.map((item) => item.text));
      if (batchEmbeddings.length !== batch.length) {
        throw new Error(
          `Embedding model returned ${batchEmbeddings.length} embeddings for ${batch.length} texts`
        );
      }
      return batch.map((item, index) => ({
        documentIndex: item.documentIndex,
        embedding: batchEmbeddings[index]
      }));
    }
  );
  const byDocument = /* @__PURE__ */ new Map();
  for (const item of embeddings.flat()) {
    const list = byDocument.get(item.documentIndex) ?? [];
    list.push(item.embedding);
    byDocument.set(item.documentIndex, list);
  }
  return prepared.map((item, index) => ({
    id: item.id,
    document: item.document,
    ...item.metadata !== void 0 && { metadata: item.metadata },
    embeddings: byDocument.get(index) ?? []
  }));
}
function chunk(items, size) {
  const chunks = [];
  for (let index = 0; index < items.length; index += size) {
    chunks.push(items.slice(index, index + size));
  }
  return chunks;
}

export {
  dotProduct,
  cosineSimilarity,
  angularDistance,
  euclideanDistance,
  manhattanDistance,
  chebyshevDistance,
  embedText,
  embedTexts,
  embedDocuments
};
