// @anvia/core@0.12.8/dist/chunk-IN5NPAWL.js
import {
  AgentBuilder
} from "./chunk-WRNPY5C6.js";
import {
  extractRagText
} from "./chunk-EGZCTSEW.js";
import {
  createTool
} from "./chunk-IZM6FNCT.js";
import {
  CompletionRequestBuilder
} from "./chunk-FUJ3POKH.js";
import {
  CompletionCapabilityError,
  Message,
  Usage
} from "./chunk-QVFOCNF6.js";

// src/extractor/extractor.ts
var SUBMIT_TOOL_NAME = "submit";
var DEFAULT_EXTRACTOR_INSTRUCTIONS = "You are an AI assistant whose purpose is to extract structured data from the provided text.\nYou have access to a `submit` function that defines the structure of the data to extract.\nAlways call the `submit` function with the structured data. Use default or null values when information is missing.";
var ExtractionError = class extends Error {
  constructor(message, cause) {
    super(message);
    this.cause = cause;
    this.name = "ExtractionError";
  }
  cause;
};
var Extractor = class {
  constructor(agent, schema, retryCount) {
    this.agent = agent;
    this.schema = schema;
    this.retryCount = retryCount;
  }
  agent;
  schema;
  retryCount;
  async extract(text) {
    return (await this.extractWithUsage(text)).data;
  }
  async extractWithUsage(text) {
    return this.run(text);
  }
  async extractWithHistory(text, history) {
    return (await this.run(text, history)).data;
  }
  getInner() {
    return this.agent;
  }
  async run(text, history) {
    let usage = Usage.empty();
    let lastError;
    const prompt = typeof text === "string" ? Message.user(text) : text;
    for (let attempt = 0; attempt <= this.retryCount; attempt += 1) {
      try {
        const toolDefs = await this.agent.toolSet.getToolDefinitions(extractRagText(prompt));
        const response = await new CompletionRequestBuilder(this.agent.model, prompt).instructions(this.agent.instructions).messages(history ?? []).documents(this.agent.staticContext).tools(toolDefs).temperature(this.agent.temperature).maxTokens(this.agent.maxTokens).additionalParams(this.agent.additionalParams).toolChoice(this.agent.toolChoice).send();
        usage = Usage.add(usage, response.usage);
        const data = extractSubmittedData(response, this.schema);
        return {
          data,
          usage,
          messages: [
            ...history ?? [],
            prompt,
            Message.assistant(response.choice, response.messageId)
          ]
        };
      } catch (error) {
        if (error instanceof CompletionCapabilityError) {
          throw error;
        }
        lastError = error;
      }
    }
    throw new ExtractionError("No data extracted", lastError);
  }
};
var ExtractorBuilder = class {
  constructor(model, schema) {
    this.schema = schema;
    this.agentBuilder = new AgentBuilder("extractor", model).instructions(DEFAULT_EXTRACTOR_INSTRUCTIONS).tool(
      createTool({
        name: SUBMIT_TOOL_NAME,
        description: "Submit the structured data extracted from the provided text.",
        input: schema,
        output: schema,
        execute: (args) => args
      })
    ).toolChoice("required");
  }
  schema;
  agentBuilder;
  retryCount = 0;
  instructions(instructions) {
    this.agentBuilder.instructions(instructions);
    return this;
  }
  context(text, id) {
    this.agentBuilder.context(text, id);
    return this;
  }
  temperature(temperature) {
    this.agentBuilder.temperature(temperature);
    return this;
  }
  maxTokens(maxTokens) {
    this.agentBuilder.maxTokens(maxTokens);
    return this;
  }
  additionalParams(params) {
    this.agentBuilder.additionalParams(params);
    return this;
  }
  toolChoice(toolChoice) {
    this.agentBuilder.toolChoice(toolChoice);
    return this;
  }
  retries(retries) {
    this.retryCount = Math.max(0, Math.trunc(retries));
    return this;
  }
  build() {
    return new Extractor(this.agentBuilder.build(), this.schema, this.retryCount);
  }
};
function extractSubmittedData(response, schema) {
  const submitted = response.choice.filter((content) => content.type === "tool_call").filter((toolCall) => toolCall.function.name === SUBMIT_TOOL_NAME).at(-1);
  if (submitted === void 0) {
    throw new ExtractionError("The model did not call the submit tool");
  }
  return schema.parse(submitted.function.arguments);
}

export {
  ExtractionError,
  Extractor,
  ExtractorBuilder
};
