// @anvia/core@0.12.8/dist/chunk-QVFOCNF6.js
// src/completion/types.ts
var UserContent = {
  text(text) {
    return { type: "text", text };
  },
  imageUrl(url, options = {}) {
    const image = { type: "image", source: { type: "url", url } };
    if (options.detail !== void 0) {
      image.detail = options.detail;
    }
    return image;
  },
  imageBase64(data, mediaType, options = {}) {
    const image = {
      type: "image",
      source: { type: "base64", data, mediaType }
    };
    if (options.detail !== void 0) {
      image.detail = options.detail;
    }
    return image;
  },
  documentUrl(url, mediaType, options = {}) {
    return {
      type: "document",
      source: options.filename === void 0 ? { type: "url", url, mediaType } : { type: "url", url, mediaType, filename: options.filename }
    };
  },
  documentBase64(data, mediaType, options = {}) {
    return {
      type: "document",
      source: options.filename === void 0 ? { type: "base64", data, mediaType } : { type: "base64", data, mediaType, filename: options.filename }
    };
  },
  documentText(text) {
    return { type: "text", text };
  }
};
var ToolContent = {
  toolResult(id, content, callIdOrOptions, toolName) {
    const normalized = typeof content === "string" ? [{ type: "text", text: content }] : content;
    const options = normalizeToolResultOptions(callIdOrOptions, toolName);
    const result = { type: "tool_result", id, content: normalized };
    if (options.callId !== void 0) {
      result.callId = options.callId;
    }
    if (options.toolName !== void 0) {
      result.toolName = options.toolName;
    }
    return result;
  }
};
function normalizeToolResultOptions(callIdOrOptions, toolName) {
  if (callIdOrOptions === void 0) {
    return toolName === void 0 ? {} : { toolName };
  }
  if (typeof callIdOrOptions === "string") {
    return toolName === void 0 ? { callId: callIdOrOptions } : { callId: callIdOrOptions, toolName };
  }
  return toolName === void 0 ? callIdOrOptions : { ...callIdOrOptions, toolName };
}
function serializeToolResultOutput(output) {
  if (typeof output === "string") {
    return output;
  }
  try {
    const serialized = JSON.stringify(output);
    return serialized === void 0 ? String(output) : serialized;
  } catch {
    return String(output);
  }
}
function isToolResultContentArray(value) {
  return Array.isArray(value) && value.length > 0 && value.every((item) => {
    if (typeof item !== "object" || item === null || !("type" in item)) {
      return false;
    }
    if (item.type === "text") {
      return "text" in item && typeof item.text === "string";
    }
    if (item.type === "image") {
      return "data" in item && typeof item.data === "string" && (!("mediaType" in item) || item.mediaType === void 0 || typeof item.mediaType === "string");
    }
    return false;
  });
}
var AssistantContent = {
  text(text) {
    return { type: "text", text };
  },
  imageUrl(url, options = {}) {
    const image = { type: "image", source: { type: "url", url } };
    if (options.detail !== void 0) {
      image.detail = options.detail;
    }
    return image;
  },
  imageBase64(data, mediaType, options = {}) {
    const image = {
      type: "image",
      source: { type: "base64", data, mediaType }
    };
    if (options.detail !== void 0) {
      image.detail = options.detail;
    }
    return image;
  },
  reasoning(text, id) {
    return id === void 0 ? { type: "reasoning", text } : { type: "reasoning", text, id };
  },
  reasoningFromContent(content, id) {
    const text = reasoningDisplayText(content);
    const reasoning = { type: "reasoning", text, content };
    return id === void 0 ? reasoning : { ...reasoning, id };
  },
  reasoningSummary(text, id) {
    return AssistantContent.reasoningFromContent([{ type: "summary", text }], id);
  },
  reasoningEncrypted(data, id) {
    return AssistantContent.reasoningFromContent([{ type: "encrypted", data }], id);
  },
  reasoningRedacted(data, id) {
    return AssistantContent.reasoningFromContent([{ type: "redacted", data }], id);
  },
  toolCall(id, name, args, callId) {
    const base = {
      type: "tool_call",
      id,
      function: {
        name,
        arguments: args
      }
    };
    return callId === void 0 ? base : { ...base, callId };
  }
};
function reasoningDisplayText(reasoning) {
  const content = Array.isArray(reasoning) ? reasoning : reasoning.content;
  if (content === void 0) {
    return Array.isArray(reasoning) ? "" : reasoning.text;
  }
  return content.flatMap((item) => {
    if (item.type === "text" || item.type === "summary") {
      return [item.text];
    }
    return [];
  }).join("");
}
var Message = {
  system(content) {
    return { role: "system", content };
  },
  user(content) {
    return {
      role: "user",
      content: typeof content === "string" ? [UserContent.text(content)] : content
    };
  },
  assistant(content, id) {
    const normalized = typeof content === "string" ? [AssistantContent.text(content)] : content;
    return id === void 0 ? { role: "assistant", content: normalized } : { role: "assistant", id, content: normalized };
  },
  tool(content) {
    return {
      role: "tool",
      content: Array.isArray(content) ? content : [content]
    };
  },
  toolResult(id, output, options = {}) {
    const content = isToolResultContentArray(output) ? output : serializeToolResultOutput(output);
    return Message.tool(ToolContent.toolResult(id, content, options));
  }
};
var Usage = {
  empty() {
    return {
      inputTokens: 0,
      outputTokens: 0,
      totalTokens: 0,
      cachedInputTokens: 0,
      cacheCreationInputTokens: 0
    };
  },
  add(left, right) {
    return {
      inputTokens: left.inputTokens + right.inputTokens,
      outputTokens: left.outputTokens + right.outputTokens,
      totalTokens: left.totalTokens + right.totalTokens,
      cachedInputTokens: left.cachedInputTokens + right.cachedInputTokens,
      cacheCreationInputTokens: left.cacheCreationInputTokens + right.cacheCreationInputTokens
    };
  }
};
var CompletionCapabilityError = class extends Error {
  constructor(message) {
    super(message);
    this.name = "CompletionCapabilityError";
  }
};
function assertCompletionRequestSupported(model, request, options = {}) {
  const modelLabel = `${model.provider}:${request.model ?? model.defaultModel}`;
  const capabilities = model.capabilities;
  if (options.streaming === true && !capabilities.streaming) {
    throw new CompletionCapabilityError(`${modelLabel} does not support streaming completions.`);
  }
  if (request.tools.length > 0 && !capabilities.tools) {
    throw new CompletionCapabilityError(`${modelLabel} does not support tool definitions.`);
  }
  if (request.toolChoice !== void 0 && !capabilities.toolChoice) {
    throw new CompletionCapabilityError(`${modelLabel} does not support tool choice.`);
  }
  if (request.outputSchema !== void 0 && !capabilities.outputSchema) {
    throw new CompletionCapabilityError(`${modelLabel} does not support output schemas.`);
  }
  if (!capabilities.imageInput && requestHasImageInput(request)) {
    throw new CompletionCapabilityError(`${modelLabel} does not support image input.`);
  }
  if (!capabilities.documentInput && requestHasFileDocumentInput(request)) {
    throw new CompletionCapabilityError(`${modelLabel} does not support document file input.`);
  }
}
function textFromAssistantContent(content) {
  return content.flatMap((item) => item.type === "text" ? [item.text] : []).join("\n");
}
function requestHasImageInput(request) {
  return request.chatHistory.some(
    (message) => message.role === "system" ? false : message.content.some((content) => content.type === "image")
  );
}
function requestHasFileDocumentInput(request) {
  return request.chatHistory.some(
    (message) => message.role === "user" ? message.content.some(
      (content) => content.type === "document" && content.source.type !== "text"
    ) : false
  );
}

export {
  UserContent,
  ToolContent,
  serializeToolResultOutput,
  isToolResultContentArray,
  AssistantContent,
  reasoningDisplayText,
  Message,
  Usage,
  CompletionCapabilityError,
  assertCompletionRequestSupported,
  textFromAssistantContent
};
