// @anvia/core@0.12.8/dist/chunk-WRNPY5C6.js
import {
  Agent,
  normalizeAgentId
} from "./chunk-2STANSN6.js";
import {
  ToolSet
} from "./chunk-SHORX7TZ.js";
import {
  resolveMemoryOptions
} from "./chunk-M6RPFHG7.js";
import {
  toProviderJsonSchema
} from "./chunk-WQKHFADH.js";
import {
  appendGuardrailPolicies
} from "./chunk-CWUJUSOS.js";

// src/agent/builder.ts
var AgentBuilder = class {
  constructor(agentId, completionModel) {
    this.completionModel = completionModel;
    this.agentId = normalizeAgentId(agentId);
  }
  completionModel;
  agentId;
  agentName;
  agentDescription;
  instructionBlocks = [];
  contextDocs = [];
  temp;
  maxTokenCount;
  params;
  choice;
  turns;
  requestHook;
  schema;
  approvalOptions;
  guardrailPolicies = [];
  skillInstructionBlocks = [];
  observerRegistrations = [];
  dynamicContextRegistrations = [];
  dynamicToolRegistrations = [];
  middlewareRegistrations = [];
  memoryRegistration;
  eventStoreRegistration;
  activeToolSet = new ToolSet();
  name(name) {
    this.agentName = name;
    return this;
  }
  description(description) {
    this.agentDescription = description;
    return this;
  }
  instructions(instructions) {
    if (instructions.length > 0) {
      this.instructionBlocks.push(instructions);
    }
    return this;
  }
  context(text, id = `static_doc_${this.contextDocs.length}`) {
    this.contextDocs.push({ id, text });
    return this;
  }
  dynamicContext(index, options) {
    this.dynamicContextRegistrations.push({ index, options });
    return this;
  }
  dynamicTools(index, options) {
    this.dynamicToolRegistrations.push({ index, options });
    return this;
  }
  tool(tool) {
    this.activeToolSet.addTool(tool);
    return this;
  }
  tools(tools) {
    this.activeToolSet.addTools(tools);
    return this;
  }
  mcp(servers) {
    for (const server of servers) {
      this.activeToolSet.addTools(server.tools);
    }
    return this;
  }
  skills(skillSet) {
    if (skillSet.instructions.length > 0) {
      this.skillInstructionBlocks.push(skillSet.instructions);
    }
    this.activeToolSet.addTools(skillSet.tools);
    return this;
  }
  useToolSet(toolSet) {
    toolSet.addTools(this.activeToolSet);
    this.activeToolSet = toolSet;
    return this;
  }
  temperature(temperature) {
    this.temp = temperature;
    return this;
  }
  maxTokens(maxTokens) {
    this.maxTokenCount = maxTokens;
    return this;
  }
  additionalParams(params) {
    this.params = params;
    return this;
  }
  toolChoice(toolChoice) {
    this.choice = toolChoice;
    return this;
  }
  defaultMaxTurns(defaultMaxTurns) {
    this.turns = defaultMaxTurns;
    return this;
  }
  hook(hook) {
    this.requestHook = hook;
    return this;
  }
  middleware(middleware) {
    this.middlewareRegistrations.push(middleware);
    return this;
  }
  middlewares(middlewares) {
    this.middlewareRegistrations.push(...middlewares);
    return this;
  }
  /**
   * @deprecated Use `middleware` instead.
   */
  toolMiddleware(middleware) {
    return this.middleware(middleware);
  }
  /**
   * @deprecated Use `middlewares` instead.
   */
  toolMiddlewares(middlewares) {
    return this.middlewares(middlewares);
  }
  observe(observer, options = {}) {
    this.observerRegistrations.push({
      observer,
      failOnObserverError: options.failOnObserverError
    });
    return this;
  }
  approvals(options) {
    this.approvalOptions = options;
    return this;
  }
  guardrails(policies) {
    this.guardrailPolicies = appendGuardrailPolicies(this.guardrailPolicies, policies);
    return this;
  }
  memory(store, options = {}) {
    this.memoryRegistration = {
      store,
      options: resolveMemoryOptions(options)
    };
    return this;
  }
  eventStore(store, options = {}) {
    this.eventStoreRegistration = {
      store,
      options: {
        include: options.include ?? "all"
      }
    };
    return this;
  }
  outputSchema(schema) {
    this.schema = toProviderJsonSchema(schema);
    return this;
  }
  build() {
    return new Agent({
      id: this.agentId,
      name: this.agentName,
      description: this.agentDescription,
      model: this.completionModel,
      instructions: this.buildInstructions(),
      staticContext: this.contextDocs,
      temperature: this.temp,
      maxTokens: this.maxTokenCount,
      additionalParams: this.params,
      toolSet: this.activeToolSet,
      toolChoice: this.choice,
      defaultMaxTurns: this.turns,
      hook: this.requestHook,
      outputSchema: this.schema,
      observers: this.observerRegistrations,
      approvals: this.approvalOptions,
      guardrails: this.guardrailPolicies,
      dynamicContexts: this.dynamicContextRegistrations,
      dynamicTools: this.dynamicToolRegistrations,
      middlewares: this.middlewareRegistrations,
      memory: this.memoryRegistration,
      eventStore: this.eventStoreRegistration
    });
  }
  buildInstructions() {
    const parts = [...this.instructionBlocks, ...this.skillInstructionBlocks].filter(
      (part) => part !== void 0 && part.length > 0
    );
    return parts.length === 0 ? void 0 : parts.join("\n\n");
  }
};

export {
  AgentBuilder
};
