// @anvia/core@0.12.8/dist/chunk-HF4KHHNM.js
import {
  markSkillTool
} from "./chunk-YK4WAAS4.js";
import {
  createTool
} from "./chunk-IZM6FNCT.js";
import {
  isRecord
} from "./chunk-MMHG7WAM.js";

// src/skills/instructions.ts
function skillInstructions(skills) {
  if (skills.length === 0) {
    return "";
  }
  return [
    "You have access to Agent Skills.",
    "Skills are compact capability packages. Use the skill tools to load full instructions, references, or scripts only when a skill is relevant.",
    "",
    "Available skills:",
    ...skills.map(formatSkill),
    "",
    "Skill tools:",
    "- get_skill_instructions: load full SKILL.md guidance for a skill.",
    "- get_skill_reference: read a reference file from a skill.",
    "- get_skill_script: read a script file from a skill.",
    "- run_skill_script: execute a script from a skill with arguments."
  ].join("\n");
}
function formatSkill(skill2) {
  const lines = [`- ${skill2.name}: ${skill2.description}`];
  if (skill2.references.length > 0) {
    lines.push(`  references: ${skill2.references.join(", ")}`);
  }
  if (skill2.scripts.length > 0) {
    lines.push(`  scripts: ${skill2.scripts.join(", ")}`);
  }
  return lines.join("\n");
}

// src/skills/tools.ts
import { spawn } from "child_process";
import { readFile } from "fs/promises";
import { isAbsolute, relative, resolve } from "path";
import { z } from "zod";
var DEFAULT_TIMEOUT_MS = 3e4;
var MAX_OUTPUT_CHARS = 2e4;
function createSkillTools(skills) {
  const registry = new SkillRegistry(skills);
  return [
    markSkillTool(
      createTool({
        name: "get_skill_instructions",
        description: "Load the full SKILL.md instructions for an Agent Skill.",
        input: z.object({
          skillName: z.string().describe("The name of the skill to load.")
        }),
        output: z.string(),
        execute: ({ skillName }) => registry.get(skillName).instructions
      })
    ),
    markSkillTool(
      createTool({
        name: "get_skill_reference",
        description: "Read a reference file from an Agent Skill.",
        input: z.object({
          skillName: z.string().describe("The name of the skill."),
          referencePath: z.string().describe("A path listed in the skill references.")
        }),
        output: z.string(),
        execute: ({ skillName, referencePath }) => registry.readReference(skillName, referencePath)
      })
    ),
    markSkillTool(
      createTool({
        name: "get_skill_script",
        description: "Read a script file from an Agent Skill.",
        input: z.object({
          skillName: z.string().describe("The name of the skill."),
          scriptPath: z.string().describe("A path listed in the skill scripts.")
        }),
        output: z.string(),
        execute: ({ skillName, scriptPath }) => registry.readScript(skillName, scriptPath)
      })
    ),
    markSkillTool(
      createTool({
        name: "run_skill_script",
        description: "Execute a script from an Agent Skill with optional arguments.",
        input: z.object({
          skillName: z.string().describe("The name of the skill."),
          scriptPath: z.string().describe("A path listed in the skill scripts."),
          args: z.array(z.string()).optional().describe("Arguments passed to the script."),
          timeoutMs: z.number().int().positive().optional().describe("Execution timeout in ms.")
        }),
        output: z.string(),
        execute: ({ skillName, scriptPath, args = [], timeoutMs = DEFAULT_TIMEOUT_MS }) => registry.runScript(skillName, scriptPath, args, timeoutMs)
      })
    )
  ];
}
var SkillRegistry = class {
  skills = /* @__PURE__ */ new Map();
  constructor(skills) {
    for (const skill2 of skills) {
      this.skills.set(skill2.name, skill2);
    }
  }
  get(skillName) {
    const skill2 = this.skills.get(skillName);
    if (skill2 === void 0) {
      throw new Error(`Skill not found: ${skillName}`);
    }
    return skill2;
  }
  async readReference(skillName, referencePath) {
    const skill2 = this.get(skillName);
    const path = this.resolveContainedPath(skill2, "references", referencePath);
    if (!skill2.references.includes(referencePath)) {
      throw new Error(`Skill reference not found: ${skillName}/${referencePath}`);
    }
    return readFile(path, "utf8");
  }
  async readScript(skillName, scriptPath) {
    const skill2 = this.get(skillName);
    const path = this.resolveContainedPath(skill2, "scripts", scriptPath);
    if (!skill2.scripts.includes(scriptPath)) {
      throw new Error(`Skill script not found: ${skillName}/${scriptPath}`);
    }
    return readFile(path, "utf8");
  }
  async runScript(skillName, scriptPath, args, timeoutMs) {
    const skill2 = this.get(skillName);
    const script = this.resolveContainedPath(skill2, "scripts", scriptPath);
    if (!skill2.scripts.includes(scriptPath)) {
      throw new Error(`Skill script not found: ${skillName}/${scriptPath}`);
    }
    return runExecutable(script, args, skill2.directory, timeoutMs);
  }
  resolveContainedPath(skill2, section, requestedPath) {
    if (requestedPath.length === 0 || isAbsolute(requestedPath)) {
      throw new Error(`Invalid skill path: ${requestedPath}`);
    }
    const root = resolve(skill2.directory, section);
    const resolved = resolve(root, requestedPath);
    const rel = relative(root, resolved);
    if (rel === ".." || rel.startsWith(`..${"/"}`) || rel.startsWith(`..${"\\"}`)) {
      throw new Error(`Invalid skill path: ${requestedPath}`);
    }
    return resolved;
  }
};
function runExecutable(command, args, cwd, timeoutMs) {
  return new Promise((resolvePromise, reject) => {
    const child = spawn(command, args, {
      cwd,
      shell: false,
      stdio: ["ignore", "pipe", "pipe"]
    });
    let stdout = "";
    let stderr = "";
    let timedOut = false;
    const timeout = setTimeout(() => {
      timedOut = true;
      child.kill();
    }, timeoutMs);
    child.stdout.setEncoding("utf8");
    child.stderr.setEncoding("utf8");
    child.stdout.on("data", (chunk) => {
      stdout = appendLimited(stdout, chunk);
    });
    child.stderr.on("data", (chunk) => {
      stderr = appendLimited(stderr, chunk);
    });
    child.on("error", (error) => {
      clearTimeout(timeout);
      reject(error);
    });
    child.on("close", (code, signal) => {
      clearTimeout(timeout);
      if (timedOut) {
        reject(new Error(`Skill script timed out after ${timeoutMs}ms`));
        return;
      }
      const output = formatProcessOutput(stdout, stderr);
      if (code !== 0) {
        reject(new Error(`Skill script exited with code ${code ?? "unknown"}: ${output}`));
        return;
      }
      if (signal !== null) {
        reject(new Error(`Skill script exited with signal ${signal}: ${output}`));
        return;
      }
      resolvePromise(output);
    });
  });
}
function formatProcessOutput(stdout, stderr) {
  const parts = [];
  if (stdout.length > 0) {
    parts.push(`stdout:
${stdout}`);
  }
  if (stderr.length > 0) {
    parts.push(`stderr:
${stderr}`);
  }
  return parts.length === 0 ? "" : parts.join("\n\n");
}
function appendLimited(current, chunk) {
  const next = current + chunk;
  if (next.length <= MAX_OUTPUT_CHARS) {
    return next;
  }
  return `${next.slice(0, MAX_OUTPUT_CHARS)}
[truncated]`;
}

// src/skills/load.ts
async function loadSkills(loaders) {
  const ordered = Array.isArray(loaders) ? loaders : [loaders];
  const byName = /* @__PURE__ */ new Map();
  for (const loader of ordered) {
    for (const loaded of await loader.load()) {
      byName.delete(loaded.name);
      byName.set(loaded.name, loaded);
    }
  }
  const skills = [...byName.values()];
  return {
    skills,
    instructions: skillInstructions(skills),
    tools: skills.length === 0 ? [] : createSkillTools(skills)
  };
}

// src/skills/local.ts
import { readdir, readFile as readFile2, stat } from "fs/promises";
import { basename, join, relative as relative2, resolve as resolve2, sep } from "path";
import { parse as parseYaml } from "yaml";

// src/skills/types.ts
var SkillValidationError = class extends Error {
  issues;
  constructor(message, issues) {
    super(message);
    this.name = "SkillValidationError";
    this.issues = issues;
  }
};

// src/skills/local.ts
var NAME_PATTERN = /^[a-z0-9]+(?:-[a-z0-9]+)*$/;
var MAX_NAME_LENGTH = 64;
var MAX_DESCRIPTION_LENGTH = 1024;
var skill = {
  local(path) {
    return {
      load: () => loadLocalSkills(path)
    };
  }
};
async function loadLocalSkills(path) {
  const root = resolve2(path);
  if (await hasSkillFile(root)) {
    return [await readSkill(root)];
  }
  const entries = await readdir(root, { withFileTypes: true });
  const skills = [];
  for (const entry of entries) {
    if (!entry.isDirectory()) {
      continue;
    }
    const directory = join(root, entry.name);
    if (await hasSkillFile(directory)) {
      skills.push(await readSkill(directory));
    }
  }
  return skills;
}
async function hasSkillFile(directory) {
  try {
    const info = await stat(join(directory, "SKILL.md"));
    return info.isFile();
  } catch {
    return false;
  }
}
async function readSkill(directory) {
  const skillPath = join(directory, "SKILL.md");
  const markdown = await readFile2(skillPath, "utf8");
  const parsed = parseSkillMarkdown(markdown, skillPath);
  validateSkillFrontmatter(parsed.frontmatter, directory, skillPath);
  const name = parsed.frontmatter.name;
  const description = parsed.frontmatter.description;
  const license = typeof parsed.frontmatter.license === "string" ? parsed.frontmatter.license : void 0;
  const metadata = isRecord(parsed.frontmatter.metadata) ? parsed.frontmatter.metadata : void 0;
  return {
    name,
    description,
    instructions: parsed.body.trim(),
    directory,
    references: await listRelativeFiles(join(directory, "references")),
    scripts: await listRelativeFiles(join(directory, "scripts")),
    license,
    metadata
  };
}
function parseSkillMarkdown(markdown, path) {
  if (!markdown.startsWith("---\n") && !markdown.startsWith("---\r\n")) {
    throw new SkillValidationError("Skill validation failed", [
      { path, message: "SKILL.md must start with YAML frontmatter" }
    ]);
  }
  const newline = markdown.startsWith("---\r\n") ? "\r\n" : "\n";
  const marker = `${newline}---${newline}`;
  const end = markdown.indexOf(marker, 3);
  if (end === -1) {
    throw new SkillValidationError("Skill validation failed", [
      { path, message: "SKILL.md frontmatter must end with ---" }
    ]);
  }
  const rawFrontmatter = markdown.slice(3 + newline.length, end);
  const body = markdown.slice(end + marker.length);
  const parsed = parseYaml(rawFrontmatter);
  if (!isRecord(parsed)) {
    throw new SkillValidationError("Skill validation failed", [
      { path, message: "SKILL.md frontmatter must be a YAML object" }
    ]);
  }
  return { frontmatter: parsed, body };
}
function validateSkillFrontmatter(frontmatter, directory, path) {
  const issues = [];
  const name = frontmatter.name;
  const description = frontmatter.description;
  if (typeof name !== "string" || name.length === 0) {
    issues.push({ path, message: "name is required" });
  } else {
    if (name.length > MAX_NAME_LENGTH) {
      issues.push({ path, message: `name must be at most ${MAX_NAME_LENGTH} characters` });
    }
    if (!NAME_PATTERN.test(name)) {
      issues.push({
        path,
        message: "name must contain lowercase letters, numbers, and hyphens only"
      });
    }
    if (basename(directory) !== name) {
      issues.push({ path, message: "name must match the skill directory name" });
    }
  }
  if (typeof description !== "string" || description.length === 0) {
    issues.push({ path, message: "description is required" });
  } else if (description.length > MAX_DESCRIPTION_LENGTH) {
    issues.push({
      path,
      message: `description must be at most ${MAX_DESCRIPTION_LENGTH} characters`
    });
  }
  if (issues.length > 0) {
    throw new SkillValidationError("Skill validation failed", issues);
  }
}
async function listRelativeFiles(directory) {
  try {
    const info = await stat(directory);
    if (!info.isDirectory()) {
      return [];
    }
  } catch {
    return [];
  }
  const files = [];
  await collectFiles(directory, directory, files);
  return files.sort();
}
async function collectFiles(root, directory, files) {
  const entries = await readdir(directory, { withFileTypes: true });
  for (const entry of entries) {
    const path = join(directory, entry.name);
    if (entry.isDirectory()) {
      await collectFiles(root, path, files);
    } else if (entry.isFile()) {
      files.push(toPortablePath(relative2(root, path)));
    }
  }
}
function toPortablePath(path) {
  return sep === "/" ? path : path.split(sep).join("/");
}

export {
  loadSkills,
  SkillValidationError,
  skill
};
