// @anvia/core@0.12.8/dist/chunk-UDISNNKR.js
import {
  AssistantContent,
  Message,
  ToolContent,
  UserContent,
  reasoningDisplayText,
  serializeToolResultOutput
} from "./chunk-QVFOCNF6.js";

// src/ui/messages.ts
function uiMessagesToCoreMessages(messages) {
  const coreMessages = [];
  for (const message of messages) {
    const text = textFromUIParts(message.parts);
    if (message.role === "system") {
      if (text.length > 0) {
        coreMessages.push(Message.system(text));
      }
      continue;
    }
    if (message.role === "user") {
      const content = [];
      for (const part of message.parts) {
        if (part.type === "text") {
          content.push(UserContent.text(part.text));
          continue;
        }
        if (part.type === "attachment") {
          content.push(attachmentToUserContent(part.attachment));
          continue;
        }
        if (part.type === "data" && isUserContent(part.data)) {
          content.push(part.data);
          continue;
        }
        throw new TypeError(
          "User UI messages can only be converted from text, attachment, or image/document data parts."
        );
      }
      if (content.length > 0) {
        coreMessages.push(Message.user(content));
      }
      continue;
    }
    if (message.role === "assistant") {
      const content = [];
      const toolResults2 = [];
      for (const part of message.parts) {
        if (part.type === "text" && part.text.length > 0) {
          content.push(AssistantContent.text(part.text));
          continue;
        }
        if (part.type === "reasoning" && part.text.length > 0) {
          content.push(AssistantContent.reasoning(part.text, part.reasoningId));
          continue;
        }
        if (part.type === "tool" && (part.state === "input-streaming" || part.state === "input-available" || part.state === "output-available")) {
          content.push(
            AssistantContent.toolCall(
              part.toolCallId,
              part.toolName,
              part.input ?? {},
              part.callId
            )
          );
          if (part.state === "output-available") {
            toolResults2.push(
              ToolContent.toolResult(part.toolCallId, outputToToolResultContent(part.output), {
                callId: part.callId,
                toolName: part.toolName
              })
            );
          }
        }
      }
      if (content.length > 0) {
        coreMessages.push(Message.assistant(content, message.id));
      }
      if (toolResults2.length > 0) {
        coreMessages.push(Message.tool(toolResults2));
      }
      continue;
    }
    const toolResults = [];
    for (const part of message.parts) {
      if (part.type !== "tool" || part.state !== "output-available") {
        continue;
      }
      toolResults.push(
        ToolContent.toolResult(part.toolCallId, outputToToolResultContent(part.output), {
          callId: part.callId,
          toolName: part.toolName
        })
      );
    }
    if (toolResults.length > 0) {
      coreMessages.push(Message.tool(toolResults));
    }
  }
  return coreMessages;
}
function coreMessagesToUIMessages(messages) {
  const uiMessages = [];
  const toolPartLocations = {
    byToolCallId: /* @__PURE__ */ new Map(),
    byCallId: /* @__PURE__ */ new Map()
  };
  for (const message of messages) {
    if (message.role === "system") {
      uiMessages.push({
        id: createId("msg"),
        role: "system",
        parts: [{ id: createId("part"), type: "text", text: message.content }]
      });
      continue;
    }
    if (message.role === "user") {
      const parts2 = [];
      for (const content of message.content) {
        if (content.type === "text") {
          parts2.push({ id: createId("part"), type: "text", text: content.text });
        } else {
          parts2.push({
            id: createId("part"),
            type: "attachment",
            attachment: userContentToAttachment(content)
          });
        }
      }
      uiMessages.push({ id: createId("msg"), role: "user", parts: parts2 });
      continue;
    }
    if (message.role === "assistant") {
      const parts2 = [];
      for (const content of message.content) {
        if (content.type === "text") {
          parts2.push({ id: createId("part"), type: "text", text: content.text });
          continue;
        }
        if (content.type === "reasoning") {
          const part = {
            id: createId("part"),
            type: "reasoning",
            text: reasoningDisplayText(content)
          };
          if (content.id !== void 0) {
            part.reasoningId = content.id;
          }
          parts2.push(part);
          continue;
        }
        if (content.type === "tool_call") {
          const part = {
            id: toolPartId(content.id),
            type: "tool",
            toolName: content.function.name,
            toolCallId: content.id,
            state: "input-available",
            input: content.function.arguments
          };
          if (content.callId !== void 0) {
            part.callId = content.callId;
          }
          parts2.push(part);
          continue;
        }
        parts2.push({
          id: createId("part"),
          type: "data",
          name: content.type,
          data: content
        });
      }
      const messageIndex = uiMessages.length;
      uiMessages.push({ id: message.id ?? createId("msg"), role: "assistant", parts: parts2 });
      registerToolPartLocations(parts2, messageIndex, toolPartLocations);
      continue;
    }
    const parts = [];
    for (const content of message.content) {
      const part = {
        id: toolPartId(content.id),
        type: "tool",
        toolName: toolNameForResult(content, uiMessages, toolPartLocations),
        toolCallId: content.id,
        state: "output-available",
        output: toolResultContentToJson(content.content)
      };
      if (content.callId !== void 0) {
        part.callId = content.callId;
      }
      if (mergeToolResultPart(uiMessages, toolPartLocations, part)) {
        continue;
      }
      parts.push(part);
    }
    if (parts.length > 0) {
      uiMessages.push({ id: createId("msg"), role: "tool", parts });
    }
  }
  return uiMessages;
}
function registerToolPartLocations(parts, messageIndex, toolPartLocations) {
  for (const [partIndex, part] of parts.entries()) {
    if (part.type !== "tool") {
      continue;
    }
    const location = { messageIndex, partIndex };
    toolPartLocations.byToolCallId.set(part.toolCallId, location);
    if (part.callId !== void 0) {
      toolPartLocations.byCallId.set(part.callId, location);
    }
  }
}
function mergeToolResultPart(uiMessages, toolPartLocations, resultPart) {
  if (resultPart.type !== "tool") {
    return false;
  }
  const location = findToolPartLocation(
    resultPart.toolCallId,
    resultPart.callId,
    toolPartLocations
  );
  if (location === void 0) {
    return false;
  }
  const message = uiMessages[location.messageIndex];
  const currentPart = message?.parts[location.partIndex];
  if (message === void 0 || currentPart?.type !== "tool") {
    return false;
  }
  const mergedPart = {
    ...currentPart,
    toolName: resultPart.toolName === "tool" ? currentPart.toolName : resultPart.toolName,
    state: "output-available",
    ...resultPart.output === void 0 ? {} : { output: resultPart.output }
  };
  if (currentPart.callId === void 0 && resultPart.callId !== void 0) {
    mergedPart.callId = resultPart.callId;
    toolPartLocations.byCallId.set(resultPart.callId, location);
  }
  const nextParts = [...message.parts];
  nextParts[location.partIndex] = mergedPart;
  uiMessages[location.messageIndex] = { ...message, parts: nextParts };
  return true;
}
function findToolPartLocation(toolCallId, callId, toolPartLocations) {
  return (callId === void 0 ? void 0 : toolPartLocations.byCallId.get(callId)) ?? toolPartLocations.byToolCallId.get(toolCallId);
}
function toolNameForResult(content, uiMessages, toolPartLocations) {
  if (content.toolName !== void 0) {
    return content.toolName;
  }
  const location = findToolPartLocation(content.id, content.callId, toolPartLocations);
  const part = location === void 0 ? void 0 : uiMessages[location.messageIndex]?.parts[location.partIndex];
  return part?.type === "tool" ? part.toolName : "tool";
}
function textFromUIParts(parts) {
  return parts.flatMap((part) => part.type === "text" ? [part.text] : []).join("");
}
function outputToToolResultContent(output) {
  return [{ type: "text", text: serializeToolResultOutput(output ?? null) }];
}
function attachmentToUserContent(attachment) {
  if (attachment.type === "image" || isImageFileAttachment(attachment)) {
    if (attachment.url !== void 0) {
      return UserContent.imageUrl(attachment.url, imageOptions(attachment));
    }
    if (attachment.data !== void 0 && attachment.mediaType !== void 0) {
      return UserContent.imageBase64(
        attachment.data,
        attachment.mediaType,
        imageOptions(attachment)
      );
    }
    throw new TypeError("Image attachments require a url or base64 data with mediaType.");
  }
  if (attachment.text !== void 0) {
    return {
      type: "document",
      source: {
        type: "text",
        text: attachment.text,
        ...attachment.mediaType === void 0 ? {} : { mediaType: attachment.mediaType },
        ...attachment.name === void 0 ? {} : { filename: attachment.name }
      }
    };
  }
  const mediaType = attachment.mediaType ?? "application/octet-stream";
  if (attachment.url !== void 0) {
    return UserContent.documentUrl(attachment.url, mediaType, documentOptions(attachment));
  }
  if (attachment.data !== void 0) {
    return UserContent.documentBase64(attachment.data, mediaType, documentOptions(attachment));
  }
  throw new TypeError("Document attachments require a url, base64 data, or text.");
}
function userContentToAttachment(content) {
  if (content.type === "image") {
    const attachment2 = {
      id: createId("attachment"),
      type: "image",
      ...content.detail === void 0 ? {} : { detail: content.detail }
    };
    if (content.source.type === "url") {
      attachment2.url = content.source.url;
    } else {
      attachment2.data = content.source.data;
      attachment2.mediaType = content.source.mediaType;
    }
    return attachment2;
  }
  const attachment = {
    id: createId("attachment"),
    type: "document"
  };
  if (content.source.type === "url") {
    attachment.url = content.source.url;
    attachment.mediaType = content.source.mediaType;
    if (content.source.filename !== void 0) {
      attachment.name = content.source.filename;
    }
  } else if (content.source.type === "base64") {
    attachment.data = content.source.data;
    attachment.mediaType = content.source.mediaType;
    if (content.source.filename !== void 0) {
      attachment.name = content.source.filename;
    }
  } else {
    attachment.text = content.source.text;
    if (content.source.mediaType !== void 0) {
      attachment.mediaType = content.source.mediaType;
    }
    if (content.source.filename !== void 0) {
      attachment.name = content.source.filename;
    }
  }
  return attachment;
}
function isImageFileAttachment(attachment) {
  return attachment.type === "file" && attachment.mediaType?.startsWith("image/") === true;
}
function imageOptions(attachment) {
  return attachment.detail === void 0 ? {} : { detail: attachment.detail };
}
function documentOptions(attachment) {
  return attachment.name === void 0 ? {} : { filename: attachment.name };
}
function toolResultContentToJson(content) {
  if (content.length === 1 && content[0]?.type === "text") {
    return content[0].text;
  }
  return content;
}
function isUserContent(value) {
  if (!isRecord(value)) {
    return false;
  }
  if (value.type === "text") {
    return typeof value.text === "string";
  }
  if (value.type === "image") {
    return isImageContent(value);
  }
  if (value.type === "document") {
    return isDocumentContent(value);
  }
  return false;
}
function isImageContent(value) {
  if (!isRecord(value.source)) {
    return false;
  }
  if (value.source.type === "url") {
    return typeof value.source.url === "string";
  }
  if (value.source.type === "base64") {
    return typeof value.source.data === "string" && typeof value.source.mediaType === "string";
  }
  return false;
}
function isDocumentContent(value) {
  if (!isRecord(value.source)) {
    return false;
  }
  if (value.source.type === "url") {
    return typeof value.source.url === "string" && typeof value.source.mediaType === "string";
  }
  if (value.source.type === "base64") {
    return typeof value.source.data === "string" && typeof value.source.mediaType === "string";
  }
  if (value.source.type === "text") {
    return typeof value.source.text === "string" && (value.source.mediaType === void 0 || typeof value.source.mediaType === "string");
  }
  return false;
}
function toolPartId(toolCallId) {
  return `tool_${toolCallId}`;
}
var nextId = 0;
function createId(prefix) {
  const random = globalThis.crypto?.randomUUID?.();
  if (random !== void 0) {
    return `${prefix}_${random}`;
  }
  nextId += 1;
  return `${prefix}_${nextId.toString(36)}`;
}
function isRecord(value) {
  return typeof value === "object" && value !== null;
}

export {
  uiMessagesToCoreMessages,
  coreMessagesToUIMessages
};
