// @anvia/react@0.8.5/dist/index.js
// src/direct.ts
function createDirectTransport(handler) {
  return {
    async *send(request, options) {
      const iterable = handler(request);
      const iterator = iterable[Symbol.asyncIterator]();
      try {
        while (true) {
          if (options?.signal?.aborted) {
            break;
          }
          const next = await iterator.next();
          if (next.done) {
            break;
          }
          yield next.value;
        }
      } finally {
        await iterator.return?.();
      }
    }
  };
}

// src/streams.ts
async function* readJsonlStream(stream) {
  const reader = stream.getReader();
  const decoder = new TextDecoder();
  let buffer = "";
  try {
    while (true) {
      const next = await reader.read();
      if (next.done === true) {
        break;
      }
      buffer += decoder.decode(next.value, { stream: true });
      const lines = buffer.split(/\r?\n/);
      buffer = lines.pop() ?? "";
      for (const line of lines) {
        const trimmed2 = line.trim();
        if (trimmed2.length > 0) {
          yield JSON.parse(trimmed2);
        }
      }
    }
    buffer += decoder.decode();
    const trimmed = buffer.trim();
    if (trimmed.length > 0) {
      yield JSON.parse(trimmed);
    }
  } finally {
    reader.releaseLock();
  }
}
async function* readSseStream(stream) {
  const reader = stream.getReader();
  const decoder = new TextDecoder();
  let buffer = "";
  let event = createEmptySseEvent();
  try {
    while (true) {
      const next = await reader.read();
      if (next.done === true) {
        break;
      }
      buffer += decoder.decode(next.value, { stream: true });
      const lines = buffer.split(/\r?\n/);
      buffer = lines.pop() ?? "";
      for (const line of lines) {
        const parsed = parseSseLine(line, event);
        event = parsed.event;
        if (parsed.complete === true && parsed.data !== void 0) {
          yield JSON.parse(parsed.data);
        }
      }
    }
    buffer += decoder.decode();
    if (buffer.length > 0) {
      const parsed = parseSseLine(buffer, event);
      event = parsed.event;
      if (parsed.complete === true && parsed.data !== void 0) {
        yield JSON.parse(parsed.data);
      }
    }
    const data = flushSseEvent(event);
    if (data !== void 0) {
      yield JSON.parse(data);
    }
  } finally {
    reader.releaseLock();
  }
}
function createEmptySseEvent() {
  return { data: [] };
}
function parseSseLine(line, event) {
  if (line === "") {
    const data = flushSseEvent(event);
    return data === void 0 ? { event: createEmptySseEvent(), complete: true } : { event: createEmptySseEvent(), complete: true, data };
  }
  if (line.startsWith(":")) {
    return { event };
  }
  const separator = line.indexOf(":");
  const field = separator === -1 ? line : line.slice(0, separator);
  const value = separator === -1 ? "" : line.slice(separator + 1).replace(/^ /, "");
  if (field === "data") {
    event.data.push(value);
  }
  return { event };
}
function flushSseEvent(event) {
  if (event.data.length === 0) {
    return void 0;
  }
  return event.data.join("\n");
}

// src/fetch.ts
var EventStreamHttpError = class extends Error {
  constructor(response, body) {
    super(`Event stream request failed with status ${response.status}`);
    this.response = response;
    this.body = body;
    this.name = "EventStreamHttpError";
  }
  response;
  body;
};
async function* fetchEventStream(input, options = {}) {
  const fetchImpl = options.fetch ?? globalThis.fetch;
  if (fetchImpl === void 0) {
    throw new Error("fetchEventStream requires a fetch implementation");
  }
  const response = await fetchImpl(input, fetchOptions(options));
  if (!response.ok) {
    throw new EventStreamHttpError(response, await response.text());
  }
  if (response.body === null) {
    throw new Error("Event stream response does not include a body");
  }
  const format = options.format ?? inferEventStreamFormat(response.headers.get("content-type"));
  if (format === "sse") {
    yield* readSseStream(response.body);
    return;
  }
  yield* readJsonlStream(response.body);
}
function fetchOptions(options) {
  const { format: _format, fetch: _fetch, ...init } = options;
  return init;
}
function inferEventStreamFormat(contentType) {
  return contentType?.toLowerCase().includes("text/event-stream") ? "sse" : "jsonl";
}

// src/human-input.ts
function defaultEventToApproval(event) {
  if (!isRecord(event)) {
    return void 0;
  }
  if (event.type !== "tool_approval_request" && event.type !== "tool_approval_result") {
    return void 0;
  }
  return isToolApproval(event.approval) ? event.approval : void 0;
}
function defaultEventToQuestion(event) {
  if (!isRecord(event)) {
    return void 0;
  }
  if (event.type !== "tool_question_request" && event.type !== "tool_question_result") {
    return void 0;
  }
  return isToolQuestion(event.question) ? event.question : void 0;
}
async function defaultDecideApproval(input, options) {
  const endpoint = requireEndpoint(options.endpoint, "decideApproval");
  const fetchImpl = options.fetch ?? globalThis.fetch;
  if (fetchImpl === void 0) {
    throw new Error("humanInput approval decisions require a fetch implementation");
  }
  const response = await fetchImpl(
    endpointUrl(endpoint, `approvals/${input.approvalId}/decision`),
    {
      method: "POST",
      headers: { "content-type": "application/json" },
      body: JSON.stringify({
        approved: input.approved,
        ...input.reason === void 0 ? {} : { reason: input.reason }
      })
    }
  );
  if (!response.ok) {
    throw new Error(`Tool approval decision failed with status ${response.status}`);
  }
  return responseJson(response);
}
async function defaultAnswerQuestion(input, options) {
  const endpoint = requireEndpoint(options.endpoint, "answerQuestion");
  const fetchImpl = options.fetch ?? globalThis.fetch;
  if (fetchImpl === void 0) {
    throw new Error("humanInput question answers require a fetch implementation");
  }
  const response = await fetchImpl(endpointUrl(endpoint, `questions/${input.questionId}/answer`), {
    method: "POST",
    headers: { "content-type": "application/json" },
    body: JSON.stringify({ answers: input.answers })
  });
  if (!response.ok) {
    throw new Error(`Tool question answer failed with status ${response.status}`);
  }
  return responseJson(response);
}
function upsertById(items, item) {
  const index = items.findIndex((current) => current.id === item.id);
  if (index === -1) {
    return [...items, item];
  }
  const next = [...items];
  next[index] = item;
  return next;
}
function requireEndpoint(endpoint, operation) {
  if (endpoint === void 0) {
    throw new Error(`humanInput.${operation} requires endpoint or a custom handler`);
  }
  return endpoint;
}
function endpointUrl(endpoint, path) {
  const suffix = path.replace(/^\/+/, "");
  if (endpoint instanceof URL) {
    const url = new URL(endpoint.toString());
    url.pathname = joinPath(url.pathname, suffix);
    return url;
  }
  if (/^[a-z][a-z\d+\-.]*:\/\//i.test(endpoint)) {
    const url = new URL(endpoint);
    url.pathname = joinPath(url.pathname, suffix);
    return url.toString();
  }
  return joinPath(endpoint, suffix);
}
function joinPath(base, suffix) {
  const trimmedBase = base.replace(/\/+$/, "");
  const trimmedSuffix = suffix.replace(/^\/+/, "");
  if (trimmedBase.length === 0) {
    return `/${trimmedSuffix}`;
  }
  return `${trimmedBase}/${trimmedSuffix}`;
}
async function responseJson(response) {
  const text = await response.text();
  if (text.trim().length === 0) {
    return void 0;
  }
  return JSON.parse(text);
}
function isToolApproval(value) {
  return isRecord(value) && typeof value.id === "string" && typeof value.toolName === "string" && typeof value.status === "string";
}
function isToolQuestion(value) {
  return isRecord(value) && typeof value.id === "string" && typeof value.toolName === "string" && typeof value.status === "string" && Array.isArray(value.questions);
}
function isRecord(value) {
  return typeof value === "object" && value !== null;
}

// src/memory.ts
import { coreMessagesToUIMessages } from "@anvia/core/ui";
function initialMessagesFromMemory(messages) {
  return coreMessagesToUIMessages(messages);
}

// src/transport.ts
function createFetchTransport(options) {
  return {
    async *send(request, transportOptions = {}) {
      const endpoint = typeof options.endpoint === "function" ? options.endpoint(request) : options.endpoint;
      const requestHeaders = await resolveHeaders(options.headers, request);
      const headers = mergeHeaders(requestHeaders, transportOptions.headers);
      const method = options.method ?? "POST";
      const body = await resolveBody(options.body, request, headers, method);
      const init = {
        ...options.init ?? {},
        method,
        headers,
        format: options.format ?? "jsonl"
      };
      if (body !== void 0) {
        init.body = body;
      }
      if (transportOptions.signal !== void 0) {
        init.signal = transportOptions.signal;
      }
      if (options.fetch !== void 0) {
        init.fetch = options.fetch;
      }
      for await (const event of fetchEventStream(endpoint, init)) {
        yield options.mapEvent === void 0 ? event : options.mapEvent(event);
      }
    }
  };
}
function createChatTransport(options) {
  return createFetchTransport(options);
}
async function resolveHeaders(headers, request) {
  return typeof headers === "function" ? headers(request) : headers;
}
async function resolveBody(body, request, headers, method) {
  if (body !== void 0) {
    return body(request);
  }
  if (!methodAllowsDefaultBody(method)) {
    return void 0;
  }
  if (!headers.has("content-type")) {
    headers.set("content-type", "application/json");
  }
  return JSON.stringify(request);
}
function methodAllowsDefaultBody(method) {
  const normalized = method.toUpperCase();
  return normalized !== "GET" && normalized !== "HEAD";
}
function mergeHeaders(...values) {
  const headers = new Headers();
  for (const value of values) {
    if (value === void 0) {
      continue;
    }
    new Headers(value).forEach((headerValue, key) => {
      headers.set(key, headerValue);
    });
  }
  return headers;
}

// src/use-chat.ts
import {
  uiMessagesToCoreMessages
} from "@anvia/core/ui";
import { useCallback, useEffect, useMemo, useRef, useState } from "react";

// src/resume.ts
var storageKeyPrefix = "anvia:chat-resume:";
function loadChatResumeState(options) {
  const storage = resolveResumeStorage(options);
  const key = resumeStorageKey(options);
  if (storage === void 0 || key === void 0) {
    return void 0;
  }
  let raw;
  try {
    raw = storage.getItem(key);
  } catch {
    return void 0;
  }
  if (raw === null) {
    return void 0;
  }
  try {
    const parsed = JSON.parse(raw);
    return isChatResumeState(parsed) ? parsed : void 0;
  } catch {
    return void 0;
  }
}
function saveChatResumeState(options, state) {
  const storage = resolveResumeStorage(options);
  const key = resumeStorageKey(options);
  if (storage === void 0 || key === void 0) {
    return;
  }
  try {
    storage.setItem(key, JSON.stringify(state));
  } catch {
  }
}
function clearChatResumeState(options) {
  const storage = resolveResumeStorage(options);
  const key = resumeStorageKey(options);
  if (storage === void 0 || key === void 0) {
    return;
  }
  try {
    storage.removeItem(key);
  } catch {
  }
}
function isResumableStreamEnvelope(value) {
  if (!isRecord2(value)) {
    return false;
  }
  if (value.type === "stream_start" && typeof value.streamId === "string" && value.eventId === 0) {
    return true;
  }
  if (value.type === "stream_event" && typeof value.streamId === "string" && typeof value.eventId === "number" && "event" in value) {
    return true;
  }
  return value.type === "stream_end" && typeof value.streamId === "string" && typeof value.eventId === "number" && typeof value.status === "string";
}
function resumeStorageKey(options) {
  if (options === void 0) {
    return void 0;
  }
  return `${storageKeyPrefix}${options.key}`;
}
function resolveResumeStorage(options) {
  if (options === void 0) {
    return void 0;
  }
  if (typeof options.storage === "object") {
    return options.storage;
  }
  const storageName = options.storage ?? "sessionStorage";
  if (typeof globalThis.window === "undefined") {
    return void 0;
  }
  try {
    return storageName === "localStorage" ? window.localStorage : window.sessionStorage;
  } catch {
    return void 0;
  }
}
function isChatResumeState(value) {
  return isRecord2(value) && value.version === 1 && typeof value.streamId === "string" && typeof value.lastEventId === "number" && Number.isFinite(value.lastEventId) && value.lastEventId >= 0 && Array.isArray(value.messages) && value.messages.every(isUIMessage);
}
function isUIMessage(value) {
  return isRecord2(value) && typeof value.id === "string" && typeof value.role === "string" && Array.isArray(value.parts);
}
function isRecord2(value) {
  return typeof value === "object" && value !== null;
}

// src/ui-messages.ts
import {
  coreMessagesToUIMessages as coreMessagesToUIMessages2
} from "@anvia/core/ui";
function createUserMessage(input) {
  if (isUIMessage2(input)) {
    return input;
  }
  const text = typeof input === "string" ? input : input.text ?? "";
  const attachments = typeof input === "string" ? [] : input.attachments ?? [];
  if (text.trim().length === 0 && attachments.length === 0) {
    return void 0;
  }
  const parts = [];
  if (text.trim().length > 0) {
    parts.push({ id: createId("part"), type: "text", text });
  }
  for (const attachment of attachments) {
    const normalized = createAttachment(attachment);
    parts.push({
      id: createId("part"),
      type: "attachment",
      attachment: normalized
    });
  }
  return {
    id: typeof input === "string" || input.id === void 0 ? createId("msg") : input.id,
    role: "user",
    parts,
    ...typeof input === "string" || input.metadata === void 0 ? {} : { metadata: input.metadata }
  };
}
function createAttachment(attachment) {
  return {
    ...attachment,
    id: attachment.id ?? createId("attachment")
  };
}
function applyUIStreamEvent(messages, event) {
  if (event.type === "message_start") {
    const index = messages.findIndex((message) => message.id === event.message.id);
    if (index === -1) {
      return [...messages, event.message];
    }
    const next = [...messages];
    next[index] = event.message;
    return next;
  }
  if (event.type === "text_delta") {
    return updateMessagePart(messages, event.messageId, {
      id: event.partId,
      type: "text",
      text: event.delta
    });
  }
  if (event.type === "reasoning_delta") {
    return updateMessagePart(messages, event.messageId, {
      id: event.partId,
      type: "reasoning",
      text: event.delta
    });
  }
  if (event.type === "tool_update") {
    return updateMessagePart(messages, event.messageId, event.part);
  }
  if (event.type === "error") {
    const messageId = createId("msg");
    return [
      ...messages,
      {
        id: messageId,
        role: "assistant",
        parts: [{ id: createId("part"), type: "error", error: event.error }]
      }
    ];
  }
  if (event.type === "message_end") {
    const { metadata } = event;
    if (metadata === void 0) {
      return messages;
    }
    return messages.map(
      (message) => message.id === event.messageId ? { ...message, metadata } : message
    );
  }
  return messages;
}
function applyAnviaStreamEvent(messages, event) {
  const uiEvent = asUIStreamEvent(event);
  if (uiEvent !== void 0) {
    return applyUIStreamEvent(messages, uiEvent);
  }
  if (!isRecord3(event) || typeof event.type !== "string") {
    return void 0;
  }
  if (event.type === "text_delta" && typeof event.delta === "string") {
    return appendAssistantDelta(messages, event.delta);
  }
  if (event.type === "reasoning_delta" && typeof event.delta === "string") {
    return appendAssistantReasoningDelta(
      messages,
      event.delta,
      typeof event.id === "string" ? event.id : void 0
    );
  }
  if (event.type === "tool_call_delta" && typeof event.id === "string") {
    const turn = eventTurn(event);
    const part = {
      id: toolPartId(event.id, turn),
      type: "tool",
      toolName: typeof event.name === "string" ? event.name : "",
      toolCallId: event.id,
      ...typeof event.callId === "string" ? { callId: event.callId } : {},
      ...turn === void 0 ? {} : { turn },
      state: "input-streaming",
      ...typeof event.argumentsDelta === "string" ? { input: event.argumentsDelta } : {}
    };
    return updateAssistantToolPart(messages, part);
  }
  if (event.type === "tool_call" && isToolCall(event.toolCall)) {
    const turn = eventTurn(event);
    const part = {
      id: toolPartId(event.toolCall.id, turn),
      type: "tool",
      toolName: event.toolCall.function.name,
      toolCallId: event.toolCall.id,
      ...typeof event.toolCall.callId === "string" ? { callId: event.toolCall.callId } : {},
      ...turn === void 0 ? {} : { turn },
      state: "input-available",
      input: event.toolCall.function.arguments
    };
    return updateAssistantToolPart(messages, part);
  }
  if (event.type === "tool_result") {
    const providerCallId = typeof event.toolCallId === "string" ? event.toolCallId : void 0;
    const internalCallId = typeof event.internalCallId === "string" ? event.internalCallId : void 0;
    const toolCallId = providerCallId ?? internalCallId;
    if (toolCallId === void 0 || typeof event.toolName !== "string") {
      return void 0;
    }
    const turn = eventTurn(event);
    const part = {
      id: toolPartId(toolCallId, turn),
      type: "tool",
      toolName: event.toolName,
      toolCallId,
      ...providerCallId === void 0 ? {} : { callId: providerCallId },
      ...turn === void 0 ? {} : { turn },
      state: "output-available",
      output: "structuredResult" in event && event.structuredResult !== void 0 ? valueToJson(event.structuredResult) : valueToJson(event.result)
    };
    return updateAssistantToolPart(
      messages,
      part,
      [providerCallId, internalCallId],
      parseJsonValue(event.args)
    );
  }
  if (event.type === "message_id" && typeof event.id === "string") {
    return setLastAssistantMetadata(messages, { providerMessageId: event.id });
  }
  if (event.type === "final") {
    const finalMessages = finalMessagesFromEvent(event);
    if (finalMessages !== void 0) {
      const next = replaceCurrentRunMessages(messages, finalMessages);
      return typeof event.runId === "string" ? setLastAssistantMetadata(next, { runId: event.runId }) : next;
    }
    if (isRecord3(event.response) && Array.isArray(event.response.choice)) {
      const finalText = textFromAssistantContent(event.response.choice);
      const next = finalText.length > 0 ? replaceAssistantText(messages, finalText) : messages;
      const providerMessageId = event.response.messageId;
      return typeof providerMessageId === "string" ? setLastAssistantMetadata(next, { providerMessageId }) : next;
    }
    if (typeof event.output === "string") {
      const next = event.output.length > 0 ? replaceAssistantText(messages, event.output) : messages;
      return typeof event.runId === "string" ? setLastAssistantMetadata(next, { runId: event.runId }) : next;
    }
  }
  if (event.type === "error" && "error" in event) {
    return appendAssistantError(messages, errorFromUnknown(event.error));
  }
  return void 0;
}
function assistantText(messages) {
  const assistant = [...messages].reverse().find((message) => message.role === "assistant");
  if (assistant === void 0) {
    return "";
  }
  return messageText(assistant);
}
function messageText(message) {
  return message.parts.flatMap((part) => part.type === "text" ? [part.text] : []).join("");
}
function assistantTextPartId(assistant) {
  const lastPart = assistant.parts.at(-1);
  if (lastPart?.type === "text") {
    return lastPart.id;
  }
  const textCount = assistant.parts.filter((part) => part.type === "text").length;
  return textCount === 0 ? `${assistant.id}_text` : `${assistant.id}_text_${textCount + 1}`;
}
function assistantReasoningPartId(assistant, reasoningId) {
  if (reasoningId !== void 0) {
    return `${assistant.id}_reasoning_${reasoningId}`;
  }
  const lastPart = assistant.parts.at(-1);
  if (lastPart?.type === "reasoning" && lastPart.reasoningId === void 0) {
    return lastPart.id;
  }
  const reasoningCount = assistant.parts.filter((part) => part.type === "reasoning").length;
  return reasoningCount === 0 ? `${assistant.id}_reasoning` : `${assistant.id}_reasoning_${reasoningCount + 1}`;
}
function appendAssistantDelta(messages, delta) {
  const current = ensureAssistantMessage(messages);
  const assistant = current[current.length - 1];
  if (assistant === void 0) {
    return current;
  }
  return updateMessagePart(current, assistant.id, {
    id: assistantTextPartId(assistant),
    type: "text",
    text: delta
  });
}
function appendAssistantReasoningDelta(messages, delta, reasoningId) {
  const current = ensureAssistantMessage(messages);
  const assistant = current[current.length - 1];
  if (assistant === void 0) {
    return current;
  }
  return updateMessagePart(current, assistant.id, {
    id: assistantReasoningPartId(assistant, reasoningId),
    type: "reasoning",
    text: delta,
    ...reasoningId === void 0 ? {} : { reasoningId }
  });
}
function replaceAssistantText(messages, text) {
  const current = ensureAssistantMessage(messages);
  const assistant = current[current.length - 1];
  if (assistant === void 0) {
    return current;
  }
  return current.map(
    (message) => message.id === assistant.id ? { ...message, parts: replaceTextPart(message, text) } : message
  );
}
function finalMessagesFromEvent(event) {
  if (!Array.isArray(event.messages) || event.messages.length === 0) {
    return void 0;
  }
  try {
    return coreMessagesToUIMessages2(event.messages);
  } catch {
    return void 0;
  }
}
function replaceCurrentRunMessages(messages, finalMessages) {
  if (messages.length === 0 || finalMessages.length === 0) {
    return finalMessages;
  }
  const lastUserIndex = findLastIndex(messages, (message) => message.role === "user");
  if (lastUserIndex === -1) {
    return finalMessages;
  }
  const finalFirst = finalMessages[0];
  if (finalFirst?.role !== "user") {
    return [...messages.slice(0, lastUserIndex + 1), ...finalMessages];
  }
  const currentPrompt = messages[lastUserIndex];
  if (currentPrompt === void 0 || messageText(currentPrompt) !== messageText(finalFirst)) {
    return finalMessages;
  }
  return [
    ...messages.slice(0, lastUserIndex),
    { ...finalFirst, id: currentPrompt.id },
    ...finalMessages.slice(1)
  ];
}
function updateMessagePart(messages, messageId, part) {
  const existingMessage = messages.find((message) => message.id === messageId);
  const current = existingMessage === void 0 ? [...messages, { id: messageId, role: "assistant", parts: [] }] : messages;
  return current.map((message) => {
    if (message.id !== messageId) {
      return message;
    }
    const partIndex = message.parts.findIndex((item) => item.id === part.id);
    if (partIndex === -1) {
      return { ...message, parts: [...message.parts, part] };
    }
    const parts = [...message.parts];
    const currentPart = parts[partIndex];
    if (currentPart?.type === "text" && part.type === "text" || currentPart?.type === "reasoning" && part.type === "reasoning") {
      parts[partIndex] = { ...currentPart, text: `${currentPart.text}${part.text}` };
    } else if (currentPart?.type === "tool" && part.type === "tool") {
      parts[partIndex] = mergeToolPart(currentPart, part);
    } else {
      parts[partIndex] = part;
    }
    return { ...message, parts };
  });
}
function updateAssistantToolPart(messages, part, aliases = [], inputMatch) {
  const current = ensureAssistantMessage(messages);
  const assistant = current[current.length - 1];
  if (assistant === void 0) {
    return current;
  }
  const existingPart = findMatchingToolPart(assistant, part, aliases, inputMatch);
  const nextPart = existingPart === void 0 ? part : {
    ...part,
    id: existingPart.id,
    toolCallId: existingPart.toolCallId,
    ...existingPart.callId !== void 0 ? { callId: existingPart.callId } : part.callId !== void 0 ? { callId: part.callId } : {}
  };
  return updateMessagePart(current, assistant.id, nextPart);
}
function findMatchingToolPart(message, part, aliases, inputMatch) {
  const candidates = new Set(
    [part.toolCallId, part.callId, ...aliases].filter(
      (candidate) => candidate !== void 0 && candidate.length > 0
    )
  );
  const exactMatch = message.parts.find(
    (item) => item.type === "tool" && hasCompatibleTurn(item, part) && (item.id === part.id || candidates.has(item.toolCallId) || item.callId !== void 0 && candidates.has(item.callId))
  );
  if (exactMatch !== void 0 || inputMatch === void 0) {
    return exactMatch;
  }
  const inputMatches = message.parts.filter(
    (item) => item.type === "tool" && hasCompatibleTurn(item, part) && item.toolName === part.toolName && item.state !== "output-available" && item.state !== "error" && jsonEquals(item.input ?? {}, inputMatch)
  );
  return inputMatches.length === 1 ? inputMatches[0] : void 0;
}
function mergeToolPart(currentPart, nextPart) {
  const merged = { ...currentPart, ...nextPart };
  if (nextPart.toolName.length === 0) {
    merged.toolName = currentPart.toolName;
  }
  if (nextPart.callId === void 0 && currentPart.callId !== void 0) {
    merged.callId = currentPart.callId;
  }
  if (nextPart.input === void 0 && currentPart.input !== void 0) {
    merged.input = currentPart.input;
  } else if (typeof currentPart.input === "string" && typeof nextPart.input === "string") {
    merged.input = `${currentPart.input}${nextPart.input}`;
  }
  if (nextPart.output === void 0 && currentPart.output !== void 0) {
    merged.output = currentPart.output;
  }
  if (nextPart.error === void 0 && currentPart.error !== void 0) {
    merged.error = currentPart.error;
  }
  return merged;
}
function hasCompatibleTurn(currentPart, nextPart) {
  return nextPart.turn === void 0 ? currentPart.turn === void 0 : currentPart.turn === nextPart.turn;
}
function appendAssistantError(messages, error) {
  const messageId = createId("msg");
  return [
    ...messages,
    {
      id: messageId,
      role: "assistant",
      parts: [{ id: createId("part"), type: "error", error }]
    }
  ];
}
function setLastAssistantMetadata(messages, metadata) {
  const current = ensureAssistantMessage(messages);
  const assistant = current[current.length - 1];
  if (assistant === void 0) {
    return current;
  }
  return current.map(
    (message) => message.id === assistant.id ? { ...message, metadata: mergeMetadata(message.metadata, metadata) } : message
  );
}
function replaceTextPart(message, text) {
  const index = message.parts.findIndex((part) => part.type === "text");
  if (index === -1) {
    return [...message.parts, { id: `${message.id}_text`, type: "text", text }];
  }
  const parts = [...message.parts];
  const current = parts[index];
  if (current?.type === "text") {
    parts[index] = { ...current, text };
  }
  return parts;
}
function ensureAssistantMessage(messages) {
  const last = messages.at(-1);
  if (last?.role === "assistant") {
    return messages;
  }
  return [
    ...messages,
    {
      id: createId("msg"),
      role: "assistant",
      parts: []
    }
  ];
}
var nextId = 0;
function createId(prefix) {
  const random = globalThis.crypto?.randomUUID?.();
  if (random !== void 0) {
    return `${prefix}_${random}`;
  }
  nextId += 1;
  return `${prefix}_${nextId.toString(36)}`;
}
function isUIMessage2(input) {
  return typeof input !== "string" && "role" in input && Array.isArray(input.parts);
}
function asUIStreamEvent(event) {
  if (!isRecord3(event) || typeof event.type !== "string") {
    return void 0;
  }
  if (event.type === "message_start" && isStreamMessage(event.message)) {
    return event;
  }
  if ((event.type === "text_delta" || event.type === "reasoning_delta") && typeof event.messageId === "string" && typeof event.partId === "string" && typeof event.delta === "string") {
    return event;
  }
  if (event.type === "tool_update" && typeof event.messageId === "string" && typeof event.partId === "string" && isUIMessagePart(event.part)) {
    return event;
  }
  if (event.type === "message_end" && typeof event.messageId === "string") {
    return event;
  }
  if (event.type === "error" && isUIError(event.error)) {
    return event;
  }
  return void 0;
}
function isStreamMessage(value) {
  return isRecord3(value) && typeof value.id === "string" && typeof value.role === "string" && Array.isArray(value.parts);
}
function isUIMessagePart(value) {
  return isRecord3(value) && typeof value.id === "string" && typeof value.type === "string";
}
function isUIError(value) {
  return isRecord3(value) && typeof value.message === "string";
}
function isToolCall(value) {
  return isRecord3(value) && typeof value.id === "string" && isRecord3(value.function) && typeof value.function.name === "string" && "arguments" in value.function;
}
function eventTurn(event) {
  return typeof event.turn === "number" && Number.isFinite(event.turn) ? event.turn : void 0;
}
function textFromAssistantContent(content) {
  return content.flatMap(
    (item) => isRecord3(item) && item.type === "text" && typeof item.text === "string" ? [item.text] : []
  ).join("\n");
}
function errorFromUnknown(error) {
  if (error instanceof Error) {
    return { name: error.name, message: error.message };
  }
  return { message: typeof error === "string" ? error : stringifyUnknown(error) };
}
function stringifyUnknown(value) {
  try {
    const serialized = JSON.stringify(value);
    return serialized === void 0 ? String(value) : serialized;
  } catch {
    return String(value);
  }
}
function parseJsonValue(value) {
  if (typeof value === "string") {
    try {
      return valueToJson(JSON.parse(value));
    } catch {
      return void 0;
    }
  }
  if (value === void 0) {
    return void 0;
  }
  return valueToJson(value);
}
function valueToJson(value) {
  if (value === null || typeof value === "string" || typeof value === "number" || typeof value === "boolean" || Array.isArray(value) || isRecord3(value)) {
    return value;
  }
  return stringifyUnknown(value);
}
function jsonEquals(left, right) {
  try {
    return JSON.stringify(normalizeJson(left)) === JSON.stringify(normalizeJson(right));
  } catch {
    return false;
  }
}
function normalizeJson(value) {
  if (Array.isArray(value)) {
    return value.map((item) => normalizeJson(item));
  }
  if (isJsonObject(value)) {
    return Object.fromEntries(
      Object.entries(value).filter((entry) => entry[1] !== void 0).sort(([left], [right]) => left.localeCompare(right)).map(([key, item]) => [key, normalizeJson(item)])
    );
  }
  return value;
}
function mergeMetadata(current, next) {
  if (isJsonObject(current) && isJsonObject(next)) {
    return { ...current, ...next };
  }
  return next;
}
function isJsonObject(value) {
  return isRecord3(value) && !Array.isArray(value);
}
function toolPartId(toolCallId, turn) {
  return turn === void 0 ? `tool_${toolCallId}` : `tool_${turn}_${toolCallId}`;
}
function findLastIndex(items, predicate) {
  for (let index = items.length - 1; index >= 0; index -= 1) {
    if (predicate(items[index])) {
      return index;
    }
  }
  return -1;
}
function isRecord3(value) {
  return typeof value === "object" && value !== null;
}

// src/use-chat.ts
function useChat(options = {}) {
  const [messages, setMessagesState] = useState(() => [...options.initialMessages ?? []]);
  const [events, setEvents] = useState([]);
  const [status, setStatus] = useState("idle");
  const [error, setError] = useState();
  const [approvals, setApprovals] = useState([]);
  const [questions, setQuestions] = useState([]);
  const [decidingApprovals, setDecidingApprovals] = useState(() => /* @__PURE__ */ new Set());
  const [answeringQuestions, setAnsweringQuestions] = useState(() => /* @__PURE__ */ new Set());
  const [streamId, setStreamIdState] = useState();
  const [isResuming, setIsResuming] = useState(false);
  const abortRef = useRef(void 0);
  const messagesRef = useRef(messages);
  const approvalsRef = useRef(approvals);
  const questionsRef = useRef(questions);
  const decidingApprovalsRef = useRef(decidingApprovals);
  const answeringQuestionsRef = useRef(answeringQuestions);
  const streamIdRef = useRef(void 0);
  const lastEventIdRef = useRef(0);
  const autoResumeStartedRef = useRef(false);
  const humanInputVersionRef = useRef(0);
  useEffect(() => {
    messagesRef.current = messages;
  }, [messages]);
  useEffect(() => {
    approvalsRef.current = approvals;
  }, [approvals]);
  useEffect(() => {
    questionsRef.current = questions;
  }, [questions]);
  useEffect(() => {
    decidingApprovalsRef.current = decidingApprovals;
  }, [decidingApprovals]);
  useEffect(() => {
    answeringQuestionsRef.current = answeringQuestions;
  }, [answeringQuestions]);
  useEffect(() => {
    return () => {
      abortRef.current?.abort();
    };
  }, []);
  const transport = useMemo(() => {
    if (options.transport !== void 0) {
      return options.transport;
    }
    if (options.endpoint === void 0) {
      return void 0;
    }
    return createChatTransport({
      endpoint: options.endpoint,
      format: options.format ?? "jsonl"
    });
  }, [options.transport, options.endpoint, options.format]);
  const setMessages = useCallback((nextMessages) => {
    const next = typeof nextMessages === "function" ? nextMessages(messagesRef.current) : nextMessages;
    messagesRef.current = next;
    setMessagesState(next);
  }, []);
  const setStreamId = useCallback((nextStreamId) => {
    streamIdRef.current = nextStreamId;
    setStreamIdState(nextStreamId);
  }, []);
  const persistResumeState = useCallback(() => {
    const currentStreamId = streamIdRef.current;
    if (currentStreamId === void 0) {
      return;
    }
    saveChatResumeState(options.resume, {
      version: 1,
      streamId: currentStreamId,
      lastEventId: lastEventIdRef.current,
      messages: messagesRef.current
    });
  }, [options.resume]);
  const clearResumeState = useCallback(() => {
    lastEventIdRef.current = 0;
    setStreamId(void 0);
    clearChatResumeState(options.resume);
  }, [options.resume, setStreamId]);
  const updateApproval = useCallback((approval) => {
    setApprovals((current) => {
      const next = upsertById(current, approval);
      approvalsRef.current = next;
      return next;
    });
  }, []);
  const updateQuestion = useCallback((question) => {
    setQuestions((current) => {
      const next = upsertById(current, question);
      questionsRef.current = next;
      return next;
    });
  }, []);
  const clearHumanInput = useCallback(() => {
    humanInputVersionRef.current += 1;
    approvalsRef.current = [];
    questionsRef.current = [];
    decidingApprovalsRef.current = /* @__PURE__ */ new Set();
    answeringQuestionsRef.current = /* @__PURE__ */ new Set();
    setApprovals([]);
    setQuestions([]);
    setDecidingApprovals(/* @__PURE__ */ new Set());
    setAnsweringQuestions(/* @__PURE__ */ new Set());
  }, []);
  const applyEvent = useCallback(
    (event) => {
      const mappedUIEvent = options.eventToUIEvent?.(event);
      if (mappedUIEvent !== void 0) {
        setMessages((current) => applyUIStreamEvent(current, mappedUIEvent));
        return;
      }
      const hasCustomEventMapper = options.eventToUIEvent !== void 0 || options.eventToDelta !== void 0 || options.eventToFinal !== void 0;
      if (!hasCustomEventMapper) {
        let handled = false;
        setMessages((current) => {
          const next = applyAnviaStreamEvent(current, event);
          if (next === void 0) {
            return current;
          }
          handled = true;
          return next;
        });
        if (handled) {
          return;
        }
      }
      const delta = options.eventToDelta?.(event);
      if (delta !== void 0 && delta.length > 0) {
        setMessages((current) => appendAssistantDelta(current, delta));
      }
      const final = options.eventToFinal?.(event);
      if (final !== void 0) {
        setMessages((current) => replaceAssistantText(current, final));
      }
    },
    [options, setMessages]
  );
  const applyHumanInputEvent = useCallback(
    (event) => {
      const humanInputOptions = options.humanInput;
      if (humanInputOptions === void 0) {
        return;
      }
      const eventToApproval = humanInputOptions.eventToApproval ?? defaultEventToApproval;
      const approval = eventToApproval(event);
      if (approval !== void 0) {
        updateApproval(approval);
      }
      const eventToQuestion = humanInputOptions.eventToQuestion ?? defaultEventToQuestion;
      const question = eventToQuestion(event);
      if (question !== void 0) {
        updateQuestion(question);
      }
    },
    [options.humanInput, updateApproval, updateQuestion]
  );
  const sendMessages = useCallback(
    async (nextMessages, runOptions = {}) => {
      if (transport === void 0) {
        throw new Error("useChat requires either transport or endpoint");
      }
      abortRef.current?.abort();
      const abortController = new AbortController();
      abortRef.current = abortController;
      const createRequest = options.createRequest ?? ((args) => ({
        messages: args.coreMessages,
        stream: true,
        ...args.resume === void 0 ? {} : { resume: args.resume }
      }));
      if (runOptions.resume === void 0) {
        clearResumeState();
      } else {
        lastEventIdRef.current = runOptions.resume.after;
        setStreamId(runOptions.resume.streamId);
      }
      setMessages(nextMessages);
      setEvents([]);
      setError(void 0);
      clearHumanInput();
      setStatus("streaming");
      setIsResuming(runOptions.isResuming === true);
      try {
        const coreMessages = uiMessagesToCoreMessages(nextMessages);
        const request = createRequest({
          messages: nextMessages,
          uiMessages: nextMessages,
          coreMessages,
          resume: runOptions.resume
        });
        for await (const rawEvent of transport.send(request, { signal: abortController.signal })) {
          if (abortRef.current !== abortController || abortController.signal.aborted) {
            return;
          }
          if (isResumableStreamEnvelope(rawEvent)) {
            if (rawEvent.type === "stream_start") {
              setStreamId(rawEvent.streamId);
              persistResumeState();
              continue;
            }
            if (rawEvent.type === "stream_end") {
              lastEventIdRef.current = rawEvent.eventId;
              clearResumeState();
              continue;
            }
            lastEventIdRef.current = rawEvent.eventId;
            const event2 = rawEvent.event;
            setEvents((current) => [...current, event2]);
            applyHumanInputEvent(event2);
            options.onEvent?.(event2);
            if (abortRef.current !== abortController || abortController.signal.aborted) {
              return;
            }
            applyEvent(event2);
            persistResumeState();
            continue;
          }
          const event = rawEvent;
          setEvents((current) => [...current, event]);
          applyHumanInputEvent(event);
          options.onEvent?.(event);
          if (abortRef.current !== abortController || abortController.signal.aborted) {
            return;
          }
          applyEvent(event);
        }
        if (abortRef.current === abortController && !abortController.signal.aborted) {
          setStatus("idle");
        }
      } catch (caught) {
        if (isAbortError(caught)) {
          if (abortRef.current === abortController) {
            setStatus("idle");
          }
          return;
        }
        if (abortRef.current !== abortController) {
          return;
        }
        setError(caught);
        setStatus("error");
        options.onError?.(caught);
      } finally {
        if (abortRef.current === abortController) {
          abortRef.current = void 0;
          setIsResuming(false);
        }
      }
    },
    [
      applyEvent,
      applyHumanInputEvent,
      clearHumanInput,
      clearResumeState,
      options,
      persistResumeState,
      setMessages,
      setStreamId,
      transport
    ]
  );
  const resume = useCallback(async () => {
    const resumeState = loadChatResumeState(options.resume);
    if (resumeState === void 0) {
      return;
    }
    await sendMessages(resumeState.messages, {
      resume: {
        streamId: resumeState.streamId,
        after: resumeState.lastEventId
      },
      isResuming: true
    });
  }, [options.resume, sendMessages]);
  useEffect(() => {
    if (options.resume?.auto === false || autoResumeStartedRef.current) {
      return;
    }
    autoResumeStartedRef.current = true;
    void resume();
  }, [options.resume?.auto, resume]);
  const sendMessage = useCallback(
    async (input) => {
      const message = createUserMessage(input);
      if (message === void 0) {
        return;
      }
      const currentMessages = messagesRef.current;
      const baseMessages = abortRef.current !== void 0 && currentMessages.at(-1)?.role === "assistant" ? currentMessages.slice(0, -1) : currentMessages;
      await sendMessages([...baseMessages, message]);
    },
    [sendMessages]
  );
  const send = useCallback(
    async (input = "") => {
      await sendMessage(input);
    },
    [sendMessage]
  );
  const regenerate = useCallback(async () => {
    const lastUserIndex = findLastUserIndex(messagesRef.current);
    if (lastUserIndex === -1) {
      return;
    }
    await sendMessages(messagesRef.current.slice(0, lastUserIndex + 1));
  }, [sendMessages]);
  const stop = useCallback(() => {
    abortRef.current?.abort();
    abortRef.current = void 0;
    clearResumeState();
    setStatus("idle");
    setIsResuming(false);
  }, [clearResumeState]);
  const decideToolApproval = useCallback(
    async (approvalId, approved, reason) => {
      const humanInputOptions = options.humanInput;
      if (humanInputOptions === void 0) {
        throw new Error("useChat humanInput is not configured");
      }
      if (decidingApprovalsRef.current.has(approvalId)) {
        return;
      }
      const humanInputVersion = humanInputVersionRef.current;
      const nextDeciding = new Set(decidingApprovalsRef.current).add(approvalId);
      decidingApprovalsRef.current = nextDeciding;
      setDecidingApprovals(nextDeciding);
      try {
        const approval = approvalsRef.current.find((item) => item.id === approvalId);
        const input = {
          approvalId,
          approved,
          ...reason === void 0 ? {} : { reason },
          ...approval === void 0 ? {} : { approval }
        };
        const result = humanInputOptions.decideApproval === void 0 ? await defaultDecideApproval(input, humanInputOptions) : await humanInputOptions.decideApproval(input);
        if (result !== void 0 && humanInputVersionRef.current === humanInputVersion) {
          updateApproval(result);
        }
      } finally {
        if (humanInputVersionRef.current === humanInputVersion) {
          const next = new Set(decidingApprovalsRef.current);
          next.delete(approvalId);
          decidingApprovalsRef.current = next;
          setDecidingApprovals(next);
        }
      }
    },
    [options.humanInput, updateApproval]
  );
  const approveTool = useCallback(
    async (approvalId, reason) => {
      await decideToolApproval(approvalId, true, reason);
    },
    [decideToolApproval]
  );
  const rejectTool = useCallback(
    async (approvalId, reason) => {
      await decideToolApproval(approvalId, false, reason);
    },
    [decideToolApproval]
  );
  const answerToolQuestion = useCallback(
    async (questionId, answers) => {
      const humanInputOptions = options.humanInput;
      if (humanInputOptions === void 0) {
        throw new Error("useChat humanInput is not configured");
      }
      if (answeringQuestionsRef.current.has(questionId)) {
        return;
      }
      const humanInputVersion = humanInputVersionRef.current;
      const nextAnswering = new Set(answeringQuestionsRef.current).add(questionId);
      answeringQuestionsRef.current = nextAnswering;
      setAnsweringQuestions(nextAnswering);
      try {
        const question = questionsRef.current.find((item) => item.id === questionId);
        const input = {
          questionId,
          answers,
          ...question === void 0 ? {} : { question }
        };
        const result = humanInputOptions.answerQuestion === void 0 ? await defaultAnswerQuestion(input, humanInputOptions) : await humanInputOptions.answerQuestion(input);
        if (result !== void 0 && humanInputVersionRef.current === humanInputVersion) {
          updateQuestion(result);
        }
      } finally {
        if (humanInputVersionRef.current === humanInputVersion) {
          const next = new Set(answeringQuestionsRef.current);
          next.delete(questionId);
          answeringQuestionsRef.current = next;
          setAnsweringQuestions(next);
        }
      }
    },
    [options.humanInput, updateQuestion]
  );
  const reset = useCallback(
    (nextMessages = []) => {
      abortRef.current?.abort();
      abortRef.current = void 0;
      clearResumeState();
      setMessages(nextMessages);
      setEvents([]);
      clearHumanInput();
      setError(void 0);
      setStatus("idle");
      setIsResuming(false);
    },
    [clearHumanInput, clearResumeState, setMessages]
  );
  return {
    messages,
    events,
    suggestions: options.suggestions ?? [],
    setMessages,
    sendMessage,
    send,
    regenerate,
    stop,
    reset,
    status,
    error,
    text: assistantText(messages),
    streamId,
    isResuming,
    resume,
    humanInput: {
      approvals: {
        all: approvals,
        pending: approvals.filter((approval) => approval.status === "pending")
      },
      questions: {
        all: questions,
        pending: questions.filter((question) => question.status === "pending")
      }
    },
    decidingApprovals: new Set(decidingApprovals),
    answeringQuestions: new Set(answeringQuestions),
    approveTool,
    rejectTool,
    answerToolQuestion
  };
}
function findLastUserIndex(messages) {
  for (let index = messages.length - 1; index >= 0; index -= 1) {
    if (messages[index]?.role === "user") {
      return index;
    }
  }
  return -1;
}
function isAbortError(error) {
  return error instanceof DOMException && error.name === "AbortError";
}

// src/use-completion.ts
import {
  uiMessagesToCoreMessages as uiMessagesToCoreMessages2
} from "@anvia/core/ui";
import { useCallback as useCallback2, useEffect as useEffect2, useMemo as useMemo2, useRef as useRef2, useState as useState2 } from "react";
function useCompletion(options = {}) {
  const initialMessages = useMemo2(
    () => options.initialMessages ?? (options.initialCompletion === void 0 ? [] : [
      {
        id: "initial_assistant",
        role: "assistant",
        parts: [
          {
            id: "initial_assistant_text",
            type: "text",
            text: options.initialCompletion
          }
        ]
      }
    ]),
    [options.initialCompletion, options.initialMessages]
  );
  const [messages, setMessagesState] = useState2(() => [...initialMessages]);
  const [input, setInput] = useState2("");
  const [events, setEvents] = useState2([]);
  const [status, setStatus] = useState2("idle");
  const [error, setError] = useState2();
  const abortRef = useRef2(void 0);
  const messagesRef = useRef2(messages);
  useEffect2(() => {
    messagesRef.current = messages;
  }, [messages]);
  useEffect2(() => {
    return () => {
      abortRef.current?.abort();
    };
  }, []);
  const transport = useMemo2(() => {
    if (options.transport !== void 0) {
      return options.transport;
    }
    if (options.endpoint === void 0) {
      return void 0;
    }
    return createFetchTransport({
      endpoint: options.endpoint,
      format: options.format ?? "jsonl"
    });
  }, [options.transport, options.endpoint, options.format]);
  const setMessages = useCallback2((nextMessages) => {
    messagesRef.current = nextMessages;
    setMessagesState(nextMessages);
  }, []);
  const applyEvent = useCallback2(
    (event) => {
      const mappedUIEvent = options.eventToUIEvent?.(event);
      if (mappedUIEvent !== void 0) {
        setMessagesState((current) => {
          const next = applyUIStreamEvent(current, mappedUIEvent);
          messagesRef.current = next;
          return next;
        });
        return;
      }
      const hasCustomEventMapper = options.eventToUIEvent !== void 0 || options.eventToDelta !== void 0 || options.eventToFinal !== void 0;
      if (!hasCustomEventMapper) {
        let handled = false;
        setMessagesState((current) => {
          const next = applyAnviaStreamEvent(current, event);
          if (next === void 0) {
            return current;
          }
          handled = true;
          messagesRef.current = next;
          return next;
        });
        if (handled) {
          return;
        }
      }
      const delta = options.eventToDelta?.(event);
      if (delta !== void 0 && delta.length > 0) {
        setMessagesState((current) => {
          const next = appendAssistantDelta(current, delta);
          messagesRef.current = next;
          return next;
        });
      }
      const final = options.eventToFinal?.(event);
      if (final !== void 0) {
        setMessagesState((current) => {
          const next = replaceAssistantText(current, final);
          messagesRef.current = next;
          return next;
        });
      }
    },
    [options]
  );
  const complete = useCallback2(
    async (nextInput) => {
      if (transport === void 0) {
        throw new Error("useCompletion requires either transport or endpoint");
      }
      const prompt = nextInput ?? input;
      const userMessage = createUserMessage(prompt);
      if (userMessage === void 0) {
        return;
      }
      const previousAbortController = abortRef.current;
      previousAbortController?.abort();
      const currentMessages = messagesRef.current;
      const baseMessages = previousAbortController !== void 0 && currentMessages.at(-1)?.role === "assistant" ? currentMessages.slice(0, -1) : currentMessages;
      const nextMessages = [...baseMessages, userMessage];
      const abortController = new AbortController();
      abortRef.current = abortController;
      const createRequest = options.createRequest ?? ((args) => ({ messages: args.coreMessages, stream: true }));
      setMessages(nextMessages);
      setInput("");
      setEvents([]);
      setError(void 0);
      setStatus("streaming");
      try {
        const coreMessages = uiMessagesToCoreMessages2(nextMessages);
        const request = createRequest({
          messages: nextMessages,
          uiMessages: nextMessages,
          coreMessages
        });
        for await (const event of transport.send(request, { signal: abortController.signal })) {
          setEvents((current) => [...current, event]);
          options.onEvent?.(event);
          applyEvent(event);
        }
        if (!abortController.signal.aborted) {
          setStatus("idle");
        }
      } catch (caught) {
        if (isAbortError2(caught)) {
          if (abortRef.current === abortController) {
            setStatus("idle");
          }
          return;
        }
        setError(caught);
        setStatus("error");
        options.onError?.(caught);
      } finally {
        if (abortRef.current === abortController) {
          abortRef.current = void 0;
        }
      }
    },
    [applyEvent, input, options, setMessages, transport]
  );
  const stop = useCallback2(() => {
    abortRef.current?.abort();
    abortRef.current = void 0;
    setStatus("idle");
  }, []);
  const reset = useCallback2(
    (messagesOrCompletion) => {
      abortRef.current?.abort();
      abortRef.current = void 0;
      if (Array.isArray(messagesOrCompletion)) {
        setMessages(messagesOrCompletion);
      } else if (typeof messagesOrCompletion === "string") {
        setMessages([
          {
            id: "reset_assistant",
            role: "assistant",
            parts: [{ id: "reset_assistant_text", type: "text", text: messagesOrCompletion }]
          }
        ]);
      } else {
        setMessages([]);
      }
      setEvents([]);
      setError(void 0);
      setInput("");
      setStatus("idle");
    },
    [setMessages]
  );
  return {
    messages,
    completion: assistantText(messages),
    input,
    setInput,
    complete,
    stop,
    reset,
    status,
    error,
    events
  };
}
function isAbortError2(error) {
  return error instanceof DOMException && error.name === "AbortError";
}
export {
  EventStreamHttpError,
  createChatTransport,
  createDirectTransport,
  createFetchTransport,
  defaultAnswerQuestion,
  defaultDecideApproval,
  defaultEventToApproval,
  defaultEventToQuestion,
  fetchEventStream,
  initialMessagesFromMemory,
  readJsonlStream,
  readSseStream,
  useChat,
  useCompletion
};
