// @anvia/core@0.8.0/dist/chunk-USWULPZJ.js
import {
  InMemoryVectorStore,
  createTool
} from "./chunk-VQE5PAAT.js";
import {
  compact
} from "./chunk-MMHG7WAM.js";
import {
  embedDocuments
} from "./chunk-FTO3W4UP.js";
import {
  isToolResultContentArray,
  serializeToolResultOutput
} from "./chunk-WJO6NVE2.js";

// src/tool/errors.ts
var ToolCallError = class extends Error {
  constructor(message, cause) {
    super(message);
    this.cause = cause;
    this.name = "ToolCallError";
  }
  cause;
};
var ToolNotFoundError = class extends Error {
  constructor(toolName) {
    super(`Tool not found: ${toolName}`);
    this.toolName = toolName;
    this.name = "ToolNotFoundError";
  }
  toolName;
};
var ToolJsonError = class extends Error {
  constructor(message, cause) {
    super(message);
    this.cause = cause;
    this.name = "ToolJsonError";
  }
  cause;
};

// src/tool/tool.ts
var ToolOutput = {
  content(content) {
    return content;
  }
};
function normalizeToolResultOutput(output) {
  return isToolResultContentArray(output) ? output : serializeToolResultOutput(output);
}
function toolResultContentToText(content) {
  return content.map((item) => item.type === "text" ? item.text : `[image:${item.mediaType ?? "image/png"}]`).join("\n");
}
function parseToolArgs(args) {
  if (args.trim() === "") {
    return {};
  }
  return JSON.parse(args);
}

// src/tool/tool-set.ts
var ToolSet = class _ToolSet {
  tools = /* @__PURE__ */ new Map();
  static fromTools(tools) {
    const toolSet = new _ToolSet();
    for (const tool of tools) {
      toolSet.addTool(tool);
    }
    return toolSet;
  }
  addTool(tool) {
    this.tools.set(tool.name, tool);
    return this;
  }
  addTools(tools) {
    const values = Array.isArray(tools) ? tools : tools.values();
    for (const tool of values) {
      this.addTool(tool);
    }
    return this;
  }
  deleteTool(toolName) {
    return this.tools.delete(toolName);
  }
  contains(toolName) {
    return this.tools.has(toolName);
  }
  get(toolName) {
    return this.tools.get(toolName);
  }
  values() {
    return [...this.tools.values()];
  }
  async getToolDefinitions(prompt = "") {
    const defs = [];
    for (const tool of this.tools.values()) {
      defs.push(await tool.definition(prompt));
    }
    return defs;
  }
  async call(toolName, args, context) {
    const tool = this.tools.get(toolName);
    if (tool === void 0) {
      throw new ToolNotFoundError(toolName);
    }
    let parsedArgs;
    try {
      parsedArgs = parseToolArgs(args);
    } catch (error) {
      throw new ToolJsonError(`Invalid JSON arguments for tool ${toolName}`, error);
    }
    try {
      const output = await tool.call(parsedArgs, context);
      return normalizeToolResultOutput(output);
    } catch (error) {
      if (error instanceof Error) {
        throw new ToolCallError(error.message, error);
      }
      throw new ToolCallError(`Tool ${toolName} failed`, error);
    }
  }
};

// src/tool/dynamic-tools.ts
async function embedTools(model, tools, options = {}) {
  const toolList = Array.isArray(tools) ? tools : tools.values();
  const definitions = await Promise.all(
    toolList.map(async (tool) => ({ tool, definition: await tool.definition("") }))
  );
  const documents = definitions.map(({ tool, definition }) => {
    const content = options.content?.(tool, definition) ?? defaultToolEmbeddingText(definition);
    const texts = Array.isArray(content) ? content : [content];
    const metadata = options.metadata?.(tool, definition);
    const document = compact({
      toolName: tool.name,
      definition,
      text: texts.join("\n"),
      metadata
    });
    return { tool, document, texts, metadata };
  });
  return embedDocuments(model, documents, {
    id: (item) => item.tool.name,
    content: (item) => item.texts,
    metadata: (item) => item.metadata,
    concurrency: options.concurrency
  }).then(
    (embedded) => embedded.map((item) => ({
      ...item,
      document: item.document.document
    }))
  );
}
async function createToolIndex(model, tools, options = {}) {
  const toolSet = Array.isArray(tools) ? ToolSet.fromTools(tools) : tools;
  const embedded = await embedTools(model, toolSet, options);
  const index = InMemoryVectorStore.fromDocuments(embedded).index(model);
  return new DynamicToolSearchIndex(index, toolSet);
}
function isDynamicToolIndex(value) {
  return typeof value === "object" && value !== null && "toolSet" in value && value.toolSet instanceof ToolSet;
}
var DynamicToolSearchIndex = class {
  constructor(index, toolSet) {
    this.index = index;
    this.toolSet = toolSet;
    if (index.inspect !== void 0) {
      this.inspect = (request) => index.inspect?.(request);
    }
  }
  index;
  toolSet;
  inspect;
  search(request) {
    return this.index.search(request);
  }
  searchIds(request) {
    return this.index.searchIds(request);
  }
  asTool(options) {
    return this.index.asTool(options);
  }
};
function defaultToolEmbeddingText(definition) {
  return [definition.name, definition.description, JSON.stringify(definition.parameters)];
}

// src/tool/middleware.ts
function createMiddleware(middleware) {
  return middleware;
}
function createToolMiddleware(middleware) {
  return middleware;
}

// src/tool/think-tool.ts
import { z } from "zod";
var defaultThinkToolDescription = "Use this tool to record a thought while reasoning through a complex task. It does not retrieve information, store memory, or change external state.";
var thinkToolInput = z.object({
  thought: z.string().describe("A thought to record while reasoning through a task.")
});
function createThinkTool(options = {}) {
  return createTool({
    name: options.name ?? "think",
    description: options.description ?? defaultThinkToolDescription,
    input: thinkToolInput,
    output: z.string(),
    execute: (args) => args.thought
  });
}

export {
  ToolCallError,
  ToolNotFoundError,
  ToolJsonError,
  ToolOutput,
  normalizeToolResultOutput,
  toolResultContentToText,
  parseToolArgs,
  ToolSet,
  embedTools,
  createToolIndex,
  isDynamicToolIndex,
  createMiddleware,
  createToolMiddleware,
  createThinkTool
};
