// @anvia/server@0.5.11/dist/index.js
// src/errors.ts
function errorEvent(error) {
  return {
    type: "error",
    error: serializeError(error)
  };
}
function serializeError(error) {
  if (error instanceof Error) {
    return {
      name: error.name,
      message: error.message
    };
  }
  return error;
}

// src/jsonl.ts
function createJsonlStream(events, options = {}) {
  const encoder = new TextEncoder();
  const iterator = events[Symbol.asyncIterator]();
  const serialize = options.serialize ?? serializeJson;
  return new ReadableStream({
    async pull(controller) {
      try {
        const next = await iterator.next();
        if (next.done === true) {
          controller.close();
          return;
        }
        controller.enqueue(encoder.encode(`${serialize(next.value)}
`));
      } catch (error) {
        controller.enqueue(encoder.encode(`${serialize(errorEvent(error))}
`));
        controller.close();
      }
    },
    async cancel() {
      await iterator.return?.();
    }
  });
}
function serializeJson(value) {
  return JSON.stringify(value);
}

// src/resumable.ts
function createResumableStream(events, options) {
  const streamId = options.id;
  let openPromise;
  function start() {
    if (openPromise !== void 0) {
      return openPromise;
    }
    openPromise = options.store.open({ streamId }).then(() => void 0);
    void drainResumableStream(events, options, openPromise);
    return openPromise;
  }
  return {
    async *[Symbol.asyncIterator]() {
      await start();
      yield* resumeStreamEvents({ id: streamId, store: options.store });
    }
  };
}
async function* resumeStreamEvents(options) {
  const streamId = options.id;
  yield { type: "stream_start", streamId, eventId: 0 };
  const subscribeInput = { streamId };
  if (options.after !== void 0) {
    subscribeInput.after = options.after;
  }
  for await (const record of options.store.subscribe(subscribeInput)) {
    yield recordToEnvelope(record);
  }
  const state = await options.store.status({ streamId });
  yield stateToEndEnvelope(streamId, state);
}
function createMemoryResumableStreamStore() {
  const streams = /* @__PURE__ */ new Map();
  function state(stream) {
    if (stream === void 0) {
      return { status: "missing", lastEventId: 0 };
    }
    return {
      status: stream.status,
      lastEventId: stream.records.at(-1)?.eventId ?? 0
    };
  }
  return {
    async open(input) {
      const existing = streams.get(input.streamId);
      if (existing !== void 0) {
        closeSubscribers(existing);
      }
      const stream = {
        records: [],
        status: "running",
        subscribers: /* @__PURE__ */ new Set()
      };
      streams.set(input.streamId, stream);
      return state(stream);
    },
    async append(input) {
      const stream = streams.get(input.streamId);
      if (stream === void 0) {
        throw new Error(`Resumable stream "${input.streamId}" is not open`);
      }
      if (stream.status !== "running") {
        throw new Error(`Resumable stream "${input.streamId}" is already ${stream.status}`);
      }
      const record = {
        streamId: input.streamId,
        eventId: stream.records.length + 1,
        event: input.event,
        createdAt: /* @__PURE__ */ new Date()
      };
      stream.records.push(record);
      for (const subscriber of stream.subscribers) {
        if (record.eventId > subscriber.after) {
          subscriber.queue.enqueue(record);
        }
      }
      return record;
    },
    subscribe(input) {
      const after = input.after ?? 0;
      return {
        [Symbol.asyncIterator]() {
          const stream = streams.get(input.streamId);
          const queue = createAsyncQueue();
          if (stream === void 0) {
            queue.close();
            return queue[Symbol.asyncIterator]();
          }
          for (const record of stream.records) {
            if (record.eventId > after) {
              queue.enqueue(record);
            }
          }
          if (stream.status !== "running") {
            queue.close();
            return queue[Symbol.asyncIterator]();
          }
          const subscriber = { after, queue };
          stream.subscribers.add(subscriber);
          const iterator = queue[Symbol.asyncIterator]();
          return {
            next: () => iterator.next(),
            async return() {
              stream.subscribers.delete(subscriber);
              queue.close();
              return { value: void 0, done: true };
            }
          };
        }
      };
    },
    async status(input) {
      return state(streams.get(input.streamId));
    },
    async close(input) {
      const stream = streams.get(input.streamId);
      if (stream === void 0) {
        return { status: input.status, lastEventId: 0 };
      }
      stream.status = input.status;
      closeSubscribers(stream);
      return state(stream);
    }
  };
}
async function drainResumableStream(events, options, openPromise) {
  const streamId = options.id;
  try {
    await openPromise;
    for await (const event of events) {
      await options.store.append({ streamId, event });
    }
    await options.store.close({ streamId, status: "completed" });
  } catch (error) {
    const event = errorEvent(error);
    try {
      await options.store.append({ streamId, event });
    } catch {
    }
    try {
      await options.store.close({ streamId, status: "error" });
    } catch {
    }
  }
}
function recordToEnvelope(record) {
  return {
    type: "stream_event",
    streamId: record.streamId,
    eventId: record.eventId,
    event: record.event
  };
}
function stateToEndEnvelope(streamId, state) {
  return {
    type: "stream_end",
    streamId,
    eventId: state.lastEventId,
    status: state.status
  };
}
function closeSubscribers(stream) {
  for (const subscriber of stream.subscribers) {
    subscriber.queue.close();
  }
  stream.subscribers.clear();
}
function createAsyncQueue() {
  const values = [];
  const waiters = [];
  let closed = false;
  let error;
  function flush() {
    while (waiters.length > 0 && values.length > 0) {
      const waiter = waiters.shift();
      const value = values.shift();
      waiter?.resolve({ value, done: false });
    }
    if (values.length > 0 || waiters.length === 0 || !closed) {
      return;
    }
    while (waiters.length > 0) {
      const waiter = waiters.shift();
      if (waiter === void 0) {
        continue;
      }
      if (error !== void 0) {
        waiter.reject(error);
      } else {
        waiter.resolve({ value: void 0, done: true });
      }
    }
  }
  return {
    enqueue(value) {
      if (closed) {
        return;
      }
      values.push(value);
      flush();
    },
    close() {
      closed = true;
      flush();
    },
    throw(thrown) {
      closed = true;
      error = thrown;
      flush();
    },
    [Symbol.asyncIterator]() {
      return {
        next() {
          if (values.length > 0) {
            const value = values.shift();
            return Promise.resolve({ value, done: false });
          }
          if (error !== void 0) {
            return Promise.reject(error);
          }
          if (closed) {
            return Promise.resolve({ value: void 0, done: true });
          }
          return new Promise((resolve, reject) => {
            waiters.push({ resolve, reject });
          });
        }
      };
    }
  };
}

// src/sse.ts
function createSseStream(events, options = {}) {
  validateSseOptions(options);
  const encoder = new TextEncoder();
  const iterator = events[Symbol.asyncIterator]();
  const serialize = options.serialize ?? serializeJson2;
  return new ReadableStream({
    start(controller) {
      if (options.retry !== void 0) {
        controller.enqueue(encoder.encode(`retry: ${options.retry}

`));
      }
    },
    async pull(controller) {
      try {
        const next = await iterator.next();
        if (next.done === true) {
          controller.close();
          return;
        }
        controller.enqueue(
          encoder.encode(formatSseEvent(next.value, serialize, options.eventName))
        );
      } catch (error) {
        const event = errorEvent(error);
        controller.enqueue(encoder.encode(formatSseEvent(event, serialize, options.eventName)));
        controller.close();
      }
    },
    async cancel() {
      await iterator.return?.();
    }
  });
}
function formatSseEvent(event, serialize, eventName) {
  const name = typeof eventName === "function" ? eventName(event) : eventName;
  const lines = [];
  if (name !== void 0 && name.length > 0) {
    validateSseEventName(name);
    lines.push(`event: ${name}`);
  }
  for (const line of serialize(event).split(/\r?\n/)) {
    lines.push(`data: ${line}`);
  }
  lines.push("", "");
  return lines.join("\n");
}
function serializeJson2(value) {
  return JSON.stringify(value);
}
function validateSseOptions(options) {
  if (options.retry !== void 0 && !isValidRetry(options.retry)) {
    throw new TypeError("SSE retry must be a finite non-negative integer");
  }
  if (typeof options.eventName === "string") {
    validateSseEventName(options.eventName);
  }
}
function isValidRetry(value) {
  return Number.isFinite(value) && Number.isInteger(value) && value >= 0;
}
function validateSseEventName(name) {
  for (const character of name) {
    const code = character.charCodeAt(0);
    if (code === 0 || code === 10 || code === 13) {
      throw new TypeError("SSE event names must not contain null bytes or line breaks");
    }
  }
}

// src/response.ts
function createEventStream(events, options = {}) {
  const format = options.format ?? "jsonl";
  const headers = new Headers(options.headers);
  if (!headers.has("cache-control")) {
    headers.set("cache-control", "no-cache, no-transform");
  }
  if (!headers.has("connection")) {
    headers.set("connection", "keep-alive");
  }
  if (!headers.has("x-accel-buffering")) {
    headers.set("x-accel-buffering", "no");
  }
  const eventsForResponse = options.resumable === void 0 ? events : createResumableStream(events, options.resumable);
  const body = format === "sse" ? createSseStream(
    eventsForResponse,
    options.sse
  ) : createJsonlStream(
    eventsForResponse,
    options.jsonl
  );
  if (!headers.has("content-type")) {
    headers.set(
      "content-type",
      format === "sse" ? "text/event-stream; charset=utf-8" : "application/x-ndjson; charset=utf-8"
    );
  }
  const responseInit = { headers };
  if (options.status !== void 0) {
    responseInit.status = options.status;
  }
  if (options.statusText !== void 0) {
    responseInit.statusText = options.statusText;
  }
  return new Response(body, responseInit);
}
function createUIStreamResponse(events, options = {}) {
  return createEventStream(events, options);
}
export {
  createEventStream,
  createJsonlStream,
  createMemoryResumableStreamStore,
  createResumableStream,
  createSseStream,
  createUIStreamResponse,
  resumeStreamEvents
};
