// @anvia/core@0.13.5/dist/chunk-2ODTMRHP.js
// src/hooks/control.ts
function createHook(hook) {
  return hook;
}
function cancelPrompt(reason) {
  return { type: "terminate", reason };
}
function skipTool(reason) {
  return { type: "skip", reason };
}
function requestToolApproval(options = {}) {
  const action = {
    type: "approval_request"
  };
  if (options.reason !== void 0) {
    action.reason = options.reason;
  }
  if (options.rejectMessage !== void 0) {
    action.rejectMessage = options.rejectMessage;
  }
  return action;
}
var runControl = {
  continue() {
    return { type: "continue" };
  },
  cancel(reason) {
    return cancelPrompt(reason);
  }
};
var toolCallControl = {
  run() {
    return { type: "continue" };
  },
  skip(reason) {
    return skipTool(reason);
  },
  cancel(reason) {
    return { type: "terminate", reason };
  },
  requestApproval(options) {
    return requestToolApproval(options);
  }
};

export {
  createHook,
  cancelPrompt,
  skipTool,
  requestToolApproval,
  runControl,
  toolCallControl
};
