// @anvia/core@0.13.5/dist/chunk-FE5QL4G3.js
// src/completion/json.ts
function isJsonValue(value) {
  return isJsonValueWithAncestors(value, /* @__PURE__ */ new WeakSet());
}
function isJsonValueWithAncestors(value, ancestors) {
  if (value === null || typeof value === "string" || typeof value === "boolean") {
    return true;
  }
  if (typeof value === "number") {
    return Number.isFinite(value);
  }
  if (typeof value !== "object" || ancestors.has(value)) {
    return false;
  }
  ancestors.add(value);
  try {
    if (Array.isArray(value)) {
      if (!hasOnlyDenseArrayProperties(value)) {
        return false;
      }
      for (let index = 0; index < value.length; index += 1) {
        if (!isJsonValueWithAncestors(value[index], ancestors)) {
          return false;
        }
      }
      return true;
    }
    const prototype = Object.getPrototypeOf(value);
    if (prototype !== Object.prototype && prototype !== null) {
      return false;
    }
    for (const key of Reflect.ownKeys(value)) {
      if (typeof key !== "string") {
        return false;
      }
      const descriptor = Object.getOwnPropertyDescriptor(value, key);
      if (descriptor === void 0 || !descriptor.enumerable || !("value" in descriptor) || !isJsonValueWithAncestors(descriptor.value, ancestors)) {
        return false;
      }
    }
    return true;
  } catch {
    return false;
  } finally {
    ancestors.delete(value);
  }
}
function hasOnlyDenseArrayProperties(value) {
  const keys = Reflect.ownKeys(value);
  if (keys.length !== value.length + 1 || keys.at(-1) !== "length") {
    return false;
  }
  for (let index = 0; index < value.length; index += 1) {
    const key = String(index);
    if (keys[index] !== key) {
      return false;
    }
    const descriptor = Object.getOwnPropertyDescriptor(value, key);
    if (descriptor === void 0 || !descriptor.enumerable || !("value" in descriptor)) {
      return false;
    }
  }
  return true;
}

// src/completion/types.ts
var UserContent = {
  text(text) {
    return { type: "text", text };
  },
  imageUrl(url, options = {}) {
    const image = { type: "image", source: { type: "url", url } };
    if (options.detail !== void 0) {
      image.detail = options.detail;
    }
    return image;
  },
  imageBase64(data, mediaType, options = {}) {
    const image = {
      type: "image",
      source: { type: "base64", data, mediaType }
    };
    if (options.detail !== void 0) {
      image.detail = options.detail;
    }
    return image;
  },
  documentUrl(url, mediaType, options = {}) {
    return {
      type: "document",
      source: options.filename === void 0 ? { type: "url", url, mediaType } : { type: "url", url, mediaType, filename: options.filename }
    };
  },
  documentBase64(data, mediaType, options = {}) {
    return {
      type: "document",
      source: options.filename === void 0 ? { type: "base64", data, mediaType } : { type: "base64", data, mediaType, filename: options.filename }
    };
  },
  documentText(text) {
    return { type: "text", text };
  }
};
var ToolContent = {
  toolResult(id, content, callIdOrOptions, toolName) {
    const normalized = typeof content === "string" ? [{ type: "text", text: content }] : content;
    const options = normalizeToolResultOptions(callIdOrOptions, toolName);
    const result = { type: "tool_result", id, content: normalized };
    if (options.callId !== void 0) {
      result.callId = options.callId;
    }
    if (options.toolName !== void 0) {
      result.toolName = options.toolName;
    }
    return result;
  }
};
function normalizeToolResultOptions(callIdOrOptions, toolName) {
  if (callIdOrOptions === void 0) {
    return toolName === void 0 ? {} : { toolName };
  }
  if (typeof callIdOrOptions === "string") {
    return toolName === void 0 ? { callId: callIdOrOptions } : { callId: callIdOrOptions, toolName };
  }
  return toolName === void 0 ? callIdOrOptions : { ...callIdOrOptions, toolName };
}
function serializeToolResultOutput(output) {
  if (typeof output === "string") {
    return output;
  }
  try {
    const serialized = JSON.stringify(output);
    return serialized === void 0 ? String(output) : serialized;
  } catch {
    return String(output);
  }
}
function isToolResultContentArray(value) {
  return Array.isArray(value) && value.length > 0 && value.every((item) => {
    if (typeof item !== "object" || item === null || !("type" in item)) {
      return false;
    }
    if (item.type === "text") {
      return "text" in item && typeof item.text === "string";
    }
    if (item.type === "image") {
      return "data" in item && typeof item.data === "string" && (!("mediaType" in item) || item.mediaType === void 0 || typeof item.mediaType === "string");
    }
    return false;
  });
}
var AssistantContent = {
  text(text) {
    return { type: "text", text };
  },
  imageUrl(url, options = {}) {
    const image = { type: "image", source: { type: "url", url } };
    if (options.detail !== void 0) {
      image.detail = options.detail;
    }
    return image;
  },
  imageBase64(data, mediaType, options = {}) {
    const image = {
      type: "image",
      source: { type: "base64", data, mediaType }
    };
    if (options.detail !== void 0) {
      image.detail = options.detail;
    }
    return image;
  },
  reasoning(text, id) {
    return id === void 0 ? { type: "reasoning", text } : { type: "reasoning", text, id };
  },
  reasoningFromContent(content, id) {
    const text = reasoningDisplayText(content);
    const reasoning = { type: "reasoning", text, content };
    return id === void 0 ? reasoning : { ...reasoning, id };
  },
  reasoningSummary(text, id) {
    return AssistantContent.reasoningFromContent([{ type: "summary", text }], id);
  },
  reasoningEncrypted(data, id) {
    return AssistantContent.reasoningFromContent([{ type: "encrypted", data }], id);
  },
  reasoningRedacted(data, id) {
    return AssistantContent.reasoningFromContent([{ type: "redacted", data }], id);
  },
  toolCall(id, name, args, callId) {
    const base = {
      type: "tool_call",
      id,
      function: {
        name,
        arguments: args
      }
    };
    return callId === void 0 ? base : { ...base, callId };
  }
};
function reasoningDisplayText(reasoning) {
  const content = Array.isArray(reasoning) ? reasoning : reasoning.content;
  if (content === void 0) {
    return Array.isArray(reasoning) ? "" : reasoning.text;
  }
  return content.flatMap((item) => {
    if (item.type === "text" || item.type === "summary") {
      return [item.text];
    }
    return [];
  }).join("");
}
var Message = {
  system(content, options = {}) {
    return { role: "system", content, ...messageMetadata(options) };
  },
  user(content, options = {}) {
    return {
      role: "user",
      content: typeof content === "string" ? [UserContent.text(content)] : content,
      ...messageMetadata(options)
    };
  },
  assistant(content, idOrOptions) {
    const normalized = typeof content === "string" ? [AssistantContent.text(content)] : content;
    const options = typeof idOrOptions === "string" ? { id: idOrOptions } : idOrOptions ?? {};
    const metadata = messageMetadata(options).metadata;
    const message = options.id === void 0 ? { role: "assistant", content: normalized } : { role: "assistant", id: options.id, content: normalized };
    if (metadata !== void 0) {
      message.metadata = metadata;
    }
    return message;
  },
  tool(content, options = {}) {
    return {
      role: "tool",
      content: Array.isArray(content) ? content : [content],
      ...messageMetadata(options)
    };
  },
  toolResult(id, output, options = {}) {
    const content = isToolResultContentArray(output) ? output : serializeToolResultOutput(output);
    return Message.tool(
      ToolContent.toolResult(id, content, {
        callId: options.callId,
        toolName: options.toolName
      }),
      { metadata: options.metadata }
    );
  }
};
function messageMetadata(options) {
  if (options.metadata === void 0) {
    return {};
  }
  if (!isJsonValue(options.metadata)) {
    throw new TypeError("Message metadata must be a strict JSON value.");
  }
  return { metadata: options.metadata };
}
var Usage = {
  empty() {
    return {
      inputTokens: 0,
      outputTokens: 0,
      totalTokens: 0,
      cachedInputTokens: 0,
      cacheCreationInputTokens: 0
    };
  },
  add(left, right) {
    return {
      inputTokens: left.inputTokens + right.inputTokens,
      outputTokens: left.outputTokens + right.outputTokens,
      totalTokens: left.totalTokens + right.totalTokens,
      cachedInputTokens: left.cachedInputTokens + right.cachedInputTokens,
      cacheCreationInputTokens: left.cacheCreationInputTokens + right.cacheCreationInputTokens
    };
  }
};
function getAssistantGenerationMetadata(message) {
  if (message.role !== "assistant" || !isJsonObjectValue(message.metadata)) {
    return void 0;
  }
  const frameworkMetadata = message.metadata.anvia;
  if (!isJsonObjectValue(frameworkMetadata)) {
    return void 0;
  }
  const generation = frameworkMetadata.generation;
  if (!isJsonObjectValue(generation) || typeof generation.provider !== "string" || typeof generation.model !== "string" || !isUsageValue(generation.usage)) {
    return void 0;
  }
  return {
    provider: generation.provider,
    model: generation.model,
    usage: { ...generation.usage }
  };
}
function isJsonObjectValue(value) {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}
function isUsageValue(value) {
  if (!isJsonObjectValue(value)) {
    return false;
  }
  return isNonnegativeFiniteNumber(value.inputTokens) && isNonnegativeFiniteNumber(value.outputTokens) && isNonnegativeFiniteNumber(value.totalTokens) && isNonnegativeFiniteNumber(value.cachedInputTokens) && isNonnegativeFiniteNumber(value.cacheCreationInputTokens);
}
function isNonnegativeFiniteNumber(value) {
  return typeof value === "number" && Number.isFinite(value) && value >= 0;
}
var CompletionCapabilityError = class extends Error {
  constructor(message) {
    super(message);
    this.name = "CompletionCapabilityError";
  }
};
function assertCompletionRequestSupported(model, request, options = {}) {
  const modelLabel = `${model.provider}:${request.model ?? model.defaultModel}`;
  const capabilities = model.capabilities;
  if (options.streaming === true && !capabilities.streaming) {
    throw new CompletionCapabilityError(`${modelLabel} does not support streaming completions.`);
  }
  if (request.tools.length > 0 && !capabilities.tools) {
    throw new CompletionCapabilityError(`${modelLabel} does not support tool definitions.`);
  }
  if (request.toolChoice !== void 0 && !capabilities.toolChoice) {
    throw new CompletionCapabilityError(`${modelLabel} does not support tool choice.`);
  }
  if (request.outputSchema !== void 0 && !capabilities.outputSchema) {
    throw new CompletionCapabilityError(`${modelLabel} does not support output schemas.`);
  }
  if (!capabilities.imageInput && requestHasImageInput(request)) {
    throw new CompletionCapabilityError(`${modelLabel} does not support image input.`);
  }
  if (!capabilities.documentInput && requestHasFileDocumentInput(request)) {
    throw new CompletionCapabilityError(`${modelLabel} does not support document file input.`);
  }
}
function textFromAssistantContent(content) {
  return content.flatMap((item) => item.type === "text" ? [item.text] : []).join("\n");
}
function requestHasImageInput(request) {
  return request.chatHistory.some(
    (message) => message.role === "system" ? false : message.content.some((content) => content.type === "image")
  );
}
function requestHasFileDocumentInput(request) {
  return request.chatHistory.some(
    (message) => message.role === "user" ? message.content.some(
      (content) => content.type === "document" && content.source.type !== "text"
    ) : false
  );
}

export {
  isJsonValue,
  UserContent,
  ToolContent,
  serializeToolResultOutput,
  isToolResultContentArray,
  AssistantContent,
  reasoningDisplayText,
  Message,
  Usage,
  getAssistantGenerationMetadata,
  CompletionCapabilityError,
  assertCompletionRequestSupported,
  textFromAssistantContent
};
