// @anvia/core@0.8.0/dist/evals/index.js
import {
  ExtractorBuilder
} from "../chunk-4KQ536VD.js";
import "../chunk-RAHDO2B2.js";
import "../chunk-EIBAVYOR.js";
import "../chunk-XUUY2L2D.js";
import "../chunk-XXT2UCAR.js";
import "../chunk-YK4WAAS4.js";
import "../chunk-USWULPZJ.js";
import "../chunk-VQE5PAAT.js";
import {
  compact
} from "../chunk-MMHG7WAM.js";
import {
  cosineSimilarity,
  embedText
} from "../chunk-FTO3W4UP.js";
import {
  mapWithConcurrency
} from "../chunk-OIMLU4SF.js";
import "../chunk-S5IGJ4RB.js";
import "../chunk-WJO6NVE2.js";
import "../chunk-WQKHFADH.js";

// src/evals/agent-target.ts
function agentEvalTarget(agent, options = {}) {
  return async (input, testCase) => {
    const prompt = options.prompt?.(input, testCase) ?? String(input);
    const response = await agent.prompt(prompt).send();
    return options.output === void 0 ? response : options.output(response, testCase);
  };
}

// src/evals/metric.ts
function defineMetric(metric) {
  return metric;
}

// src/evals/metrics.ts
import { z } from "zod";

// src/evals/format.ts
function defaultOutputValue(output) {
  if (typeof output === "object" && output !== null && "output" in output && typeof output.output === "string") {
    return output.output;
  }
  return output;
}
function stableComparable(value) {
  if (typeof value === "string") {
    return value;
  }
  return JSON.stringify(value);
}
function formatValue(value) {
  if (typeof value === "string") {
    return value;
  }
  try {
    return JSON.stringify(value);
  } catch {
    return String(value);
  }
}
function errorMessage(error) {
  return error instanceof Error ? error.message : String(error);
}

// src/evals/outcome.ts
var EvalOutcome = {
  pass(score, options = {}) {
    return compact({
      outcome: "pass",
      score,
      comment: options.comment,
      metadata: options.metadata
    });
  },
  fail(score, options = {}) {
    return compact({
      outcome: "fail",
      score,
      comment: options.comment,
      metadata: options.metadata
    });
  },
  invalid(reason, options = {}) {
    return compact({
      outcome: "invalid",
      reason,
      score: options.score,
      comment: options.comment,
      metadata: options.metadata
    });
  }
};

// src/evals/selectors.ts
async function resolveActual(selector, args) {
  return selector === void 0 ? defaultOutputValue(args.output) : selector(args);
}
async function resolveActualText(selector, args) {
  const value = selector === void 0 ? defaultOutputValue(args.output) : await selector(args);
  return typeof value === "string" ? value : JSON.stringify(value);
}
async function resolveExpected(selectorOrValue, args) {
  if (selectorOrValue === void 0) {
    return args.case.expected;
  }
  return typeof selectorOrValue === "function" ? selectorOrValue(args) : selectorOrValue;
}
async function resolveJudgePrompt(selector, args) {
  if (selector !== void 0) {
    return selector(args);
  }
  return [
    `Suite: ${args.suiteName}`,
    `Case: ${args.case.id}`,
    `Input: ${formatValue(args.case.input)}`,
    `Expected: ${formatValue(args.case.expected)}`,
    `Output: ${formatValue(defaultOutputValue(args.output))}`
  ].join("\n\n");
}

// src/evals/metrics.ts
function exactMatch(options = {}) {
  return {
    name: options.name ?? "exact_match",
    async evaluate(args) {
      const actual = await resolveActual(options.actual, args);
      const expected = await resolveExpected(options.expected, args);
      if (expected === void 0) {
        return EvalOutcome.invalid("No expected value provided for exact match.");
      }
      const passed = stableComparable(actual) === stableComparable(expected);
      return passed ? EvalOutcome.pass(true) : EvalOutcome.fail(false, { comment: `Expected ${formatValue(expected)}.` });
    }
  };
}
function contains(options = {}) {
  return {
    name: options.name ?? "contains",
    async evaluate(args) {
      const actual = await resolveActualText(options.actual, args);
      const expected = await resolveExpected(options.expected, args);
      if (expected === void 0) {
        return EvalOutcome.invalid("No expected value provided for contains.");
      }
      if (typeof expected !== "string" && !(expected instanceof RegExp)) {
        return EvalOutcome.invalid("Contains expected value must be a string or RegExp.");
      }
      const passed = expected instanceof RegExp ? expected.test(actual) : actual.includes(expected);
      return passed ? EvalOutcome.pass(true) : EvalOutcome.fail(false, { comment: `Output did not contain ${String(expected)}.` });
    }
  };
}
function semanticSimilarity(options) {
  return {
    name: options.name ?? "semantic_similarity",
    async evaluate(args) {
      const actual = await resolveActualText(options.actual, args);
      const expected = await resolveExpected(options.expected, args);
      if (expected === void 0) {
        return EvalOutcome.invalid("No expected value provided for semantic similarity.");
      }
      if (typeof expected !== "string") {
        return EvalOutcome.invalid("Semantic similarity expected value must be a string.");
      }
      const [actualEmbedding, expectedEmbedding] = await Promise.all([
        embedText(options.model, actual),
        embedText(options.model, expected)
      ]);
      const score = cosineSimilarity(actualEmbedding.vector, expectedEmbedding.vector);
      return score >= options.threshold ? EvalOutcome.pass(score) : EvalOutcome.fail(score, { comment: `Similarity below threshold ${options.threshold}.` });
    }
  };
}
function llmJudge(options) {
  const extractor = new ExtractorBuilder(options.model, options.schema).instructions(
    options.instructions ?? "Judge the eval case by the requested schema. Submit the judgment using the schema."
  ).retries(options.retries ?? 0).build();
  return {
    name: options.name ?? "llm_judge",
    async evaluate(args) {
      try {
        const judgment = await extractor.extract(await resolveJudgePrompt(options.prompt, args));
        return options.passes(judgment) ? EvalOutcome.pass(judgment) : EvalOutcome.fail(judgment);
      } catch (error) {
        return EvalOutcome.invalid(errorMessage(error));
      }
    }
  };
}
function llmScore(options) {
  const criteria = Array.isArray(options.criteria) ? options.criteria.join("\n") : options.criteria;
  const extractor = new ExtractorBuilder(
    options.model,
    z.object({
      score: z.number(),
      feedback: z.string()
    })
  ).instructions(
    options.instructions ?? `Score the eval case against these criteria:
${criteria}

Return a score between 0 and 1 and brief feedback.`
  ).retries(options.retries ?? 0).build();
  return {
    name: options.name ?? "llm_score",
    async evaluate(args) {
      try {
        const score = await extractor.extract(await resolveJudgePrompt(options.prompt, args));
        if (score.score < 0 || score.score > 1) {
          return EvalOutcome.invalid(`Score ${score.score} outside valid range [0, 1].`, {
            score
          });
        }
        return score.score >= options.threshold ? EvalOutcome.pass(score, { comment: score.feedback }) : EvalOutcome.fail(score, { comment: score.feedback });
      } catch (error) {
        return EvalOutcome.invalid(errorMessage(error));
      }
    }
  };
}

// src/evals/runner.ts
async function runEvalSuite(options) {
  const startedAt = Date.now();
  const results = await mapWithConcurrency(
    options.cases,
    Math.max(1, Math.trunc(options.concurrency ?? 1)),
    (testCase) => runEvalCase(options, testCase)
  );
  const counts = countOutcomes(results);
  return {
    name: options.name,
    results,
    ...counts,
    durationMs: Date.now() - startedAt
  };
}
async function runEvalCase(options, testCase) {
  let output;
  let targetError;
  try {
    output = await options.target(testCase.input, testCase);
  } catch (error) {
    targetError = error;
  }
  const metrics = [];
  for (const metric of options.metrics) {
    const outcome = targetError === void 0 ? await safeEvaluate(options.name, testCase, output, metric) : EvalOutcome.invalid(`Target failed: ${errorMessage(targetError)}`);
    const reporterErrors = await reportOutcome({
      suiteName: options.name,
      testCase,
      output,
      targetError,
      metric,
      outcome,
      reporters: options.reporters ?? [],
      failOnReporterError: options.failOnReporterError === true
    });
    metrics.push({ metricName: metric.name, outcome, reporterErrors });
  }
  return compact({
    case: testCase,
    output,
    targetError,
    metrics
  });
}
async function safeEvaluate(suiteName, testCase, output, metric) {
  try {
    return await metric.evaluate({ suiteName, case: testCase, output });
  } catch (error) {
    return EvalOutcome.invalid(errorMessage(error));
  }
}
async function reportOutcome(args) {
  const errors = [];
  for (const reporter of args.reporters) {
    try {
      await reporter.report({
        suiteName: args.suiteName,
        case: args.testCase,
        output: args.output,
        targetError: args.targetError,
        metric: args.metric,
        outcome: args.outcome
      });
    } catch (error) {
      if (args.failOnReporterError) {
        throw error;
      }
      errors.push(error);
    }
  }
  return errors;
}
function countOutcomes(results) {
  let passed = 0;
  let failed = 0;
  let invalid = 0;
  for (const result of results) {
    for (const metric of result.metrics) {
      if (metric.outcome.outcome === "pass") passed += 1;
      if (metric.outcome.outcome === "fail") failed += 1;
      if (metric.outcome.outcome === "invalid") invalid += 1;
    }
  }
  return { passed, failed, invalid };
}
export {
  EvalOutcome,
  agentEvalTarget,
  contains,
  defineMetric,
  exactMatch,
  llmJudge,
  llmScore,
  runEvalSuite,
  semanticSimilarity
};
