// @anvia/core@0.7.1/dist/chunk-S5IGJ4RB.js
import {
  Message,
  assertCompletionRequestSupported,
  textFromAssistantContent
} from "./chunk-WJO6NVE2.js";
import {
  toProviderJsonSchema
} from "./chunk-WQKHFADH.js";

// src/completion/create-completion.ts
function createCompletion(model, options) {
  return sendCompletion(model, options);
}
function createCompletionStream(model, options) {
  const request = toCompletionRequest(options);
  if (!isStreamingCompletionModel(model) || !model.capabilities.streaming) {
    throw new Error("This completion model does not support streaming");
  }
  assertCompletionRequestSupported(model, request, { streaming: true });
  return model.streamCompletion(request);
}
async function createParsedCompletion(model, options) {
  const { schema, ...completionOptions } = options;
  const request = toCompletionRequest(
    {
      ...completionOptions,
      outputSchema: toProviderJsonSchema(schema)
    },
    "createParsedCompletion"
  );
  assertCompletionRequestSupported(model, request);
  const response = await model.completion(request);
  const text = textFromAssistantContent(response.choice);
  return {
    data: parseCompletionData(text, schema),
    text,
    content: response.choice,
    usage: response.usage,
    response
  };
}
async function sendCompletion(model, options) {
  const request = toCompletionRequest(options);
  assertCompletionRequestSupported(model, request);
  const response = await model.completion(request);
  return {
    text: textFromAssistantContent(response.choice),
    content: response.choice,
    usage: response.usage,
    response
  };
}
function toCompletionRequest(options, helperName = "createCompletion") {
  const chatHistory = [...options.messages ?? [], ...messagesFromInput(options.input)];
  if (chatHistory.length === 0) {
    throw new Error(`${helperName} requires input or messages.`);
  }
  const request = {
    chatHistory,
    documents: [...options.documents ?? []],
    tools: [...options.tools ?? []]
  };
  if (options.instructions !== void 0 && options.instructions.length > 0) {
    request.instructions = options.instructions;
  }
  if (options.temperature !== void 0) request.temperature = options.temperature;
  if (options.maxTokens !== void 0) request.maxTokens = options.maxTokens;
  if (options.toolChoice !== void 0) request.toolChoice = options.toolChoice;
  if (options.outputSchema !== void 0) request.outputSchema = options.outputSchema;
  if (options.params !== void 0) request.additionalParams = options.params;
  return request;
}
function messagesFromInput(input) {
  if (input === void 0) {
    return [];
  }
  if (typeof input === "string") {
    return [Message.user(input)];
  }
  return Array.isArray(input) ? [...input] : [input];
}
function isStreamingCompletionModel(model) {
  return typeof model.streamCompletion === "function";
}
function parseCompletionData(text, schema) {
  let json;
  try {
    json = JSON.parse(text);
  } catch (error) {
    throw new Error("createParsedCompletion expected the model response to be valid JSON.", {
      cause: error
    });
  }
  return schema.parse(json);
}

// src/completion/documents.ts
function normalizeDocuments(documents) {
  if (documents.length === 0) {
    return void 0;
  }
  return Message.user(documents.map(formatDocument).join("\n"));
}
function formatDocument(document) {
  return `<file id: ${document.id}>
${formatDocumentBody(document)}
</file>
`;
}
function formatDocumentBody(document) {
  const metadata = formatMetadata(document.additionalProps);
  return metadata === void 0 ? document.text : `${metadata}
${document.text}`;
}
function formatMetadata(additionalProps) {
  if (additionalProps === void 0) {
    return void 0;
  }
  const entries = Object.entries(additionalProps).sort(
    ([left], [right]) => left.localeCompare(right)
  );
  if (entries.length === 0) {
    return void 0;
  }
  const metadata = entries.map(([key, value]) => `${key}: ${JSON.stringify(value)}`).join(" ");
  return `<metadata ${metadata} />`;
}

// src/completion/request.ts
var CompletionRequestBuilder = class {
  constructor(model, promptMessage) {
    this.model = model;
    this.promptMessage = promptMessage;
  }
  model;
  promptMessage;
  requestModel;
  instructionBlocks = [];
  history = [];
  docs = [];
  toolDefs = [];
  temp;
  maxTokenCount;
  choice;
  params;
  schema;
  modelOverride(model) {
    this.requestModel = model;
    return this;
  }
  instructions(instructions) {
    if (instructions !== void 0 && instructions.length > 0) {
      this.instructionBlocks.push(instructions);
    }
    return this;
  }
  messages(messages) {
    this.history.push(...messages);
    return this;
  }
  documents(documents) {
    this.docs.push(...documents);
    return this;
  }
  tools(tools) {
    this.toolDefs.push(...tools);
    return this;
  }
  temperature(temperature) {
    this.temp = temperature;
    return this;
  }
  maxTokens(maxTokens) {
    this.maxTokenCount = maxTokens;
    return this;
  }
  toolChoice(toolChoice) {
    this.choice = toolChoice;
    return this;
  }
  additionalParams(additionalParams) {
    this.params = additionalParams;
    return this;
  }
  outputSchema(outputSchema) {
    this.schema = outputSchema;
    return this;
  }
  build() {
    const instructions = this.buildInstructions();
    const request = {
      chatHistory: [...this.history, this.promptMessage],
      documents: [...this.docs],
      tools: [...this.toolDefs]
    };
    if (this.requestModel !== void 0) request.model = this.requestModel;
    if (instructions !== void 0) request.instructions = instructions;
    if (this.temp !== void 0) request.temperature = this.temp;
    if (this.maxTokenCount !== void 0) request.maxTokens = this.maxTokenCount;
    if (this.choice !== void 0) request.toolChoice = this.choice;
    if (this.params !== void 0) request.additionalParams = this.params;
    if (this.schema !== void 0) request.outputSchema = this.schema;
    return request;
  }
  async send() {
    const request = this.build();
    assertCompletionRequestSupported(this.model, request);
    return this.model.completion(request);
  }
  buildInstructions() {
    return this.instructionBlocks.length === 0 ? void 0 : this.instructionBlocks.join("\n\n");
  }
};

export {
  createCompletion,
  createCompletionStream,
  createParsedCompletion,
  isStreamingCompletionModel,
  normalizeDocuments,
  formatDocument,
  CompletionRequestBuilder
};
