// @anvia/core@0.8.0/dist/loaders/index.js
// src/loaders/documents.ts
function fileToDocument(file) {
  return {
    id: file.path,
    text: file.text,
    additionalProps: {
      source: file.path,
      mediaType: "text/plain"
    }
  };
}
async function fileLoaderToDocuments(loader) {
  const documents = [];
  for await (const file of loader) {
    documents.push(fileToDocument(file));
  }
  return documents;
}
function pdfToDocument(pdf) {
  return {
    id: pdf.path,
    text: pdf.text,
    additionalProps: {
      source: pdf.path,
      mediaType: "application/pdf"
    }
  };
}
async function pdfLoaderToDocuments(loader) {
  const documents = [];
  for await (const pdf of loader) {
    documents.push(pdfToDocument(pdf));
  }
  return documents;
}
function pdfPageToDocument(page) {
  return {
    id: `${page.path}#page=${page.pageNumber}`,
    text: page.text,
    additionalProps: {
      source: page.path,
      mediaType: "application/pdf",
      pageNumber: String(page.pageNumber)
    }
  };
}
async function pdfPageLoaderToDocuments(loader) {
  const documents = [];
  for await (const page of loader) {
    documents.push(pdfPageToDocument(page));
  }
  return documents;
}

// src/loaders/file.ts
import { readdir, readFile } from "fs/promises";
import { join } from "path";

// src/loaders/shared.ts
import { isAbsolute } from "path";
import { glob } from "tinyglobby";
function ok(value) {
  return { ok: true, value };
}
function err(error) {
  return { ok: false, error };
}
async function sortedGlob(pattern) {
  return (await glob(pattern, { absolute: isAbsolute(pattern), onlyFiles: true })).sort(
    (left, right) => left.localeCompare(right)
  );
}
function toUint8Array(bytes) {
  if (bytes instanceof Uint8Array) {
    return new Uint8Array(
      bytes.buffer.slice(bytes.byteOffset, bytes.byteOffset + bytes.byteLength)
    );
  }
  return new Uint8Array(bytes);
}

// src/loaders/file.ts
var FileLoader = class _FileLoader {
  constructor(source, mode, skipErrors) {
    this.source = source;
    this.mode = mode;
    this.skipErrors = skipErrors;
  }
  source;
  mode;
  skipErrors;
  static withGlob(pattern) {
    return new _FileLoader(
      async function* () {
        for (const path of await sortedGlob(pattern)) {
          yield ok({ path });
        }
      },
      "source",
      false
    );
  }
  static withDir(directory) {
    return new _FileLoader(
      async function* () {
        const entries = await readdir(directory, { withFileTypes: true });
        const files = entries.filter((entry) => entry.isFile()).map((entry) => join(directory, entry.name)).sort((left, right) => left.localeCompare(right));
        for (const path of files) {
          yield ok({ path });
        }
      },
      "source",
      false
    );
  }
  static fromBytes(bytes) {
    return _FileLoader.fromBytesMany([bytes]);
  }
  static fromBytesMany(bytes) {
    return new _FileLoader(
      async function* () {
        for (const value of bytes) {
          yield ok({ path: "<memory>", bytes: toUint8Array(value) });
        }
      },
      "source",
      false
    );
  }
  read() {
    return new _FileLoader(this.source, "read", this.skipErrors);
  }
  readWithPath() {
    return new _FileLoader(this.source, "readWithPath", this.skipErrors);
  }
  ignoreErrors() {
    return new _FileLoader(this.source, this.mode, true);
  }
  async *[Symbol.asyncIterator]() {
    for await (const item of this.source()) {
      const result = await mapFileResult(item, this.mode);
      if (!result.ok && this.skipErrors) {
        continue;
      }
      yield this.skipErrors && result.ok ? result.value : result;
    }
  }
};
async function mapFileResult(item, mode) {
  if (!item.ok) {
    return item;
  }
  if (mode === "source") {
    return item;
  }
  try {
    const text = await readFileSource(item.value);
    return mode === "read" ? ok(text) : ok({ path: item.value.path, text });
  } catch (error) {
    return err(error);
  }
}
async function readFileSource(source) {
  if ("bytes" in source) {
    return new TextDecoder().decode(source.bytes);
  }
  return readFile(source.path, "utf8");
}

// src/loaders/pdf.ts
import { readdir as readdir2, readFile as readFile2 } from "fs/promises";
import { join as join2 } from "path";
var PdfFileLoader = class _PdfFileLoader {
  constructor(source, mode, skipErrors) {
    this.source = source;
    this.mode = mode;
    this.skipErrors = skipErrors;
  }
  source;
  mode;
  skipErrors;
  static withGlob(pattern) {
    return new _PdfFileLoader(
      async function* () {
        for (const path of await sortedGlob(pattern)) {
          yield ok({ path });
        }
      },
      "source",
      false
    );
  }
  static withDir(directory) {
    return new _PdfFileLoader(
      async function* () {
        const entries = await readdir2(directory, { withFileTypes: true });
        const files = entries.filter((entry) => entry.isFile()).map((entry) => join2(directory, entry.name)).sort((left, right) => left.localeCompare(right));
        for (const path of files) {
          yield ok({ path });
        }
      },
      "source",
      false
    );
  }
  static fromBytes(bytes) {
    return _PdfFileLoader.fromBytesMany([bytes]);
  }
  static fromBytesMany(bytes) {
    return new _PdfFileLoader(
      async function* () {
        for (const value of bytes) {
          yield ok({ path: "<memory>", bytes: toUint8Array(value) });
        }
      },
      "source",
      false
    );
  }
  read() {
    return new _PdfFileLoader(this.source, "read", this.skipErrors);
  }
  readWithPath() {
    return new _PdfFileLoader(this.source, "readWithPath", this.skipErrors);
  }
  byPage() {
    const mode = this.mode === "readWithPath" ? "byPageWithPath" : "byPage";
    return new _PdfFileLoader(this.source, mode, this.skipErrors);
  }
  ignoreErrors() {
    return new _PdfFileLoader(this.source, this.mode, true);
  }
  async *[Symbol.asyncIterator]() {
    for await (const item of this.source()) {
      const results = await mapPdfResult(item, this.mode);
      for (const result of results) {
        if (!result.ok && this.skipErrors) {
          continue;
        }
        yield this.skipErrors && result.ok ? result.value : result;
      }
    }
  }
};
async function mapPdfResult(item, mode) {
  if (!item.ok) {
    return [item];
  }
  if (mode === "source") {
    return [item];
  }
  try {
    const pages = await readPdfPages(item.value);
    if (mode === "read") {
      return [ok(pages.map((page) => page.text).join(""))];
    }
    if (mode === "readWithPath") {
      return [ok({ path: item.value.path, text: pages.map((page) => page.text).join("") })];
    }
    if (mode === "byPageWithPath") {
      return pages.map((page) => ok({ path: item.value.path, ...page }));
    }
    return pages.map(ok);
  } catch (error) {
    return [err(error)];
  }
}
async function readPdfPages(source) {
  const pdfjs = await import("pdfjs-dist/legacy/build/pdf.mjs");
  const bytes = "bytes" in source ? source.bytes : toUint8Array(await readFile2(source.path));
  const loadingTask = pdfjs.getDocument({ data: bytes });
  const document = await loadingTask.promise;
  const pages = [];
  try {
    for (let index = 1; index <= document.numPages; index += 1) {
      const page = await document.getPage(index);
      const content = await page.getTextContent();
      const text = content.items.flatMap(
        (item) => typeof item === "object" && item !== null && "str" in item ? [String(item.str)] : []
      ).join("\n");
      pages.push({ pageNumber: index - 1, text: text.length > 0 ? `${text}
` : "" });
    }
  } finally {
    await loadingTask.destroy();
  }
  return pages;
}
export {
  FileLoader,
  PdfFileLoader,
  fileLoaderToDocuments,
  fileToDocument,
  pdfLoaderToDocuments,
  pdfPageLoaderToDocuments,
  pdfPageToDocument,
  pdfToDocument
};
