// @anvia/openai@0.3.16/dist/index.js
var __defProp = Object.defineProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};

// src/openai/index.ts
var openai_exports = {};
__export(openai_exports, {
  DALL_E_2: () => DALL_E_2,
  DALL_E_3: () => DALL_E_3,
  GPT_IMAGE_1: () => GPT_IMAGE_1,
  GPT_IMAGE_2: () => GPT_IMAGE_2,
  OpenAIAudioGenerationModel: () => OpenAIAudioGenerationModel,
  OpenAIChatCompletionModel: () => OpenAIChatCompletionModel,
  OpenAIClient: () => OpenAIClient,
  OpenAIEmbeddingModel: () => OpenAIEmbeddingModel,
  OpenAIImageGenerationModel: () => OpenAIImageGenerationModel,
  OpenAIResponsesCompletionModel: () => OpenAIResponsesCompletionModel,
  OpenAITranscriptionModel: () => OpenAITranscriptionModel,
  TTS_1: () => TTS_1,
  TTS_1_HD: () => TTS_1_HD,
  WHISPER_1: () => WHISPER_1
});

// src/utils.ts
function isPlainObject(value) {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}
function numberFrom(value) {
  return typeof value === "number" && Number.isFinite(value) ? value : 0;
}
function stringFrom(value) {
  return typeof value === "string" ? value : void 0;
}
function parseJsonValue(text) {
  try {
    return JSON.parse(text);
  } catch {
    return text;
  }
}
function schemaName(schema) {
  return typeof schema.title === "string" ? schema.title : "response_schema";
}

// src/openai/audio-generation.ts
var TTS_1 = "tts-1";
var TTS_1_HD = "tts-1-hd";
var OpenAIAudioGenerationModel = class {
  constructor(client, defaultModel = TTS_1) {
    this.client = client;
    this.defaultModel = defaultModel;
  }
  client;
  defaultModel;
  provider = "openai";
  async audioGeneration(request) {
    const params = {
      model: this.defaultModel,
      input: request.text,
      voice: request.voice,
      speed: request.speed
    };
    if (request.additionalParams !== void 0 && isPlainObject(request.additionalParams)) {
      Object.assign(params, request.additionalParams);
    }
    const response = await this.client.audio.speech.create(params);
    return {
      audio: new Uint8Array(await response.arrayBuffer()),
      mediaType: mediaTypeFromFormat(params.response_format),
      rawResponse: response
    };
  }
};
function mediaTypeFromFormat(format) {
  if (format === "wav") return "audio/wav";
  if (format === "flac") return "audio/flac";
  if (format === "opus") return "audio/opus";
  if (format === "aac") return "audio/aac";
  if (format === "pcm") return "audio/pcm";
  return "audio/mpeg";
}

// src/openai/chat-completion.ts
import {
  AssistantContent,
  assertCompletionRequestSupported,
  Usage
} from "@anvia/core/completion";

// src/request-messages.ts
import {
  Message,
  normalizeDocuments
} from "@anvia/core/completion";
function orderedRequestMessages(request, options = {}) {
  const messages = [];
  if (options.includeInstructionsAsSystem === true && request.instructions !== void 0) {
    messages.push(Message.system(request.instructions));
  }
  messages.push(...request.chatHistory.filter((message) => message.role === "system"));
  const documents = normalizeDocuments(request.documents);
  if (documents !== void 0) {
    messages.push(documents);
  }
  messages.push(...request.chatHistory.filter((message) => message.role !== "system"));
  return messages;
}

// src/openai/chat-completion.ts
var OpenAIChatCompletionModel = class {
  constructor(client, defaultModel = "openai/gpt-5.2") {
    this.client = client;
    this.defaultModel = defaultModel;
  }
  client;
  defaultModel;
  provider = "openai-chat";
  capabilities = {
    streaming: true,
    tools: true,
    toolChoice: true,
    imageInput: true,
    documentInput: false,
    outputSchema: true,
    reasoning: true
  };
  traceRequest(request, options = {}) {
    const params = toOpenAIChatCompletionParams(this.defaultModel, request);
    if (options.stream === true) {
      params.stream = true;
      const streamOptions = isPlainObject(params.stream_options) ? params.stream_options : {};
      params.stream_options = { ...streamOptions, include_usage: true };
    }
    return providerRequestSummary(params, request, options);
  }
  async completion(request) {
    assertCompletionRequestSupported(this, request);
    const params = toOpenAIChatCompletionParams(this.defaultModel, request);
    const response = await this.client.chat.completions.create(params);
    return fromOpenAIChatCompletionResponse(response);
  }
  async *streamCompletion(request) {
    assertCompletionRequestSupported(this, request, { streaming: true });
    const params = {
      ...toOpenAIChatCompletionParams(this.defaultModel, request),
      stream: true
    };
    const streamOptions = isPlainObject(params.stream_options) ? params.stream_options : {};
    params.stream_options = { ...streamOptions, include_usage: true };
    const stream = await this.client.chat.completions.create(params);
    for await (const chunk of stream) {
      for (const event of fromOpenAIChatCompletionStreamChunk(chunk)) {
        yield event;
      }
    }
  }
};
function toOpenAIChatCompletionParams(defaultModel, request) {
  const params = {
    model: request.model ?? defaultModel,
    messages: requestMessages(request).flatMap(messageToChatMessages)
  };
  if (request.tools.length > 0) {
    params.tools = request.tools.map(toolDefinitionToOpenAIChatCompletion);
  }
  if (request.temperature !== void 0) {
    params.temperature = request.temperature;
  }
  if (request.maxTokens !== void 0) {
    params.max_tokens = request.maxTokens;
  }
  if (request.toolChoice !== void 0) {
    params.tool_choice = toolChoiceToOpenAIChatCompletion(request.toolChoice);
  }
  if (request.outputSchema !== void 0) {
    params.response_format = {
      type: "json_schema",
      json_schema: {
        name: schemaName(request.outputSchema),
        strict: true,
        schema: request.outputSchema
      }
    };
  }
  if (request.additionalParams !== void 0 && isPlainObject(request.additionalParams)) {
    Object.assign(params, request.additionalParams);
  }
  return params;
}
function providerRequestSummary(params, request, options) {
  return compactJsonObject({
    provider: "openai-chat",
    api: "chat.completions",
    stream: options.stream === true,
    model: stringFrom(params.model),
    parameterKeys: Object.keys(params).sort(),
    messageCount: Array.isArray(params.messages) ? params.messages.length : void 0,
    toolCount: request.tools.length,
    toolNames: request.tools.map((tool) => tool.name),
    hasOutputSchema: request.outputSchema !== void 0,
    temperature: request.temperature,
    maxTokens: request.maxTokens,
    toolChoice: toolChoiceSummary(request.toolChoice),
    additionalParamKeys: isPlainObject(request.additionalParams) ? Object.keys(request.additionalParams).sort() : void 0
  });
}
function toolChoiceSummary(toolChoice) {
  if (toolChoice === void 0 || typeof toolChoice === "string") {
    return toolChoice;
  }
  return { type: toolChoice.type, name: toolChoice.name };
}
function compactJsonObject(values) {
  return Object.fromEntries(
    Object.entries(values).flatMap(([key, value]) => {
      if (value === void 0) {
        return [];
      }
      return [[key, toJsonValue(value)]];
    })
  );
}
function toJsonValue(value) {
  if (value === null || typeof value === "string" || typeof value === "number" || typeof value === "boolean") {
    return value;
  }
  if (Array.isArray(value)) {
    return value.map((item) => toJsonValue(item));
  }
  if (isPlainObject(value)) {
    return compactJsonObject(value);
  }
  return String(value);
}
function requestMessages(request) {
  return orderedRequestMessages(request, { includeInstructionsAsSystem: true });
}
function fromOpenAIChatCompletionResponse(response) {
  const raw = response;
  const choices = Array.isArray(raw.choices) ? raw.choices : [];
  const firstChoice = choices.find(isPlainObject);
  const message = isPlainObject(firstChoice?.message) ? firstChoice.message : {};
  const choice = [];
  if (typeof message.content === "string" && message.content.length > 0) {
    choice.push(AssistantContent.text(message.content));
  }
  if (typeof message.refusal === "string" && message.refusal.length > 0) {
    choice.push(AssistantContent.text(message.refusal));
  }
  const reasoning = stringFrom(message.reasoning) ?? stringFrom(message.reasoning_content);
  if (reasoning !== void 0 && reasoning.length > 0) {
    choice.push(AssistantContent.reasoning(reasoning));
  }
  const toolCalls = Array.isArray(message.tool_calls) ? message.tool_calls : [];
  for (const toolCall of toolCalls) {
    if (!isPlainObject(toolCall)) {
      continue;
    }
    const fn = isPlainObject(toolCall.function) ? toolCall.function : {};
    const id = typeof toolCall.id === "string" ? toolCall.id : crypto.randomUUID();
    const name = typeof fn.name === "string" ? fn.name : "";
    const argsText = typeof fn.arguments === "string" ? fn.arguments : "{}";
    choice.push(AssistantContent.toolCall(id, name, parseJsonValue(argsText)));
  }
  const result = {
    choice,
    usage: usageFromOpenAIChatCompletion(raw.usage),
    rawResponse: response
  };
  if (typeof raw.id === "string") {
    result.messageId = raw.id;
  }
  return result;
}
function fromOpenAIChatCompletionStreamChunk(chunk) {
  if (!isPlainObject(chunk)) {
    return [];
  }
  const events = [];
  const choices = Array.isArray(chunk.choices) ? chunk.choices : [];
  for (const choice of choices) {
    if (!isPlainObject(choice) || !isPlainObject(choice.delta)) {
      continue;
    }
    const delta = choice.delta;
    if (typeof delta.content === "string" && delta.content.length > 0) {
      events.push({ type: "text_delta", delta: delta.content });
    }
    if (typeof delta.refusal === "string" && delta.refusal.length > 0) {
      events.push({ type: "text_delta", delta: delta.refusal });
    }
    const reasoning = stringFrom(delta.reasoning) ?? stringFrom(delta.reasoning_content);
    if (reasoning !== void 0 && reasoning.length > 0) {
      events.push({ type: "reasoning_delta", delta: reasoning });
    }
    const toolCalls = Array.isArray(delta.tool_calls) ? delta.tool_calls : [];
    for (const toolCall of toolCalls) {
      if (!isPlainObject(toolCall)) {
        continue;
      }
      const fn = isPlainObject(toolCall.function) ? toolCall.function : {};
      const index = numberFrom(toolCall.index);
      const id = `tool_${index}`;
      events.push(
        toolCallDelta(id, {
          callId: stringFrom(toolCall.id),
          name: stringFrom(fn.name),
          argumentsDelta: stringFrom(fn.arguments)
        })
      );
    }
  }
  if (typeof chunk.id === "string") {
    events.push({ type: "message_id", id: chunk.id });
  }
  if (isPlainObject(chunk.usage)) {
    const response = {
      choice: [],
      usage: usageFromOpenAIChatCompletion(chunk.usage),
      rawResponse: chunk
    };
    if (typeof chunk.id === "string") {
      response.messageId = chunk.id;
    }
    events.push({ type: "final", response });
  }
  return events;
}
function usageFromOpenAIChatCompletion(usage) {
  const usageSource = isPlainObject(usage) ? usage : {};
  const promptDetails = isPlainObject(usageSource.prompt_tokens_details) ? usageSource.prompt_tokens_details : {};
  return {
    ...Usage.empty(),
    inputTokens: numberFrom(usageSource.prompt_tokens),
    outputTokens: numberFrom(usageSource.completion_tokens),
    totalTokens: numberFrom(usageSource.total_tokens),
    cachedInputTokens: numberFrom(promptDetails.cached_tokens)
  };
}
function messageToChatMessages(message) {
  if (message.role === "system") {
    return [{ role: "system", content: message.content }];
  }
  if (message.role === "user") {
    const contentParts = [];
    for (const content of message.content) {
      contentParts.push(...userContentToChatParts(content));
    }
    if (contentParts.length === 1 && contentParts[0]?.type === "text") {
      return [{ role: "user", content: contentParts[0].text }];
    } else if (contentParts.length > 0) {
      return [{ role: "user", content: contentParts }];
    }
    return [];
  }
  if (message.role === "tool") {
    return message.content.map(toolContentToChatMessage);
  }
  const text = message.content.flatMap((content) => content.type === "text" ? [content.text] : []).join("\n");
  const reasoning = message.content.flatMap((content) => content.type === "reasoning" ? [content.text] : []).filter((text2) => text2.length > 0).join("\n");
  if (message.content.some((content) => content.type === "image")) {
    throw new Error("OpenAI chat completions does not support image content in assistant history");
  }
  const toolCalls = message.content.filter((content) => content.type === "tool_call").map((content) => ({
    id: content.callId ?? content.id,
    type: "function",
    function: {
      name: content.function.name,
      arguments: JSON.stringify(content.function.arguments ?? {})
    }
  }));
  const chatMessage = {
    role: "assistant"
  };
  if (text.length > 0) {
    chatMessage.content = text;
  }
  if (reasoning.length > 0) {
    chatMessage.reasoning_content = reasoning;
  }
  if (toolCalls.length > 0) {
    chatMessage.tool_calls = toolCalls;
  }
  return [chatMessage];
}
function toolContentToChatMessage(content) {
  return {
    role: "tool",
    tool_call_id: content.callId ?? content.id,
    content: content.content.map(
      (item) => item.type === "text" ? item.text : `[image:${item.mediaType ?? "image/png"}]`
    ).join("\n")
  };
}
function userContentToChatParts(content) {
  if (content.type === "text") {
    return [{ type: "text", text: content.text }];
  }
  if (content.type === "image") {
    const image_url = { url: imageUrl(content) };
    if (content.detail !== void 0) {
      image_url.detail = content.detail;
    }
    return [{ type: "image_url", image_url }];
  }
  if (content.type === "document") {
    return documentToChatParts(content);
  }
  return [];
}
function imageUrl(image) {
  if (image.source.type === "url") {
    return image.source.url;
  }
  return `data:${image.source.mediaType};base64,${image.source.data}`;
}
function documentToChatParts(document) {
  if (document.source.type === "text") {
    return [{ type: "text", text: document.source.text }];
  }
  throw new Error("OpenAI chat completions does not support file document attachments");
}
function toolDefinitionToOpenAIChatCompletion(tool) {
  return {
    type: "function",
    function: {
      name: tool.name,
      description: tool.description,
      parameters: tool.parameters
    }
  };
}
function toolChoiceToOpenAIChatCompletion(toolChoice) {
  if (toolChoice === "auto" || toolChoice === "required" || toolChoice === "none") {
    return toolChoice;
  }
  return {
    type: "function",
    function: {
      name: toolChoice.name
    }
  };
}
function toolCallDelta(id, values) {
  const event = { type: "tool_call_delta", id };
  if (values.callId !== void 0) event.callId = values.callId;
  if (values.name !== void 0) event.name = values.name;
  if (values.argumentsDelta !== void 0) event.argumentsDelta = values.argumentsDelta;
  return event;
}

// src/openai/client.ts
import {
  ModelListingError
} from "@anvia/core/model-listing";
import OpenAI from "openai";

// src/openai/embedding.ts
var OpenAIEmbeddingModel = class {
  constructor(client, model, options = {}) {
    this.client = client;
    this.model = model;
    this.dimensions = options.dimensions;
    this.maxBatchSize = options.maxBatchSize ?? 1024;
    this.user = options.user;
  }
  client;
  model;
  dimensions;
  maxBatchSize;
  user;
  async embedTexts(texts) {
    const embeddings = [];
    for (let index = 0; index < texts.length; index += this.maxBatchSize) {
      const batch = texts.slice(index, index + this.maxBatchSize);
      embeddings.push(...await this.embedBatch(batch));
    }
    return embeddings;
  }
  async embedBatch(texts) {
    if (texts.length === 0) {
      return [];
    }
    const params = {
      model: this.model,
      input: texts
    };
    if (this.dimensions !== void 0) {
      params.dimensions = this.dimensions;
    }
    if (this.user !== void 0) {
      params.user = this.user;
    }
    const response = await this.client.embeddings.create(params);
    const data = embeddingDataFromResponse(response, texts.length);
    if (data.length !== texts.length) {
      throw new Error(
        `Embedding response length ${data.length} did not match input length ${texts.length}`
      );
    }
    return texts.map((document, index) => {
      const item = data[index];
      if (item === void 0) {
        throw new Error(`Embedding response did not contain index ${index}`);
      }
      return {
        document,
        vector: item.embedding
      };
    });
  }
};
function embeddingDataFromResponse(response, inputLength) {
  const raw = response;
  const data = Array.isArray(raw.data) ? raw.data : [];
  const byIndex = /* @__PURE__ */ new Map();
  for (const [position, item] of data.entries()) {
    if (!isObject(item) || typeof item.index !== "number" || !Number.isInteger(item.index)) {
      throw new Error(`Embedding response item ${position} contained an invalid index`);
    }
    const index = item.index;
    if (index < 0 || index >= inputLength) {
      throw new Error(
        `Embedding response index ${index} was outside input range 0..${inputLength - 1}`
      );
    }
    if (byIndex.has(index)) {
      throw new Error(`Embedding response contained duplicate index ${index}`);
    }
    if (!isNumberArray(item.embedding)) {
      throw new Error(`Embedding response item ${position} contained an invalid embedding`);
    }
    byIndex.set(index, { index, embedding: item.embedding });
  }
  return Array.from({ length: inputLength }, (_, index) => byIndex.get(index)).filter(
    (item) => item !== void 0
  );
}
function isObject(value) {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}
function isNumberArray(value) {
  return Array.isArray(value) && value.every((item) => typeof item === "number");
}

// src/openai/image-generation.ts
import { Buffer as Buffer2 } from "buffer";
var DALL_E_2 = "dall-e-2";
var DALL_E_3 = "dall-e-3";
var GPT_IMAGE_1 = "gpt-image-1";
var GPT_IMAGE_2 = "gpt-image-2";
var OpenAIImageGenerationModel = class {
  constructor(client, defaultModel = GPT_IMAGE_1) {
    this.client = client;
    this.defaultModel = defaultModel;
  }
  client;
  defaultModel;
  provider = "openai";
  async imageGeneration(request) {
    const params = {
      model: this.defaultModel,
      prompt: request.prompt,
      size: `${request.width}x${request.height}`
    };
    if (this.defaultModel === DALL_E_2 || this.defaultModel === DALL_E_3) {
      params.response_format = "b64_json";
    }
    if (request.additionalParams !== void 0 && isPlainObject(request.additionalParams)) {
      Object.assign(params, request.additionalParams);
    }
    const response = await this.client.images.generate(params);
    return imageResponseFromOpenAI(response);
  }
};
function imageResponseFromOpenAI(response) {
  const raw = response;
  const mediaType = mediaTypeFromFormat2(
    typeof raw.output_format === "string" ? raw.output_format : "png"
  );
  const data = Array.isArray(raw.data) ? raw.data : [];
  const images = data.flatMap((item) => {
    if (!isPlainObject(item) || typeof item.b64_json !== "string") {
      return [];
    }
    return [{ data: new Uint8Array(Buffer2.from(item.b64_json, "base64")), mediaType }];
  });
  const image = images[0]?.data;
  if (image === void 0) {
    if (data.some((item) => isPlainObject(item) && typeof item.url === "string")) {
      throw new Error(
        "OpenAI image generation response contained image URLs, which are not supported."
      );
    }
    throw new Error("OpenAI image generation response contained no base64 images.");
  }
  return {
    image,
    images,
    mediaType,
    rawResponse: response
  };
}
function mediaTypeFromFormat2(format) {
  if (format === "jpeg" || format === "jpg") {
    return "image/jpeg";
  }
  if (format === "webp") {
    return "image/webp";
  }
  return "image/png";
}

// src/openai/responses.ts
import {
  AssistantContent as AssistantContent2,
  assertCompletionRequestSupported as assertCompletionRequestSupported2,
  Usage as Usage2
} from "@anvia/core/completion";
var OpenAIResponsesCompletionModel = class {
  constructor(client, defaultModel = "gpt-5") {
    this.client = client;
    this.defaultModel = defaultModel;
  }
  client;
  defaultModel;
  provider = "openai";
  capabilities = {
    streaming: true,
    tools: true,
    toolChoice: true,
    imageInput: true,
    documentInput: true,
    outputSchema: true,
    reasoning: true
  };
  traceRequest(request, options = {}) {
    const params = toOpenAIResponsesParams(this.defaultModel, request);
    if (options.stream === true) {
      params.stream = true;
    }
    return providerRequestSummary2(params, request, options);
  }
  async completion(request) {
    assertCompletionRequestSupported2(this, request);
    const params = toOpenAIResponsesParams(this.defaultModel, request);
    const response = await this.client.responses.create(params);
    return fromOpenAIResponse(response);
  }
  async *streamCompletion(request) {
    assertCompletionRequestSupported2(this, request, { streaming: true });
    const params = { ...toOpenAIResponsesParams(this.defaultModel, request), stream: true };
    const stream = await this.client.responses.create(params);
    for await (const event of stream) {
      const mapped = fromOpenAIStreamEvent(event);
      if (mapped !== void 0) {
        yield mapped;
      }
    }
  }
};
function toOpenAIResponsesParams(defaultModel, request) {
  const params = {
    model: request.model ?? defaultModel,
    input: requestMessages2(request).flatMap(messageToResponsesInput)
  };
  if (request.instructions !== void 0) {
    params.instructions = request.instructions;
  }
  if (request.tools.length > 0) {
    params.tools = request.tools.map(toolDefinitionToOpenAI);
  }
  if (request.temperature !== void 0) {
    params.temperature = request.temperature;
  }
  if (request.maxTokens !== void 0) {
    params.max_output_tokens = request.maxTokens;
  }
  if (request.toolChoice !== void 0) {
    params.tool_choice = toolChoiceToOpenAI(request.toolChoice);
  }
  if (request.outputSchema !== void 0) {
    params.text = {
      format: {
        type: "json_schema",
        name: schemaName(request.outputSchema),
        strict: true,
        schema: request.outputSchema
      }
    };
  }
  if (request.additionalParams !== void 0 && isPlainObject(request.additionalParams)) {
    Object.assign(params, request.additionalParams);
  }
  return params;
}
function providerRequestSummary2(params, request, options) {
  return compactJsonObject2({
    provider: "openai",
    api: "responses",
    stream: options.stream === true,
    model: stringFrom(params.model),
    parameterKeys: Object.keys(params).sort(),
    inputCount: Array.isArray(params.input) ? params.input.length : void 0,
    toolCount: request.tools.length,
    toolNames: request.tools.map((tool) => tool.name),
    hasInstructions: typeof params.instructions === "string" && params.instructions.length > 0,
    hasOutputSchema: request.outputSchema !== void 0,
    temperature: request.temperature,
    maxTokens: request.maxTokens,
    toolChoice: toolChoiceSummary2(request.toolChoice),
    additionalParamKeys: isPlainObject(request.additionalParams) ? Object.keys(request.additionalParams).sort() : void 0
  });
}
function toolChoiceSummary2(toolChoice) {
  if (toolChoice === void 0 || typeof toolChoice === "string") {
    return toolChoice;
  }
  return { type: toolChoice.type, name: toolChoice.name };
}
function compactJsonObject2(values) {
  return Object.fromEntries(
    Object.entries(values).flatMap(([key, value]) => {
      if (value === void 0) {
        return [];
      }
      return [[key, toJsonValue2(value)]];
    })
  );
}
function toJsonValue2(value) {
  if (value === null || typeof value === "string" || typeof value === "number" || typeof value === "boolean") {
    return value;
  }
  if (Array.isArray(value)) {
    return value.map((item) => toJsonValue2(item));
  }
  if (isPlainObject(value)) {
    return compactJsonObject2(value);
  }
  return String(value);
}
function requestMessages2(request) {
  return orderedRequestMessages(request);
}
function fromOpenAIResponse(response) {
  const raw = response;
  const output = Array.isArray(raw.output) ? raw.output : [];
  const choice = [];
  for (const item of output) {
    if (!isPlainObject(item)) {
      continue;
    }
    if (item.type === "message") {
      choice.push(...messageOutputToAssistantContent(item));
    }
    if (item.type === "function_call") {
      const id = typeof item.id === "string" ? item.id : crypto.randomUUID();
      const callId = typeof item.call_id === "string" ? item.call_id : void 0;
      const name = typeof item.name === "string" ? item.name : "";
      const argsText = typeof item.arguments === "string" ? item.arguments : "{}";
      choice.push(AssistantContent2.toolCall(id, name, parseJsonValue(argsText), callId));
    }
    if (item.type === "reasoning") {
      choice.push(reasoningItemToAssistantContent(item));
    }
  }
  const usageSource = isPlainObject(raw.usage) ? raw.usage : {};
  const inputTokens = numberFrom(usageSource.input_tokens);
  const outputTokens = numberFrom(usageSource.output_tokens);
  const totalTokens = numberFrom(usageSource.total_tokens) || inputTokens + outputTokens;
  const details = isPlainObject(usageSource.input_tokens_details) ? usageSource.input_tokens_details : {};
  const result = {
    choice,
    usage: {
      ...Usage2.empty(),
      inputTokens,
      outputTokens,
      totalTokens,
      cachedInputTokens: numberFrom(details.cached_tokens)
    },
    rawResponse: response
  };
  if (typeof raw.id === "string") {
    result.messageId = raw.id;
  }
  return result;
}
function fromOpenAIStreamEvent(event) {
  if (!isPlainObject(event) || typeof event.type !== "string") {
    return void 0;
  }
  if (event.type === "response.output_text.delta" || event.type === "response.refusal.delta") {
    return typeof event.delta === "string" ? { type: "text_delta", delta: event.delta } : void 0;
  }
  if (event.type === "response.reasoning_text.delta" || event.type === "response.reasoning_summary_text.delta") {
    if (typeof event.delta !== "string") {
      return void 0;
    }
    const mapped = { type: "reasoning_delta", delta: event.delta };
    const id = stringFrom(event.item_id);
    if (id !== void 0) {
      mapped.id = id;
    }
    if (event.type === "response.reasoning_summary_text.delta") {
      mapped.contentType = "summary";
    } else {
      mapped.contentType = "text";
    }
    return mapped;
  }
  if (event.type === "response.output_item.added" && isPlainObject(event.item)) {
    const item = event.item;
    if (item.type === "function_call") {
      return toolCallDelta2(
        stringFrom(item.id) ?? stringFrom(event.item_id) ?? crypto.randomUUID(),
        {
          callId: stringFrom(item.call_id),
          name: stringFrom(item.name),
          argumentsDelta: typeof item.arguments === "string" ? item.arguments : void 0
        }
      );
    }
    if (typeof item.id === "string") {
      return { type: "message_id", id: item.id };
    }
  }
  if (event.type === "response.function_call_arguments.delta" || event.type === "response.function_call_arguments.done") {
    return toolCallDelta2(
      stringFrom(event.item_id) ?? stringFrom(event.output_item_id) ?? crypto.randomUUID(),
      {
        argumentsDelta: typeof event.delta === "string" ? event.delta : typeof event.arguments === "string" ? event.arguments : void 0
      }
    );
  }
  if (event.type === "response.output_item.done" && isPlainObject(event.item)) {
    const item = event.item;
    if (item.type === "function_call") {
      return {
        type: "tool_call",
        toolCall: AssistantContent2.toolCall(
          stringFrom(item.id) ?? crypto.randomUUID(),
          stringFrom(item.name) ?? "",
          parseJsonValue(typeof item.arguments === "string" ? item.arguments : "{}"),
          stringFrom(item.call_id)
        )
      };
    }
  }
  if ((event.type === "response.completed" || event.type === "response.incomplete") && isPlainObject(event.response)) {
    return {
      type: "final",
      response: fromOpenAIResponse(event.response)
    };
  }
  if (event.type === "response.failed" && isPlainObject(event.response)) {
    return { type: "error", error: event.response.error ?? event.response };
  }
  if (event.type === "error" || event.type === "response.error") {
    return { type: "error", error: event.error ?? event };
  }
  return void 0;
}
function messageToResponsesInput(message) {
  if (message.role === "system") {
    return [
      {
        role: "system",
        content: message.content
      }
    ];
  }
  if (message.role === "user") {
    const inputContent = [];
    for (const content of message.content) {
      inputContent.push(...userContentToOpenAIResponsesParts(content));
    }
    if (inputContent.length === 1 && inputContent[0]?.type === "input_text") {
      return [{ role: "user", content: inputContent[0].text }];
    } else if (inputContent.length > 0) {
      return [{ role: "user", content: inputContent }];
    }
    return [];
  }
  if (message.role === "tool") {
    return message.content.map(toolContentToOpenAIResponsesItem);
  }
  const items = [];
  const text = message.content.flatMap((content) => content.type === "text" ? [content.text] : []).join("\n");
  if (text.length > 0) {
    items.push({ role: "assistant", content: text });
  }
  for (const content of message.content) {
    if (content.type === "reasoning" && content.id !== void 0) {
      items.push(reasoningToOpenAIInput(content));
    }
    if (content.type === "tool_call") {
      items.push({
        type: "function_call",
        id: content.id,
        call_id: content.callId ?? content.id,
        name: content.function.name,
        arguments: JSON.stringify(content.function.arguments ?? {})
      });
    }
    if (content.type === "image") {
      throw new Error("OpenAI Responses does not support image content in assistant history");
    }
  }
  return items;
}
function toolContentToOpenAIResponsesItem(content) {
  return {
    type: "function_call_output",
    call_id: content.callId ?? content.id,
    output: toolResultContentToOpenAIResponsesOutput(content.content)
  };
}
function toolResultContentToOpenAIResponsesOutput(content) {
  if (content.every((item) => item.type === "text")) {
    return content.map((item) => item.text).join("\n");
  }
  return content.map((item) => {
    if (item.type === "text") {
      return { type: "input_text", text: item.text };
    }
    return {
      type: "input_image",
      image_url: `data:${item.mediaType ?? "image/png"};base64,${item.data}`,
      detail: "auto"
    };
  });
}
function reasoningItemToAssistantContent(item) {
  const content = reasoningContentFromOpenAIItem(item);
  const id = stringFrom(item.id);
  if (content.length === 0) {
    return AssistantContent2.reasoning("", id);
  }
  return AssistantContent2.reasoningFromContent(content, id);
}
function reasoningContentFromOpenAIItem(item) {
  const content = [];
  if (Array.isArray(item.content)) {
    for (const part of item.content) {
      if (!isPlainObject(part)) {
        continue;
      }
      if (part.type === "reasoning_text" && typeof part.text === "string") {
        content.push({ type: "text", text: part.text });
      }
    }
  }
  if (Array.isArray(item.summary)) {
    for (const summary of item.summary) {
      if (!isPlainObject(summary)) {
        continue;
      }
      if (typeof summary.text === "string") {
        content.push({ type: "summary", text: summary.text });
      }
    }
  }
  if (typeof item.encrypted_content === "string") {
    content.push({ type: "encrypted", data: item.encrypted_content });
  }
  return content;
}
function reasoningToOpenAIInput(reasoning) {
  const item = {
    type: "reasoning",
    id: reasoning.id,
    summary: reasoning.content?.filter((content) => {
      return content.type === "summary";
    }).map((content) => ({ type: "summary_text", text: content.text })) ?? []
  };
  const textContent = reasoning.content?.flatMap(
    (content) => content.type === "text" ? [{ type: "reasoning_text", text: content.text }] : []
  );
  if (textContent !== void 0 && textContent.length > 0) {
    item.content = textContent;
  }
  const encrypted = reasoning.content?.find((content) => content.type === "encrypted");
  if (encrypted?.type === "encrypted") {
    item.encrypted_content = encrypted.data;
  }
  return item;
}
function userContentToOpenAIResponsesParts(content) {
  if (content.type === "text") {
    return [{ type: "input_text", text: content.text }];
  }
  if (content.type === "image") {
    const part = { type: "input_image", image_url: imageUrl2(content) };
    if (content.detail !== void 0) {
      part.detail = content.detail;
    }
    return [part];
  }
  if (content.type === "document") {
    return [documentToOpenAIResponsesPart(content)];
  }
  return [];
}
function imageUrl2(image) {
  if (image.source.type === "url") {
    return image.source.url;
  }
  return `data:${image.source.mediaType};base64,${image.source.data}`;
}
function documentToOpenAIResponsesPart(document) {
  if (document.source.type === "text") {
    return { type: "input_text", text: document.source.text };
  }
  if (document.source.mediaType !== "application/pdf") {
    throw new Error(`OpenAI Responses only supports PDF document attachments`);
  }
  if (document.source.type === "url") {
    return { type: "input_file", file_url: document.source.url };
  }
  return {
    type: "input_file",
    file_data: `data:${document.source.mediaType};base64,${document.source.data}`,
    filename: document.source.filename ?? "document.pdf"
  };
}
function toolDefinitionToOpenAI(tool) {
  return {
    type: "function",
    name: tool.name,
    description: tool.description,
    parameters: tool.parameters
  };
}
function toolChoiceToOpenAI(toolChoice) {
  if (toolChoice === "auto" || toolChoice === "required" || toolChoice === "none") {
    return toolChoice;
  }
  return {
    type: "function",
    name: toolChoice.name
  };
}
function messageOutputToAssistantContent(item) {
  const content = Array.isArray(item.content) ? item.content : [];
  return content.flatMap((part) => {
    if (!isPlainObject(part)) {
      return [];
    }
    if (part.type === "output_text" && typeof part.text === "string") {
      return [AssistantContent2.text(part.text)];
    }
    if (part.type === "text" && typeof part.text === "string") {
      return [AssistantContent2.text(part.text)];
    }
    if (part.type === "refusal" && typeof part.refusal === "string") {
      return [AssistantContent2.text(part.refusal)];
    }
    return [];
  });
}
function toolCallDelta2(id, values) {
  const event = { type: "tool_call_delta", id };
  if (values.callId !== void 0) event.callId = values.callId;
  if (values.name !== void 0) event.name = values.name;
  if (values.argumentsDelta !== void 0) event.argumentsDelta = values.argumentsDelta;
  return event;
}

// src/openai/transcription.ts
import { toFile } from "openai";
var WHISPER_1 = "whisper-1";
var OpenAITranscriptionModel = class {
  constructor(client, defaultModel = WHISPER_1) {
    this.client = client;
    this.defaultModel = defaultModel;
  }
  client;
  defaultModel;
  provider = "openai";
  async transcription(request) {
    const params = {
      model: this.defaultModel,
      file: await toFile(request.data, request.filename)
    };
    if (request.language !== void 0) params.language = request.language;
    if (request.prompt !== void 0) params.prompt = request.prompt;
    if (request.temperature !== void 0) params.temperature = request.temperature;
    if (request.additionalParams !== void 0 && isPlainObject(request.additionalParams)) {
      Object.assign(params, request.additionalParams);
    }
    const response = await this.client.audio.transcriptions.create(params);
    return {
      text: transcriptionText(response),
      rawResponse: response
    };
  }
};
function transcriptionText(response) {
  if (typeof response === "string") {
    return response;
  }
  if (isPlainObject(response) && typeof response.text === "string") {
    return response.text;
  }
  throw new Error("OpenAI transcription response contained no text.");
}

// src/openai/client.ts
var OpenAIClient = class {
  client;
  completionApi;
  constructor(options = {}) {
    this.completionApi = options.completionApi ?? (options.baseUrl === void 0 ? "responses" : "chat");
    this.client = options.client ?? new OpenAI({
      apiKey: requireApiKey(options.apiKey),
      baseURL: options.baseUrl,
      defaultHeaders: options.headers
    });
  }
  completionModel(model = "gpt-5") {
    return this.completionApi === "chat" ? new OpenAIChatCompletionModel(this.client, model) : new OpenAIResponsesCompletionModel(this.client, model);
  }
  embeddingModel(model = "text-embedding-3-small", options = {}) {
    return new OpenAIEmbeddingModel(this.client, model, options);
  }
  imageGenerationModel(model = GPT_IMAGE_1) {
    return new OpenAIImageGenerationModel(this.client, model);
  }
  audioGenerationModel(model = TTS_1) {
    return new OpenAIAudioGenerationModel(this.client, model);
  }
  transcriptionModel(model = WHISPER_1) {
    return new OpenAITranscriptionModel(this.client, model);
  }
  async listModels() {
    try {
      const response = await this.client.models.list();
      const data = (await collectModelsFromResponse(response)).map(toListedModel).filter(isListedModel);
      return { data };
    } catch (error) {
      throw toModelListingError("OpenAI", error);
    }
  }
};
function requireApiKey(apiKey) {
  if (apiKey === void 0 || apiKey.length === 0) {
    throw new Error("Missing OpenAI credentials. Pass apiKey when constructing OpenAIClient.");
  }
  return apiKey;
}
async function collectModelsFromResponse(response) {
  if (isAsyncIterable(response)) {
    const models = [];
    for await (const model of response) {
      models.push(model);
    }
    return models;
  }
  if (Array.isArray(response)) {
    return response;
  }
  if (isObject2(response) && Array.isArray(response.data)) {
    return response.data;
  }
  return [];
}
function toListedModel(model) {
  if (!isObject2(model) || typeof model.id !== "string") {
    return void 0;
  }
  return {
    id: model.id,
    ...typeof model.name === "string" ? { name: model.name } : {},
    ...typeof model.description === "string" ? { description: model.description } : {},
    ...typeof model.type === "string" ? { type: model.type } : typeof model.object === "string" ? { type: model.object } : {},
    ...typeof model.created === "number" ? { createdAt: model.created } : {},
    ...typeof model.created_at === "number" ? { createdAt: model.created_at } : {},
    ...typeof model.owned_by === "string" ? { ownedBy: model.owned_by } : {},
    ...typeof model.context_length === "number" ? { contextLength: model.context_length } : {},
    ...typeof model.contextLength === "number" ? { contextLength: model.contextLength } : {}
  };
}
function isListedModel(model) {
  return model !== void 0;
}
function toModelListingError(provider, error) {
  if (error instanceof ModelListingError) {
    return error;
  }
  const statusCode = getStatusCode(error);
  return new ModelListingError(`${provider} model listing failed: ${getErrorMessage(error)}`, {
    provider,
    ...statusCode === void 0 ? {} : { statusCode },
    cause: error
  });
}
function getStatusCode(error) {
  if (!isObject2(error)) {
    return void 0;
  }
  if (typeof error.status === "number") {
    return error.status;
  }
  if (typeof error.statusCode === "number") {
    return error.statusCode;
  }
  return void 0;
}
function getErrorMessage(error) {
  return error instanceof Error ? error.message : String(error);
}
function isObject2(value) {
  return typeof value === "object" && value !== null;
}
function isAsyncIterable(value) {
  return isObject2(value) && Symbol.asyncIterator in value;
}
export {
  DALL_E_2,
  DALL_E_3,
  GPT_IMAGE_1,
  GPT_IMAGE_2,
  OpenAIAudioGenerationModel,
  OpenAIChatCompletionModel,
  OpenAIClient,
  OpenAIEmbeddingModel,
  OpenAIImageGenerationModel,
  OpenAIResponsesCompletionModel,
  OpenAITranscriptionModel,
  TTS_1,
  TTS_1_HD,
  WHISPER_1,
  openai_exports as openai
};
