// @anvia/core@0.12.4/dist/chunk-QKA6OQTH.js
import {
  compact
} from "./chunk-MMHG7WAM.js";
import {
  toProviderJsonSchema
} from "./chunk-WQKHFADH.js";
import {
  cosineSimilarity,
  embedText
} from "./chunk-JS5MGVNM.js";

// src/vector-store/index.ts
import { z } from "zod";

// src/tool/create-tool.ts
function createTool(options) {
  const parameters = toProviderJsonSchema(options.input);
  return {
    ...compact({
      name: options.name,
      approval: options.approval
    }),
    definition() {
      return {
        name: options.name,
        description: options.description,
        parameters
      };
    },
    async call(args, context = {}) {
      const parsedArgs = options.input.parse(args);
      const result = await options.execute(parsedArgs, context);
      return options.output === void 0 ? result : options.output.parse(result);
    },
    parseApprovalArgs(args) {
      return options.input.parse(args);
    }
  };
}

// src/vector-store/filter.ts
var vectorFilter = {
  eq(key, value) {
    return { type: "eq", key, value };
  },
  gt(key, value) {
    return { type: "gt", key, value };
  },
  lt(key, value) {
    return { type: "lt", key, value };
  },
  and(left, right) {
    return { type: "and", filters: [left, right] };
  },
  or(left, right) {
    return { type: "or", filters: [left, right] };
  }
};
function matchesVectorFilter(metadata, filter) {
  if (filter === void 0) {
    return true;
  }
  if (metadata === void 0) {
    return false;
  }
  switch (filter.type) {
    case "eq":
      return metadata[filter.key] === filter.value;
    case "gt":
      return compare(metadata[filter.key], filter.value) > 0;
    case "lt":
      return compare(metadata[filter.key], filter.value) < 0;
    case "and":
      return matchesVectorFilter(metadata, filter.filters[0]) && matchesVectorFilter(metadata, filter.filters[1]);
    case "or":
      return matchesVectorFilter(metadata, filter.filters[0]) || matchesVectorFilter(metadata, filter.filters[1]);
  }
}
function compare(left, right) {
  if (typeof left === "number" && typeof right === "number") {
    return left - right;
  }
  if (typeof left === "string" && typeof right === "string") {
    return left.localeCompare(right);
  }
  if (typeof left === "boolean" && typeof right === "boolean") {
    return Number(left) - Number(right);
  }
  return 0;
}

// src/vector-store/lsh.ts
var LshIndex = class {
  constructor(dimensions, options) {
    this.options = options;
    const rng = seededRandom(options.seed ?? 42);
    this.hyperplanes = [];
    for (let index = 0; index < options.numTables * options.numHyperplanes; index += 1) {
      const plane = Array.from({ length: dimensions }, () => rng() * 2 - 1);
      const norm = Math.sqrt(plane.reduce((sum, value) => sum + value ** 2, 0));
      this.hyperplanes.push(norm === 0 ? plane : plane.map((value) => value / norm));
    }
    this.tables = Array.from({ length: options.numTables }, () => /* @__PURE__ */ new Map());
  }
  options;
  hyperplanes;
  tables;
  insert(id, vector) {
    for (let table = 0; table < this.options.numTables; table += 1) {
      const hash = this.hash(vector, table);
      const bucket = this.tables[table]?.get(hash) ?? /* @__PURE__ */ new Set();
      bucket.add(id);
      this.tables[table]?.set(hash, bucket);
    }
  }
  query(vector) {
    const candidates = /* @__PURE__ */ new Set();
    for (let table = 0; table < this.options.numTables; table += 1) {
      const hash = this.hash(vector, table);
      for (const id of this.tables[table]?.get(hash) ?? []) {
        candidates.add(id);
      }
    }
    return candidates;
  }
  hash(vector, table) {
    let hash = "";
    const start = table * this.options.numHyperplanes;
    for (let offset = 0; offset < this.options.numHyperplanes; offset += 1) {
      const plane = this.hyperplanes[start + offset];
      const dot = vector.reduce((sum, value, index) => sum + value * (plane[index] ?? 0), 0);
      hash += dot >= 0 ? "1" : "0";
    }
    return hash;
  }
};
function seededRandom(seed) {
  let state = seed >>> 0;
  return () => {
    state = state * 1664525 + 1013904223 >>> 0;
    return state / 4294967296;
  };
}

// src/vector-store/index.ts
var InMemoryVectorStore = class _InMemoryVectorStore {
  documents = /* @__PURE__ */ new Map();
  indexStrategy;
  lshIndex;
  embeddingDimension;
  constructor(options = {}) {
    this.indexStrategy = options.index ?? { type: "bruteForce" };
  }
  static fromDocuments(documents, options = {}) {
    const store = new _InMemoryVectorStore(options);
    store.addDocuments(documents);
    return store;
  }
  addDocuments(documents) {
    this.validateDocumentDimensions(documents);
    for (const document of documents) {
      this.documents.set(document.id, document);
    }
    this.rebuildLshIndex();
    return this;
  }
  get(id) {
    return this.documents.get(id);
  }
  values() {
    return [...this.documents.values()];
  }
  len() {
    return this.documents.size;
  }
  isEmpty() {
    return this.documents.size === 0;
  }
  index(model) {
    return new InMemoryVectorIndex(model, this);
  }
  candidates(queryEmbedding) {
    this.validateQueryDimension(queryEmbedding);
    if (this.indexStrategy.type !== "lsh" || this.lshIndex === void 0) {
      return this.values();
    }
    const candidateIds = this.lshIndex.query(queryEmbedding.vector);
    if (candidateIds.size === 0) {
      return this.values();
    }
    return [...candidateIds].flatMap((id) => {
      const document = this.documents.get(id);
      return document === void 0 ? [] : [document];
    });
  }
  rebuildLshIndex() {
    if (this.indexStrategy.type !== "lsh") {
      this.lshIndex = void 0;
      return;
    }
    const firstEmbedding = this.values().flatMap((document) => document.embeddings)[0];
    if (firstEmbedding === void 0) {
      this.lshIndex = void 0;
      return;
    }
    const index = new LshIndex(firstEmbedding.vector.length, this.indexStrategy);
    for (const document of this.documents.values()) {
      for (const embedding of document.embeddings) {
        index.insert(document.id, embedding.vector);
      }
    }
    this.lshIndex = index;
  }
  validateDocumentDimensions(documents) {
    let dimension = this.embeddingDimension;
    for (const document of documents) {
      for (const embedding of document.embeddings) {
        dimension = validateEmbeddingDimension(dimension, embedding, document.id);
      }
    }
    this.embeddingDimension = dimension;
  }
  validateQueryDimension(queryEmbedding) {
    if (this.embeddingDimension === void 0) {
      return;
    }
    validateEmbeddingDimension(this.embeddingDimension, queryEmbedding, "query");
  }
};
var InMemoryVectorIndex = class {
  constructor(model, store) {
    this.model = model;
    this.store = store;
  }
  model;
  store;
  async search(request) {
    const queryEmbedding = await embedText(this.model, request.query);
    return this.store.candidates(queryEmbedding).filter((document) => matchesVectorFilter(document.metadata, request.filter)).flatMap((document) => {
      const score = bestScore(queryEmbedding, document.embeddings);
      if (score === void 0) {
        return [];
      }
      if (request.threshold !== void 0 && score < request.threshold) {
        return [];
      }
      return [
        {
          score,
          id: document.id,
          document: document.document,
          ...document.metadata !== void 0 && { metadata: document.metadata }
        }
      ];
    }).sort((left, right) => right.score - left.score).slice(0, Math.max(0, Math.trunc(request.topK)));
  }
  async searchIds(request) {
    return (await this.search(request)).map(({ score, id }) => ({ score, id }));
  }
  async inspect(request) {
    const limit = Math.max(0, Math.trunc(request.limit));
    const start = Math.max(0, Math.trunc(Number(request.cursor ?? "0")));
    const documents = this.store.values().filter((document) => matchesVectorFilter(document.metadata, request.filter));
    const page = documents.slice(start, start + limit);
    const nextOffset = start + page.length;
    return {
      items: page.map((document) => ({
        id: document.id,
        document: document.document,
        ...document.metadata !== void 0 && { metadata: document.metadata }
      })),
      ...nextOffset < documents.length ? { nextCursor: String(nextOffset) } : {},
      totalCount: documents.length
    };
  }
  asTool(options) {
    return createVectorSearchTool(this, options);
  }
};
function createVectorSearchTool(index, options) {
  return createTool({
    name: options.name,
    description: options.description ?? "Search a vector store for documents relevant to the provided query.",
    input: z.object({
      query: z.string().describe("The query string to search for relevant documents."),
      topK: z.number().int().positive().optional().describe("The maximum number of results.")
    }),
    output: z.array(
      z.object({
        score: z.number(),
        id: z.string(),
        document: z.any(),
        metadata: z.record(z.string(), z.union([z.string(), z.number(), z.boolean(), z.null()])).optional()
      })
    ),
    execute: ({ query, topK }) => index.search({
      query,
      topK: topK ?? options.topK ?? 5,
      threshold: options.threshold,
      filter: options.filter
    })
  });
}
function bestScore(queryEmbedding, embeddings) {
  let best;
  for (const embedding of embeddings) {
    const score = cosineSimilarity(queryEmbedding.vector, embedding.vector);
    best = best === void 0 ? score : Math.max(best, score);
  }
  return best;
}
function validateEmbeddingDimension(expectedDimension, embedding, id) {
  if (expectedDimension === void 0) {
    return embedding.vector.length;
  }
  if (embedding.vector.length !== expectedDimension) {
    throw new Error(
      `Vector dimension mismatch: expected ${expectedDimension} dimensions but received ${embedding.vector.length} for ${id}`
    );
  }
  return expectedDimension;
}

export {
  createTool,
  vectorFilter,
  InMemoryVectorStore,
  InMemoryVectorIndex,
  createVectorSearchTool
};
