// @anvia/core@0.12.4/dist/chunk-OQXRM5DO.js
import {
  PromptRequest
} from "./chunk-C7JLBS3F.js";
import {
  isSkillTool
} from "./chunk-YK4WAAS4.js";
import {
  ToolSet
} from "./chunk-YVQBC3SJ.js";
import {
  createTool
} from "./chunk-QKA6OQTH.js";
import {
  compact
} from "./chunk-MMHG7WAM.js";
import {
  isStreamingCompletionModel
} from "./chunk-SRBW3LD7.js";

// src/agent/agent.ts
import { z } from "zod";

// src/agent/ids.ts
function normalizeAgentId(id) {
  if (typeof id !== "string") {
    throw new TypeError("Agent id must be a string.");
  }
  const normalized = id.trim();
  if (normalized.length === 0) {
    throw new TypeError("Agent id must be a non-empty string.");
  }
  return normalized;
}

// src/agent/agent.ts
var DEFAULT_MAX_TURNS = 20;
var Agent = class {
  id;
  name;
  description;
  model;
  instructions;
  staticContext;
  temperature;
  maxTokens;
  additionalParams;
  toolSet;
  toolChoice;
  defaultMaxTurns;
  hook;
  outputSchema;
  observers;
  approvals;
  guardrails;
  dynamicContexts;
  dynamicTools;
  middlewares;
  /**
   * @deprecated Use `middlewares` instead.
   */
  toolMiddlewares;
  memory;
  eventStore;
  constructor(options) {
    this.id = normalizeAgentId(options.id);
    this.name = options.name;
    this.description = options.description;
    this.model = options.model;
    this.instructions = options.instructions;
    this.staticContext = [...options.staticContext ?? []];
    this.temperature = options.temperature;
    this.maxTokens = options.maxTokens;
    this.additionalParams = options.additionalParams;
    this.toolSet = options.toolSet ?? new ToolSet();
    this.toolChoice = options.toolChoice;
    this.defaultMaxTurns = options.defaultMaxTurns ?? DEFAULT_MAX_TURNS;
    this.hook = options.hook;
    this.outputSchema = options.outputSchema;
    this.observers = [...options.observers ?? []];
    this.approvals = options.approvals;
    this.guardrails = [...options.guardrails ?? []];
    this.dynamicContexts = [...options.dynamicContexts ?? []];
    this.dynamicTools = [...options.dynamicTools ?? []];
    this.middlewares = [...options.middlewares ?? options.toolMiddlewares ?? []];
    this.toolMiddlewares = this.middlewares;
    this.memory = options.memory;
    this.eventStore = options.eventStore;
  }
  prompt(prompt) {
    return PromptRequest.fromAgent(this, prompt);
  }
  session(sessionId, options = {}) {
    if (this.memory === void 0) {
      throw new Error(`Agent "${this.id}" has no memory store configured.`);
    }
    const normalized = sessionId.trim();
    if (normalized.length === 0) {
      throw new TypeError("Session id must be a non-empty string.");
    }
    return new AgentSession(this, {
      sessionId: normalized,
      ...options.userId !== void 0 && { userId: options.userId },
      ...options.metadata !== void 0 && { metadata: options.metadata }
    });
  }
  asTool(options) {
    const description = options.description ?? this.description ?? `Prompt the ${options.name} agent.`;
    return createTool({
      name: options.name,
      description,
      input: z.object({
        prompt: z.string().describe("The prompt to send to the agent.")
      }),
      output: z.string(),
      execute: async ({ prompt }, context) => {
        const request = this.prompt(prompt);
        const childRequest = options.maxTurns === void 0 ? request : request.maxTurns(options.maxTurns);
        if (options.stream === true && context.emitStreamEvent !== void 0 && this.model.capabilities.streaming && isStreamingCompletionModel(this.model)) {
          let output = "";
          for await (const event of childRequest.stream()) {
            await context.emitStreamEvent(
              compact({
                agentId: this.id,
                agentName: this.name,
                event
              })
            );
            if (event.type === "final") {
              output = event.output;
            }
          }
          return output;
        }
        const response = await childRequest.send();
        return response.output;
      }
    });
  }
  getTool(toolName) {
    const staticTool = this.toolSet.get(toolName);
    if (staticTool !== void 0) {
      return staticTool;
    }
    for (const registration of this.dynamicTools) {
      const dynamicTool = dynamicToolSetFromIndex(registration.index)?.get(toolName);
      if (dynamicTool !== void 0) {
        return dynamicTool;
      }
    }
    return void 0;
  }
  async callTool(toolName, args, context) {
    if (this.toolSet.contains(toolName)) {
      return this.toolSet.call(toolName, args, context);
    }
    for (const registration of this.dynamicTools) {
      const toolSet = dynamicToolSetFromIndex(registration.index);
      if (toolSet?.contains(toolName)) {
        return toolSet.call(toolName, args, context);
      }
    }
    return this.toolSet.call(toolName, args, context);
  }
  shouldApplyToolMiddleware(toolName) {
    return !isSkillTool(this.getTool(toolName));
  }
};
var AgentSession = class {
  constructor(agent, context) {
    this.agent = agent;
    this.context = context;
  }
  agent;
  context;
  prompt(prompt) {
    if (Array.isArray(prompt)) {
      throw new TypeError("AgentSession.prompt does not accept Message[] transcripts.");
    }
    return PromptRequest.fromAgent(this.agent, prompt, { memoryContext: this.context });
  }
  async messages() {
    const memory = this.agent.memory;
    if (memory === void 0) {
      throw new Error(`Agent "${this.agent.id}" has no memory store configured.`);
    }
    return memory.store.load(this.context);
  }
  async clear() {
    const memory = this.agent.memory;
    if (memory === void 0) {
      throw new Error(`Agent "${this.agent.id}" has no memory store configured.`);
    }
    await memory.store.clear(this.context);
  }
};
function dynamicToolSetFromIndex(index) {
  const maybeIndex = index;
  return maybeIndex.toolSet instanceof ToolSet ? maybeIndex.toolSet : void 0;
}

export {
  normalizeAgentId,
  DEFAULT_MAX_TURNS,
  Agent,
  AgentSession
};
