// @anvia/core@0.12.4/dist/chunk-CWUJUSOS.js
// src/guardrails/index.ts
function defineGuardrailPolicy(options) {
  return {
    id: options.id,
    mode: options.mode ?? "enforce",
    input: options.input ?? [],
    output: options.output ?? []
  };
}
function defineInputGuardrail(guardrail) {
  return guardrail;
}
function defineOutputGuardrail(guardrail) {
  return guardrail;
}
function allow(options = {}) {
  return { action: "allow", ...options };
}
function block(options) {
  return { action: "block", ...options };
}
var guardrails = {
  blockText,
  redactText
};
function normalizeGuardrailPolicies(policies) {
  if (policies === void 0) {
    return [];
  }
  return Array.isArray(policies) ? policies : [policies];
}
function appendGuardrailPolicies(current, next) {
  return [...current, ...normalizeGuardrailPolicies(next)];
}
function hasEnforcedOutputGuardrails(policies) {
  return policies.some((policy) => policy.mode === "enforce" && policy.output.length > 0);
}
async function runInputGuardrails(policies, context) {
  let prompt = context.prompt;
  let inputText = context.inputText;
  const decisions = [];
  for (const policy of policies) {
    for (const guardrail of policy.input) {
      const currentContext = { ...context, prompt, inputText };
      const result = await runGuardrail(
        policy,
        guardrail.id,
        "input",
        () => guardrail.check(currentContext, inputActions)
      );
      decisions.push(result.decision);
      if (policy.mode === "observe") {
        continue;
      }
      if (result.action.action === "block") {
        return {
          prompt,
          inputText,
          blocked: true,
          message: result.action.message,
          decisions
        };
      }
      if (result.action.action === "rewrite") {
        if (result.action.prompt !== void 0) {
          prompt = result.action.prompt;
          inputText = textFromMessage(prompt);
        } else if (result.action.inputText !== void 0) {
          inputText = result.action.inputText;
          prompt = rewriteMessageText(prompt, inputText);
        }
      }
    }
  }
  return { prompt, inputText, blocked: false, decisions };
}
async function runOutputGuardrails(policies, context) {
  let outputText = context.outputText;
  const decisions = [];
  for (const policy of policies) {
    for (const guardrail of policy.output) {
      const currentContext = { ...context, outputText };
      const result = await runGuardrail(
        policy,
        guardrail.id,
        "output",
        () => guardrail.check(currentContext, outputActions)
      );
      decisions.push(result.decision);
      if (policy.mode === "observe") {
        continue;
      }
      if (result.action.action === "block") {
        return {
          outputText,
          blocked: true,
          message: result.action.message,
          decisions
        };
      }
      if (result.action.action === "rewrite") {
        outputText = result.action.outputText;
      }
    }
  }
  return { outputText, blocked: false, decisions };
}
async function runGuardrail(policy, guardrailId, boundary, run) {
  const startedAt = Date.now();
  try {
    const action = await run() ?? allow();
    return {
      action,
      decision: decisionRecord({
        policy,
        guardrailId,
        boundary,
        action,
        latencyMs: Date.now() - startedAt
      })
    };
  } catch (error) {
    const message = error instanceof Error ? error.message : String(error);
    if (policy.mode === "observe") {
      return {
        action: allow(),
        decision: {
          policyId: policy.id,
          guardrailId,
          boundary,
          mode: policy.mode,
          action: "error",
          applied: false,
          reason: message,
          latencyMs: Date.now() - startedAt
        }
      };
    }
    throw error;
  }
}
function decisionRecord(args) {
  const applied = args.policy.mode === "enforce" && args.action.action !== "allow";
  const record = {
    policyId: args.policy.id,
    guardrailId: args.guardrailId,
    boundary: args.boundary,
    mode: args.policy.mode,
    action: args.action.action,
    applied,
    latencyMs: args.latencyMs
  };
  if (args.action.reason !== void 0) record.reason = args.action.reason;
  if ("message" in args.action && typeof args.action.message === "string") {
    record.message = args.action.message;
  }
  if (args.action.metadata !== void 0) record.metadata = args.action.metadata;
  return record;
}
var commonActions = {
  allow,
  block
};
var inputActions = {
  ...commonActions,
  rewrite(options) {
    return { action: "rewrite", ...options };
  }
};
var outputActions = {
  ...commonActions,
  rewrite(options) {
    return { action: "rewrite", ...options };
  }
};
function blockText(options) {
  return textPatternGuardrail(options, "block");
}
function redactText(options) {
  return textPatternGuardrail(options, "rewrite");
}
function textPatternGuardrail(options, action) {
  if (options.boundary === "input") {
    return defineInputGuardrail({
      id: options.id,
      check(ctx, actions) {
        return textPatternAction(
          ctx.inputText,
          options,
          action,
          (value) => actions.rewrite({ inputText: value, reason: options.reason })
        );
      }
    });
  }
  return defineOutputGuardrail({
    id: options.id,
    check(ctx, actions) {
      return textPatternAction(
        ctx.outputText,
        options,
        action,
        (value) => actions.rewrite({ outputText: value, reason: options.reason })
      );
    }
  });
}
function textPatternAction(text, options, action, rewrite) {
  const matched = options.patterns.some((pattern) => textPatternMatches(text, pattern));
  if (!matched) {
    return allow();
  }
  if (action === "block") {
    return block({
      reason: options.reason,
      message: options.message
    });
  }
  let current = text;
  for (const pattern of options.patterns) {
    current = replaceTextPattern(current, pattern, options.replacement ?? "[redacted]");
  }
  return rewrite(current);
}
function textPatternMatches(text, pattern) {
  if (typeof pattern === "string") {
    return text.includes(pattern);
  }
  pattern.lastIndex = 0;
  const matched = pattern.test(text);
  pattern.lastIndex = 0;
  return matched;
}
function replaceTextPattern(text, pattern, replacement) {
  if (typeof pattern === "string") {
    return text.split(pattern).join(replacement);
  }
  pattern.lastIndex = 0;
  const flags = pattern.flags.replace("y", "");
  const globalFlags = flags.includes("g") ? flags : `${flags}g`;
  return text.replace(new RegExp(pattern.source, globalFlags), replacement);
}
function textFromMessage(message) {
  if (message.role === "system") {
    return message.content;
  }
  return message.content.flatMap((content) => {
    if (content.type === "text") {
      return [content.text];
    }
    if (content.type === "document" && content.source.type === "text") {
      return [content.source.text];
    }
    return [];
  }).join("\n");
}
function rewriteMessageText(message, text) {
  if (message.role === "system") {
    return { ...message, content: text };
  }
  if (message.role === "user") {
    return {
      ...message,
      content: [
        { type: "text", text },
        ...message.content.filter(
          (item) => item.type !== "text" && !(item.type === "document" && item.source.type === "text")
        )
      ]
    };
  }
  if (message.role === "assistant") {
    return { ...message, content: [{ type: "text", text }] };
  }
  return message;
}

export {
  defineGuardrailPolicy,
  defineInputGuardrail,
  defineOutputGuardrail,
  allow,
  block,
  guardrails,
  normalizeGuardrailPolicies,
  appendGuardrailPolicies,
  hasEnforcedOutputGuardrails,
  runInputGuardrails,
  runOutputGuardrails
};
