// @anvia/core@0.12.4/dist/chunk-C7JLBS3F.js
import {
  toReadableStream
} from "./chunk-XUUY2L2D.js";
import {
  parseToolArgs,
  toolResultContentToText
} from "./chunk-YVQBC3SJ.js";
import {
  runControl,
  toolCallControl
} from "./chunk-NPZDYOE6.js";
import {
  compact
} from "./chunk-MMHG7WAM.js";
import {
  CompletionRequestBuilder,
  isStreamingCompletionModel
} from "./chunk-SRBW3LD7.js";
import {
  AssistantContent,
  Message,
  ToolContent,
  Usage,
  assertCompletionRequestSupported,
  textFromAssistantContent
} from "./chunk-5LFGEMIW.js";
import {
  mapWithConcurrency
} from "./chunk-OIMLU4SF.js";
import {
  appendGuardrailPolicies,
  hasEnforcedOutputGuardrails,
  runInputGuardrails,
  runOutputGuardrails
} from "./chunk-CWUJUSOS.js";

// src/request/errors.ts
var MaxTurnsError = class extends Error {
  constructor(maxTurns, chatHistory, prompt) {
    super(`Reached max turn limit: ${maxTurns}`);
    this.maxTurns = maxTurns;
    this.chatHistory = chatHistory;
    this.prompt = prompt;
    this.name = "MaxTurnsError";
  }
  maxTurns;
  chatHistory;
  prompt;
};
var PromptCancelledError = class extends Error {
  constructor(chatHistory, reason) {
    super(`Prompt cancelled: ${reason}`);
    this.chatHistory = chatHistory;
    this.reason = reason;
    this.name = "PromptCancelledError";
  }
  chatHistory;
  reason;
};
var ToolApprovalRequiredError = class extends Error {
  constructor(request) {
    super(`Tool approval required for ${request.toolName}, but no approval handler is configured.`);
    this.request = request;
    this.name = "ToolApprovalRequiredError";
  }
  request;
};

// src/internal/async-queue.ts
function createAsyncQueue() {
  const values = [];
  const waiters = [];
  let closed = false;
  let error;
  function flush() {
    while (waiters.length > 0 && values.length > 0) {
      const waiter = waiters.shift();
      const value = values.shift();
      if (waiter !== void 0) {
        waiter.resolve({ value, done: false });
      }
    }
    if (values.length > 0 || waiters.length === 0 || !closed) {
      return;
    }
    while (waiters.length > 0) {
      const waiter = waiters.shift();
      if (waiter === void 0) {
        continue;
      }
      if (error !== void 0) {
        waiter.reject(error);
      } else {
        waiter.resolve({ value: void 0, done: true });
      }
    }
  }
  return {
    enqueue(value) {
      if (closed) {
        return;
      }
      values.push(value);
      flush();
    },
    close() {
      closed = true;
      flush();
    },
    throw(thrown) {
      closed = true;
      error = thrown;
      flush();
    },
    [Symbol.asyncIterator]() {
      return {
        next() {
          if (values.length > 0) {
            const value = values.shift();
            return Promise.resolve({ value, done: false });
          }
          if (error !== void 0) {
            return Promise.reject(error);
          }
          if (closed) {
            return Promise.resolve({ value: void 0, done: true });
          }
          return new Promise((resolve, reject) => {
            waiters.push({ resolve, reject });
          });
        }
      };
    }
  };
}

// src/internal/prompt-runtime/memory.ts
var PromptRequestMemory = class {
  constructor(agent, memoryContext, initialHistory) {
    this.agent = agent;
    this.memoryContext = memoryContext;
    this.initialHistory = initialHistory;
  }
  agent;
  memoryContext;
  initialHistory;
  memoryPolicy() {
    return this.memory()?.options.savePolicy;
  }
  pendingTurnMessages(newMessages) {
    return this.memoryPolicy() === "turn" ? [...newMessages] : [];
  }
  async prepareRun(runId, newMessages) {
    const memory = this.memory();
    if (memory === void 0 || this.memoryContext === void 0) {
      return this.initialHistory;
    }
    const memoryHistory = await memory.store.load(this.memoryContext);
    const chatHistory = [...memoryHistory, ...this.initialHistory];
    if (memory.options.savePolicy === "message") {
      await memory.store.append({
        context: this.memoryContext,
        runId,
        turn: 1,
        messages: newMessages
      });
    }
    return chatHistory;
  }
  async commitMessages(runId, turn, messages, pendingTurnMessages) {
    const memory = this.memory();
    if (memory === void 0 || this.memoryContext === void 0 || messages.length === 0) {
      return;
    }
    if (memory.options.savePolicy === "message") {
      await memory.store.append({
        context: this.memoryContext,
        runId,
        turn,
        messages
      });
    } else if (memory.options.savePolicy === "turn") {
      pendingTurnMessages.push(...messages);
    }
  }
  async commitCompletedTurn(runId, turn, pendingTurnMessages) {
    const memory = this.memory();
    if (memory === void 0 || this.memoryContext === void 0 || memory.options.savePolicy !== "turn" || pendingTurnMessages.length === 0) {
      return;
    }
    await memory.store.append({
      context: this.memoryContext,
      runId,
      turn,
      messages: [...pendingTurnMessages]
    });
    pendingTurnMessages.length = 0;
  }
  async commitCompletedRun(runId, turn, newMessages, pendingTurnMessages) {
    await this.commitCompletedTurn(runId, turn, pendingTurnMessages);
    const memory = this.memory();
    if (memory === void 0 || this.memoryContext === void 0 || memory.options.savePolicy !== "run") {
      return;
    }
    await memory.store.append({
      context: this.memoryContext,
      runId,
      turn,
      messages: [...newMessages]
    });
  }
  async recordError(runId, error, newMessages) {
    const memory = this.memory();
    if (memory === void 0 || this.memoryContext === void 0) {
      return;
    }
    await memory.store.recordError?.({
      context: this.memoryContext,
      runId,
      error,
      messages: [...newMessages]
    });
  }
  memory() {
    return this.memoryContext === void 0 ? void 0 : this.agent.memory;
  }
};

// src/internal/prompt-runtime/retrieval.ts
async function fetchDynamicContext(agent, ragText) {
  if (ragText === void 0 || ragText.length === 0 || agent.dynamicContexts.length === 0) {
    return [];
  }
  const documents = [];
  for (const registration of agent.dynamicContexts) {
    const results = await registration.index.search({
      query: ragText,
      topK: registration.options.topK,
      threshold: registration.options.threshold,
      filter: registration.options.filter
    });
    for (const result of results) {
      const formatted = registration.options.format?.(result);
      if (formatted !== void 0) {
        documents.push(formatted);
      } else {
        const metadata = formatMetadata(result.metadata);
        documents.push(
          compact({
            id: result.id,
            text: typeof result.document === "string" ? result.document : JSON.stringify(result.document, null, 2),
            additionalProps: metadata
          })
        );
      }
    }
  }
  return documents;
}
async function fetchToolDefinitions(agent, ragText) {
  const staticDefinitions = await agent.toolSet.getToolDefinitions(ragText);
  if (ragText === void 0 || ragText.length === 0 || agent.dynamicTools.length === 0) {
    return staticDefinitions;
  }
  const definitions = [...staticDefinitions];
  const names = new Set(staticDefinitions.map((definition) => definition.name));
  for (const registration of agent.dynamicTools) {
    const results = await registration.index.search({
      query: ragText,
      topK: registration.options.topK,
      threshold: registration.options.threshold,
      filter: registration.options.filter
    });
    for (const result of results) {
      if (names.has(result.document.toolName)) {
        continue;
      }
      names.add(result.document.toolName);
      definitions.push(result.document.definition);
    }
  }
  return definitions;
}
function formatMetadata(metadata) {
  if (metadata === void 0) {
    return void 0;
  }
  return Object.fromEntries(Object.entries(metadata).map(([key, value]) => [key, String(value)]));
}

// src/internal/prompt-runtime/stream-accumulator.ts
var CompletionStreamAccumulator = class {
  orderedParts = [];
  textParts = /* @__PURE__ */ new Map();
  reasoningByKey = /* @__PURE__ */ new Map();
  reasoningKeyById = /* @__PURE__ */ new Map();
  toolCalls = /* @__PURE__ */ new Map();
  finalResponse;
  messageId;
  nextTextKey = 0;
  nextReasoningKey = 0;
  accept(event) {
    if (event.type === "text_delta") {
      this.appendText(event.delta);
      return { type: "text_delta", delta: event.delta };
    }
    if (event.type === "reasoning_delta") {
      const reasoning = this.reasoningStateForEvent(event);
      this.appendReasoning(reasoning, event);
      return reasoningDeltaEvent(event);
    }
    if (event.type === "tool_call_delta") {
      const toolCall = this.toolCallStateForId(event.id);
      if (event.callId !== void 0) toolCall.callId = event.callId;
      if (event.name !== void 0) toolCall.name = event.name;
      if (event.signature !== void 0) toolCall.signature = event.signature;
      if (event.argumentsDelta !== void 0) {
        toolCall.argumentsText += event.argumentsDelta;
      }
      return void 0;
    }
    if (event.type === "tool_call") {
      this.upsertToolCall(event.toolCall);
      return { type: "tool_call", toolCall: event.toolCall };
    }
    if (event.type === "message_id") {
      this.messageId = event.id;
      return void 0;
    }
    if (event.type === "final") {
      this.finalResponse = event.response;
      return void 0;
    }
    return void 0;
  }
  response() {
    const accumulatedResponse = this.buildAccumulatedResponse();
    if (this.finalResponse !== void 0) {
      if (accumulatedResponse.choice.length === 0) {
        return this.withMessageIdFallback(this.finalResponse, accumulatedResponse);
      }
      return this.mergeFinalResponse(accumulatedResponse, this.finalResponse);
    }
    return accumulatedResponse;
  }
  buildAccumulatedResponse() {
    const choice = [];
    for (const part of this.orderedParts) {
      if (part.type === "text") {
        const text = this.textParts.get(part.key) ?? "";
        if (text.length > 0) {
          choice.push({ type: "text", text });
        }
        continue;
      }
      if (part.type === "reasoning") {
        const reasoning = this.reasoningByKey.get(part.key);
        if (reasoning !== void 0) {
          choice.push(reasoningContent(reasoning));
        }
        continue;
      }
      const toolCall = this.toolCalls.get(part.key);
      if (toolCall !== void 0) {
        choice.push(toolCallContent(toolCall));
      }
    }
    const response = {
      choice,
      usage: Usage.empty(),
      rawResponse: void 0
    };
    if (this.messageId !== void 0) {
      response.messageId = this.messageId;
    }
    return response;
  }
  upsertToolCall(toolCall) {
    if (!this.toolCalls.has(toolCall.id)) {
      this.orderedParts.push({ type: "tool_call", key: toolCall.id });
    }
    const partial = {
      id: toolCall.id,
      name: toolCall.function.name,
      argumentsText: JSON.stringify(toolCall.function.arguments ?? {})
    };
    if (toolCall.callId !== void 0) {
      partial.callId = toolCall.callId;
    }
    if (toolCall.signature !== void 0) {
      partial.signature = toolCall.signature;
    }
    if (toolCall.additionalParams !== void 0) {
      partial.additionalParams = toolCall.additionalParams;
    }
    this.toolCalls.set(toolCall.id, partial);
  }
  mergeFinalResponse(accumulatedResponse, finalResponse) {
    if (finalResponse.choice.length === 0) {
      return this.withMessageIdFallback(
        {
          ...accumulatedResponse,
          usage: finalResponse.usage,
          rawResponse: finalResponse.rawResponse,
          ...finalResponse.messageId === void 0 ? {} : { messageId: finalResponse.messageId }
        },
        accumulatedResponse
      );
    }
    const finalById = /* @__PURE__ */ new Map();
    const finalByCallId = /* @__PURE__ */ new Map();
    for (const content of finalResponse.choice) {
      if (content.type !== "tool_call") {
        continue;
      }
      finalById.set(content.id, content);
      if (content.callId !== void 0) {
        finalByCallId.set(content.callId, content);
      }
    }
    const matchedFinalToolCalls = /* @__PURE__ */ new Set();
    const choice = accumulatedResponse.choice.map((content) => {
      if (content.type !== "tool_call") {
        return content;
      }
      const finalToolCall = finalById.get(content.id) ?? (content.callId === void 0 ? void 0 : finalByCallId.get(content.callId));
      if (finalToolCall === void 0) {
        return content;
      }
      matchedFinalToolCalls.add(finalToolCall);
      return mergeFinalToolCall(content, finalToolCall);
    });
    for (const content of finalResponse.choice) {
      if (content.type !== "tool_call") {
        continue;
      }
      if (!matchedFinalToolCalls.has(content)) {
        choice.push(content);
      }
    }
    return this.withMessageIdFallback({ ...finalResponse, choice }, accumulatedResponse);
  }
  appendText(delta) {
    const lastPart = this.orderedParts.at(-1);
    const key = lastPart?.type === "text" ? lastPart.key : this.createTextKey();
    if (lastPart?.type !== "text") {
      this.orderedParts.push({ type: "text", key });
    }
    this.textParts.set(key, `${this.textParts.get(key) ?? ""}${delta}`);
  }
  reasoningStateForEvent(event) {
    if (event.id !== void 0) {
      const existingKey = this.reasoningKeyById.get(event.id);
      if (existingKey !== void 0) {
        const existing = this.reasoningByKey.get(existingKey);
        if (existing !== void 0) {
          return existing;
        }
      }
      const key2 = this.createReasoningKey();
      const reasoning2 = { id: event.id, text: "" };
      this.reasoningKeyById.set(event.id, key2);
      this.reasoningByKey.set(key2, reasoning2);
      this.orderedParts.push({ type: "reasoning", key: key2 });
      return reasoning2;
    }
    const lastPart = this.orderedParts.at(-1);
    if (lastPart?.type === "reasoning") {
      const lastReasoning = this.reasoningByKey.get(lastPart.key);
      if (lastReasoning !== void 0 && lastReasoning.id === void 0) {
        return lastReasoning;
      }
    }
    const key = this.createReasoningKey();
    const reasoning = { text: "" };
    this.reasoningByKey.set(key, reasoning);
    this.orderedParts.push({ type: "reasoning", key });
    return reasoning;
  }
  toolCallStateForId(id) {
    const existing = this.toolCalls.get(id);
    if (existing !== void 0) {
      return existing;
    }
    const toolCall = {
      id,
      name: "",
      argumentsText: ""
    };
    this.toolCalls.set(id, toolCall);
    this.orderedParts.push({ type: "tool_call", key: id });
    return toolCall;
  }
  withMessageIdFallback(response, accumulatedResponse) {
    if (response.messageId !== void 0 || accumulatedResponse.messageId === void 0) {
      return response;
    }
    return { ...response, messageId: accumulatedResponse.messageId };
  }
  createTextKey() {
    this.nextTextKey += 1;
    return `text_${this.nextTextKey.toString()}`;
  }
  createReasoningKey() {
    this.nextReasoningKey += 1;
    return `reasoning_${this.nextReasoningKey.toString()}`;
  }
  appendReasoning(reasoning, event) {
    const contentType = event.contentType ?? "text";
    if (contentType === "text" || contentType === "summary") {
      reasoning.text += event.delta;
    }
    if (event.contentType === void 0 && event.signature === void 0) {
      return;
    }
    reasoning.content ??= [];
    const last = reasoning.content.at(-1);
    if (contentType === "text") {
      if (last?.type === "text") {
        last.text += event.delta;
        if (event.signature !== void 0) {
          last.signature = event.signature;
        }
      } else {
        reasoning.content.push(
          event.signature === void 0 ? { type: "text", text: event.delta } : { type: "text", text: event.delta, signature: event.signature }
        );
      }
      return;
    }
    if (contentType === "summary") {
      if (last?.type === "summary") {
        last.text += event.delta;
      } else {
        reasoning.content.push({ type: "summary", text: event.delta });
      }
      return;
    }
    if (contentType === "encrypted") {
      reasoning.content.push({ type: "encrypted", data: event.delta });
      return;
    }
    reasoning.content.push({ type: "redacted", data: event.delta });
  }
};
function reasoningContent(reasoning) {
  const content = reasoning.content === void 0 ? { type: "reasoning", text: reasoning.text } : { type: "reasoning", text: reasoning.text, content: reasoning.content };
  return reasoning.id === void 0 ? content : { ...content, id: reasoning.id };
}
function toolCallContent(toolCall) {
  const content = {
    type: "tool_call",
    id: toolCall.id,
    function: {
      name: toolCall.name,
      arguments: parseJsonValue(toolCall.argumentsText)
    }
  };
  if (toolCall.callId !== void 0) {
    content.callId = toolCall.callId;
  }
  if (toolCall.signature !== void 0) {
    content.signature = toolCall.signature;
  }
  if (toolCall.additionalParams !== void 0) {
    content.additionalParams = toolCall.additionalParams;
  }
  return content;
}
function mergeFinalToolCall(accumulated, finalToolCall) {
  const argumentsValue = isEmptyToolArguments(finalToolCall.function.arguments) ? accumulated.function.arguments : finalToolCall.function.arguments;
  return {
    ...accumulated,
    ...finalToolCall,
    function: {
      ...accumulated.function,
      ...finalToolCall.function,
      arguments: argumentsValue
    }
  };
}
function reasoningDeltaEvent(event) {
  const mapped = { type: "reasoning_delta", delta: event.delta };
  if (event.id !== void 0) mapped.id = event.id;
  if (event.contentType !== void 0) mapped.contentType = event.contentType;
  if (event.signature !== void 0) mapped.signature = event.signature;
  return mapped;
}
function isEmptyToolArguments(value) {
  if (value === void 0 || value === null) {
    return true;
  }
  if (typeof value === "string") {
    return value.trim().length === 0;
  }
  if (Array.isArray(value)) {
    return value.length === 0;
  }
  if (typeof value === "object") {
    return Object.values(value).every((item) => item === void 0);
  }
  return false;
}
function parseJsonValue(text) {
  if (text.trim().length === 0) {
    return {};
  }
  try {
    return JSON.parse(text);
  } catch {
    return text;
  }
}

// src/internal/prompt-runtime/stream-events.ts
function addTurn(turn, event) {
  if (event.type === "text_delta") {
    return { type: "text_delta", turn, delta: event.delta };
  }
  if (event.type === "reasoning_delta") {
    const mapped = { type: "reasoning_delta", turn, delta: event.delta };
    if (event.id !== void 0) mapped.id = event.id;
    if (event.contentType !== void 0) mapped.contentType = event.contentType;
    if (event.signature !== void 0) mapped.signature = event.signature;
    return mapped;
  }
  return { type: "tool_call", turn, toolCall: event.toolCall };
}
function isGenerationDeltaEvent(type) {
  return type === "text_delta" || type === "reasoning_delta" || type === "tool_call_delta" || type === "tool_call";
}

// src/internal/prompt-runtime/tool-execution.ts
var MCP_TOOL_METADATA_KEY = /* @__PURE__ */ Symbol.for("anvia.mcp.tool.metadata");
var ToolCallExecutor = class {
  constructor(agent, activeHook, approvals, runContext, concurrency, requestMiddlewares, cancel) {
    this.agent = agent;
    this.activeHook = activeHook;
    this.approvals = approvals;
    this.runContext = runContext;
    this.concurrency = concurrency;
    this.requestMiddlewares = requestMiddlewares;
    this.cancel = cancel;
  }
  agent;
  activeHook;
  approvals;
  runContext;
  concurrency;
  requestMiddlewares;
  cancel;
  async execute(toolCalls, onResult, onStreamEvent, observation) {
    return mapWithConcurrency(toolCalls, this.concurrency, async (toolCall) => {
      const args = JSON.stringify(toolCall.function.arguments ?? {});
      const internalCallId = globalThis.crypto.randomUUID();
      const hookArgs = {
        toolName: toolCall.function.name,
        internalCallId,
        args
      };
      if (toolCall.callId !== void 0) {
        hookArgs.toolCallId = toolCall.callId;
      }
      const tool = this.agent.getTool(toolCall.function.name);
      const toolDefinition = observation?.toolDefinitions?.find(
        (definition) => definition.name === toolCall.function.name
      );
      const toolMetadata = toolTraceMetadata(tool);
      const toolObservers = await observation?.runObservers.startTool(
        compact({
          turn: observation.turn,
          toolCall,
          toolName: toolCall.function.name,
          internalCallId,
          args,
          toolCallId: toolCall.callId,
          toolDefinition,
          toolMetadata
        })
      );
      let output;
      let skipped = false;
      let effectiveArgs = args;
      const callAction = await this.activeHook?.onToolCall?.({
        ...hookArgs,
        tool: toolCallControl
      });
      if (callAction?.type === "terminate") {
        await recordToolError(
          toolObservers,
          observation?.turn,
          toolCall,
          internalCallId,
          effectiveArgs,
          callAction.reason
        );
        throw this.cancel(callAction.reason);
      }
      if (callAction?.type === "skip") {
        output = callAction.reason;
        skipped = true;
      } else {
        let approvalDecision;
        try {
          effectiveArgs = await this.runToolInputMiddlewares({
            ...hookArgs,
            turn: observation?.turn ?? 0,
            originalArgs: args
          });
          hookArgs.args = effectiveArgs;
          approvalDecision = callAction?.type === "approval_request" ? await this.requestApproval(tool, hookArgs, callAction) : await this.evaluateToolApproval(tool, hookArgs) ?? { approved: true };
          if (!approvalDecision.approved) {
            output = approvalDecision.result;
            skipped = true;
          } else {
            output = await this.runApprovedToolCall(
              toolCall,
              hookArgs,
              effectiveArgs,
              args,
              toolObservers,
              observation,
              onStreamEvent,
              false
            );
            effectiveArgs = hookArgs.args;
          }
        } catch (error) {
          await recordToolError(
            toolObservers,
            observation?.turn,
            toolCall,
            internalCallId,
            effectiveArgs,
            error
          );
          throw error;
        }
      }
      let result = toolOutputToText(output);
      let structuredResult = toolOutputToStructuredResult(output);
      if (this.agent.shouldApplyToolMiddleware(toolCall.function.name)) {
        const middlewareReplacement = await this.runToolResultMiddlewares({
          ...hookArgs,
          args: effectiveArgs,
          result,
          originalResult: result,
          structuredResult,
          originalStructuredResult: structuredResult,
          turn: observation?.turn ?? 0
        });
        if (middlewareReplacement !== void 0) {
          output = middlewareReplacement;
          result = toolOutputToText(middlewareReplacement);
          structuredResult = toolOutputToStructuredResult(middlewareReplacement);
        }
      }
      const resultAction = await this.activeHook?.onToolResult?.({
        ...hookArgs,
        args: effectiveArgs,
        result,
        structuredResult,
        run: runControl
      });
      await toolObservers?.end({
        turn: observation?.turn ?? 0,
        toolCall,
        toolName: toolCall.function.name,
        internalCallId,
        args: effectiveArgs,
        result,
        structuredResult,
        skipped,
        toolCallId: toolCall.callId
      });
      if (resultAction?.type === "terminate") {
        throw this.cancel(resultAction.reason);
      }
      const resultPayload = {
        type: "tool_result",
        toolName: toolCall.function.name,
        internalCallId,
        args: effectiveArgs,
        result,
        structuredResult
      };
      if (toolCall.callId !== void 0) {
        resultPayload.toolCallId = toolCall.callId;
      }
      onResult?.(resultPayload);
      return ToolContent.toolResult(toolCall.id, output, toolCall.callId);
    });
  }
  async runApprovedToolCall(toolCall, hookArgs, effectiveArgs, originalArgs, toolObservers, observation, onStreamEvent, applyInputMiddleware = true) {
    const middlewareArgs = applyInputMiddleware ? await this.runToolInputMiddlewares({
      ...hookArgs,
      args: effectiveArgs,
      turn: observation?.turn ?? 0,
      originalArgs
    }) : effectiveArgs;
    hookArgs.args = middlewareArgs;
    try {
      return await this.agent.callTool(toolCall.function.name, middlewareArgs, {
        emitStreamEvent: async (event) => {
          await toolObservers?.streamEvent(
            compact({
              turn: observation?.turn ?? 0,
              toolCall,
              toolName: toolCall.function.name,
              internalCallId: hookArgs.internalCallId,
              args: middlewareArgs,
              toolCallId: toolCall.callId,
              event
            })
          );
          const payload = agentToolEventPayload(toolCall, hookArgs.internalCallId, event);
          if (payload !== void 0) {
            onStreamEvent?.(payload);
          }
        }
      });
    } catch (error) {
      const errorAction = await this.activeHook?.onToolError?.({
        ...hookArgs,
        args: middlewareArgs,
        error,
        run: runControl
      });
      await toolObservers?.error({
        turn: observation?.turn ?? 0,
        toolCall,
        toolName: toolCall.function.name,
        internalCallId: hookArgs.internalCallId,
        args: middlewareArgs,
        ...toolCall.callId !== void 0 && { toolCallId: toolCall.callId },
        error
      });
      if (errorAction?.type === "terminate") {
        throw this.cancel(errorAction.reason);
      }
      return error instanceof Error ? error.toString() : String(error);
    }
  }
  async runToolResultMiddlewares(args) {
    let result = args.result;
    let structuredResult = args.structuredResult;
    let replaced = false;
    for (const middleware of this.activeMiddlewares()) {
      const outputReplacement = await middleware.onToolOutput?.({
        ...args,
        result,
        structuredResult
      });
      if (outputReplacement !== void 0) {
        const normalized = normalizeToolOutputMiddlewareResult(outputReplacement);
        if (normalized.result !== void 0) {
          result = normalized.result;
          structuredResult = void 0;
        }
        if (normalized.structuredResult !== void 0) {
          structuredResult = normalized.structuredResult;
          result = toolResultContentToText(normalized.structuredResult);
        }
        replaced = true;
      }
      const legacyReplacement = await middleware.onResult?.({
        ...args,
        result,
        structuredResult
      });
      if (legacyReplacement !== void 0) {
        result = legacyReplacement;
        structuredResult = void 0;
        replaced = true;
      }
    }
    return replaced ? structuredResult ?? result : void 0;
  }
  async runToolInputMiddlewares(args) {
    let current = args.args;
    for (const middleware of this.activeMiddlewares()) {
      const replacement = await middleware.onToolInput?.({
        ...args,
        args: current
      });
      if (replacement?.args !== void 0) {
        current = typeof replacement.args === "string" ? replacement.args : JSON.stringify(replacement.args);
      }
    }
    return current;
  }
  activeMiddlewares() {
    return [...this.agent.middlewares, ...this.requestMiddlewares];
  }
  async evaluateToolApproval(tool, hookArgs) {
    if (tool?.approval === void 0) {
      return void 0;
    }
    const policy = tool.approval;
    const context = approvalContext(tool, hookArgs, this.agent, this.runContext);
    const required = await policy.when(context);
    if (!required) {
      return { approved: true };
    }
    const reason = await resolveApprovalText(policy.reason, context);
    const rejectMessage = await resolveApprovalText(policy.rejectMessage, context);
    return this.requestApproval(tool, hookArgs, compact({ reason, rejectMessage }));
  }
  async requestApproval(tool, hookArgs, options) {
    const request = compact({
      ...approvalContext(tool, hookArgs, this.agent, this.runContext),
      reason: options.reason,
      rejectMessage: options.rejectMessage
    });
    if (this.approvals === void 0) {
      throw new ToolApprovalRequiredError(request);
    }
    const decision = normalizeApprovalDecision(await this.approvals.handler(request));
    if (decision.approved) {
      return { approved: true };
    }
    return {
      approved: false,
      result: decision.rejectMessage ?? decision.reason ?? request.rejectMessage ?? "Tool approval was rejected."
    };
  }
};
function approvalContext(tool, hookArgs, agent, run) {
  const rawParsedArgs = parseToolArgs(hookArgs.args);
  const parsedArgs = tool?.parseApprovalArgs?.(rawParsedArgs) ?? rawParsedArgs;
  return compact({
    toolName: hookArgs.toolName,
    args: parsedArgs,
    rawArgs: hookArgs.args,
    toolCallId: hookArgs.toolCallId,
    internalCallId: hookArgs.internalCallId,
    run: compact({
      agentId: agent.id,
      runId: run.runId,
      sessionId: run.sessionId,
      metadata: run.metadata
    })
  });
}
function normalizeApprovalDecision(decision) {
  if (typeof decision === "boolean") {
    return { approved: decision };
  }
  return decision;
}
async function resolveApprovalText(value, context) {
  return typeof value === "function" ? value(context) : value;
}
function normalizeToolOutputMiddlewareResult(result) {
  if (typeof result === "string") {
    return { result };
  }
  return result ?? {};
}
function toolTraceMetadata(tool) {
  if (tool === void 0) {
    return void 0;
  }
  const metadata = tool[MCP_TOOL_METADATA_KEY];
  const mcpMetadata = typeof metadata === "object" && metadata !== null ? metadata : void 0;
  return {
    approvalRequired: tool.approval !== void 0,
    ...typeof mcpMetadata?.serverName === "string" && mcpMetadata.serverName.length > 0 ? { mcpServerName: mcpMetadata.serverName } : {}
  };
}
async function recordToolError(toolObservers, turn, toolCall, internalCallId, args, error) {
  await toolObservers?.error({
    turn: turn ?? 0,
    toolCall,
    toolName: toolCall.function.name,
    internalCallId,
    args,
    error,
    toolCallId: toolCall.callId
  });
}
function toolOutputToText(output) {
  return typeof output === "string" ? output : toolResultContentToText(output);
}
function toolOutputToStructuredResult(output) {
  return typeof output === "string" ? void 0 : output;
}
function agentToolEventPayload(toolCall, internalCallId, event) {
  if (typeof event.agentId !== "string" || event.agentId.length === 0) {
    return void 0;
  }
  return compact({
    type: "agent_tool_event",
    toolName: toolCall.function.name,
    toolCallId: toolCall.callId,
    internalCallId,
    agentId: event.agentId,
    agentName: event.agentName,
    event: event.event
  });
}

// src/internal/rag-text.ts
function extractRagText(message) {
  if (message.role === "user") {
    return message.content.flatMap((item) => item.type === "text" ? [item.text] : []).join("\n");
  }
  if (message.role === "tool") {
    return message.content.flatMap((item) => item.content.flatMap((part) => part.type === "text" ? [part.text] : [])).join("\n");
  }
  return void 0;
}

// src/observability/group.ts
async function startAgentRunObservers(registrations, args, failOnObserverError) {
  const runObservers = [];
  for (const registration of registrations) {
    try {
      const runObserver = await registration.observer.startRun(args);
      if (runObserver !== void 0) {
        runObservers.push(runObserver);
      }
    } catch (error) {
      if (failOnObserverError || registration.failOnObserverError === true) {
        throw error;
      }
    }
  }
  return new ActiveAgentRunObservers(runObservers, failOnObserverError);
}
var ActiveAgentRunObservers = class {
  constructor(runObservers, failOnObserverError) {
    this.runObservers = runObservers;
    this.failOnObserverError = failOnObserverError;
    this.trace = runObservers.find((observer) => observer.trace !== void 0)?.trace;
  }
  runObservers;
  failOnObserverError;
  trace;
  async startGeneration(args) {
    const generationObservers = [];
    for (const runObserver of this.runObservers) {
      if (runObserver.startGeneration === void 0) {
        continue;
      }
      try {
        const generationObserver = await runObserver.startGeneration(args);
        if (generationObserver !== void 0) {
          generationObservers.push(generationObserver);
        }
      } catch (error) {
        this.handleError(error);
      }
    }
    return new ActiveGenerationObservers(generationObservers, this.failOnObserverError);
  }
  async startTool(args) {
    const toolObservers = [];
    for (const runObserver of this.runObservers) {
      if (runObserver.startTool === void 0) {
        continue;
      }
      try {
        const toolObserver = await runObserver.startTool(args);
        if (toolObserver !== void 0) {
          toolObservers.push(toolObserver);
        }
      } catch (error) {
        this.handleError(error);
      }
    }
    return new ActiveToolObservers(toolObservers, this.failOnObserverError);
  }
  async end(args) {
    for (const runObserver of this.runObservers) {
      try {
        await runObserver.end(args);
      } catch (error) {
        this.handleError(error);
      }
    }
  }
  async error(args) {
    for (const runObserver of this.runObservers) {
      if (runObserver.error === void 0) {
        continue;
      }
      try {
        await runObserver.error(args);
      } catch (error) {
        this.handleError(error);
      }
    }
  }
  async event(args) {
    for (const runObserver of this.runObservers) {
      if (runObserver.event === void 0) {
        continue;
      }
      try {
        await runObserver.event(args);
      } catch (error) {
        this.handleError(error);
      }
    }
  }
  handleError(error) {
    if (this.failOnObserverError) {
      throw error;
    }
  }
};
var ActiveGenerationObservers = class {
  constructor(generationObservers, failOnObserverError) {
    this.generationObservers = generationObservers;
    this.failOnObserverError = failOnObserverError;
  }
  generationObservers;
  failOnObserverError;
  async end(args) {
    for (const observer of this.generationObservers) {
      try {
        await observer.end(args);
      } catch (error) {
        this.handleError(error);
      }
    }
  }
  async error(args) {
    for (const observer of this.generationObservers) {
      if (observer.error === void 0) {
        continue;
      }
      try {
        await observer.error(args);
      } catch (error) {
        this.handleError(error);
      }
    }
  }
  async update(args) {
    for (const observer of this.generationObservers) {
      if (observer.update === void 0) {
        continue;
      }
      try {
        await observer.update(args);
      } catch (error) {
        this.handleError(error);
      }
    }
  }
  handleError(error) {
    if (this.failOnObserverError) {
      throw error;
    }
  }
};
var ActiveToolObservers = class {
  constructor(toolObservers, failOnObserverError) {
    this.toolObservers = toolObservers;
    this.failOnObserverError = failOnObserverError;
  }
  toolObservers;
  failOnObserverError;
  async streamEvent(args) {
    for (const observer of this.toolObservers) {
      if (observer.streamEvent === void 0) {
        continue;
      }
      try {
        await observer.streamEvent(args);
      } catch (error) {
        this.handleError(error);
      }
    }
  }
  async end(args) {
    for (const observer of this.toolObservers) {
      try {
        await observer.end(args);
      } catch (error) {
        this.handleError(error);
      }
    }
  }
  async error(args) {
    for (const observer of this.toolObservers) {
      if (observer.error === void 0) {
        continue;
      }
      try {
        await observer.error(args);
      } catch (error) {
        this.handleError(error);
      }
    }
  }
  handleError(error) {
    if (this.failOnObserverError) {
      throw error;
    }
  }
};

// src/request/prompt-request.ts
var PromptRequest = class _PromptRequest {
  constructor(agent, promptMessage, initialHistory = [], memoryContext = void 0) {
    this.agent = agent;
    this.promptMessage = promptMessage;
    this.memoryContext = memoryContext;
    this.chatHistory = initialHistory;
    this.maxTurnCount = agent.defaultMaxTurns ?? 0;
    this.activeHook = agent.hook;
    this.approvalOptions = agent.approvals;
    this.guardrailPolicies = [...agent.guardrails];
    this.memoryRecorder = new PromptRequestMemory(agent, memoryContext, initialHistory);
  }
  agent;
  promptMessage;
  memoryContext;
  chatHistory;
  maxTurnCount;
  activeHook;
  approvalOptions;
  guardrailPolicies;
  guardrailDecisions = [];
  concurrency = 1;
  traceOptions;
  requestMiddlewares = [];
  steeringMessages = [];
  runState = "idle";
  memoryRecorder;
  static fromAgent(agent, prompt, options = {}) {
    const normalized = normalizePromptInput(prompt);
    return new _PromptRequest(agent, normalized.prompt, normalized.history, options.memoryContext);
  }
  maxTurns(maxTurns) {
    this.maxTurnCount = maxTurns;
    return this;
  }
  withHook(hook) {
    this.activeHook = hook;
    return this;
  }
  approvals(options) {
    this.approvalOptions = options;
    return this;
  }
  guardrails(policies) {
    this.guardrailPolicies = appendGuardrailPolicies(this.guardrailPolicies, policies);
    return this;
  }
  /**
   * @deprecated Use `withHook` instead.
   */
  requestHook(hook) {
    return this.withHook(hook);
  }
  withToolConcurrency(concurrency) {
    this.concurrency = Math.max(1, concurrency);
    return this;
  }
  withMiddleware(middleware) {
    this.requestMiddlewares.push(middleware);
    return this;
  }
  withMiddlewares(middlewares) {
    this.requestMiddlewares.push(...middlewares);
    return this;
  }
  /**
   * @deprecated Use `withMiddleware` instead.
   */
  withToolMiddleware(middleware) {
    return this.withMiddleware(middleware);
  }
  /**
   * @deprecated Use `withMiddlewares` instead.
   */
  withToolMiddlewares(middlewares) {
    return this.withMiddlewares(middlewares);
  }
  withTrace(trace) {
    this.traceOptions = trace;
    return this;
  }
  steer(input) {
    if (this.isTerminal()) {
      return false;
    }
    this.steeringMessages.push(...normalizeSteeringInput(input));
    return true;
  }
  async send() {
    this.startRun();
    const runId = globalThis.crypto.randomUUID();
    let usage = Usage.empty();
    let currentTurns = 0;
    let lastPrompt = this.promptMessage;
    let newMessages = [this.promptMessage];
    const runObservers = await this.startRunObservers();
    try {
      const inputResult = await runInputGuardrails(this.guardrailPolicies, {
        prompt: this.promptMessage,
        history: this.chatHistory,
        inputText: textFromMessage(this.promptMessage),
        run: this.guardrailRunContext(runId)
      });
      for (const decision of inputResult.decisions) {
        await this.recordGuardrailDecision(decision, runObservers);
      }
      this.promptMessage = inputResult.prompt;
      if (inputResult.blocked) {
        const output = inputResult.message ?? "The request was blocked by a guardrail.";
        const result = {
          output,
          usage: Usage.empty(),
          messages: [this.promptMessage, Message.assistant(output)],
          guardrails: [...this.guardrailDecisions]
        };
        await runObservers.end(result);
        this.runState = "completed";
        return result;
      }
      newMessages = [this.promptMessage];
      this.chatHistory = await this.memoryRecorder.prepareRun(runId, newMessages);
      const pendingTurnMessages = this.memoryRecorder.pendingTurnMessages(newMessages);
      await this.runRunStartHook(newMessages);
      while (currentTurns <= this.maxTurnCount + 1) {
        const prompt = newMessages.at(-1);
        if (prompt === void 0) {
          throw new Error("PromptRequest requires at least one message");
        }
        lastPrompt = prompt;
        currentTurns += 1;
        const historyForRequest = [...this.chatHistory, ...newMessages.slice(0, -1)];
        await this.runTurnStartHook(currentTurns, prompt, historyForRequest, newMessages);
        await this.runCompletionCallHook(prompt, historyForRequest, newMessages);
        const ragText = extractRagText(prompt);
        const dynamicContext = await fetchDynamicContext(this.agent, ragText);
        const toolDefs = await fetchToolDefinitions(this.agent, ragText);
        let request = new CompletionRequestBuilder(this.agent.model, prompt).instructions(this.agent.instructions).messages(historyForRequest).documents([...this.agent.staticContext, ...dynamicContext]).tools(toolDefs).temperature(this.agent.temperature).maxTokens(this.agent.maxTokens).additionalParams(this.agent.additionalParams).toolChoice(this.agent.toolChoice).outputSchema(this.agent.outputSchema).build();
        request = await this.runCompletionRequestMiddlewares(
          request,
          currentTurns
        );
        let response;
        try {
          response = await this.runCompletion(request, currentTurns, runObservers);
        } catch (error) {
          await this.runCompletionErrorHook(prompt, error, newMessages);
          throw error;
        }
        response = await this.runCompletionResponseMiddlewares(request, response, currentTurns);
        usage = Usage.add(usage, response.usage);
        await this.runCompletionResponseHook(prompt, response, newMessages);
        await this.runTurnEndHook(currentTurns, response, newMessages);
        const toolCalls = response.choice.filter(
          (item) => item.type === "tool_call"
        );
        const assistantMessage = Message.assistant(response.choice, response.messageId);
        newMessages.push(assistantMessage);
        if (toolCalls.length === 0) {
          if (this.steeringMessages.length > 0) {
            await this.memoryRecorder.commitMessages(
              runId,
              currentTurns,
              [assistantMessage],
              pendingTurnMessages
            );
          }
          if (await this.drainSteeringMessages(runId, currentTurns, newMessages, pendingTurnMessages)) {
            await this.memoryRecorder.commitCompletedTurn(runId, currentTurns, pendingTurnMessages);
            continue;
          }
          const guardedOutput = await this.runOutputGuardrailsForResponse(
            runId,
            usage,
            response,
            newMessages,
            runObservers
          );
          response = guardedOutput.response;
          const finalAssistantMessage = Message.assistant(response.choice, response.messageId);
          newMessages[newMessages.length - 1] = finalAssistantMessage;
          await this.memoryRecorder.commitMessages(
            runId,
            currentTurns,
            [finalAssistantMessage],
            pendingTurnMessages
          );
          const result = {
            output: guardedOutput.output,
            usage,
            messages: [...newMessages],
            trace: runObservers.trace,
            guardrails: [...this.guardrailDecisions]
          };
          await this.runRunEndHook(result, newMessages);
          await runObservers.end(result);
          await this.memoryRecorder.commitCompletedRun(
            runId,
            currentTurns,
            newMessages,
            pendingTurnMessages
          );
          this.runState = "completed";
          return result;
        }
        await this.memoryRecorder.commitMessages(
          runId,
          currentTurns,
          [assistantMessage],
          pendingTurnMessages
        );
        const toolResults = await this.executeToolCalls(
          runId,
          toolCalls,
          newMessages,
          void 0,
          void 0,
          {
            turn: currentTurns,
            runObservers,
            toolDefinitions: request.tools
          }
        );
        const toolMessage = Message.tool(toolResults);
        newMessages.push(toolMessage);
        await this.memoryRecorder.commitMessages(
          runId,
          currentTurns,
          [toolMessage],
          pendingTurnMessages
        );
        await this.drainSteeringMessages(runId, currentTurns, newMessages, pendingTurnMessages);
        await this.memoryRecorder.commitCompletedTurn(runId, currentTurns, pendingTurnMessages);
      }
      throw new MaxTurnsError(this.maxTurnCount, [...this.chatHistory, ...newMessages], lastPrompt);
    } catch (error) {
      const finalError = await this.runRunErrorHook(error, usage, newMessages);
      this.runState = finalError instanceof PromptCancelledError ? "cancelled" : "errored";
      await runObservers.error({ error: finalError, usage, messages: [...newMessages] });
      await this.memoryRecorder.recordError(runId, finalError, newMessages);
      throw finalError;
    }
  }
  async *stream() {
    if (!this.agent.model.capabilities.streaming || !isStreamingCompletionModel(this.agent.model)) {
      throw new Error("This completion model does not support streaming");
    }
    this.startRun();
    const runId = globalThis.crypto.randomUUID();
    const emit = async (event) => {
      await this.recordAgentEvent(runId, event);
      return event;
    };
    let usage = Usage.empty();
    let currentTurns = 0;
    let lastPrompt = this.promptMessage;
    let newMessages = [this.promptMessage];
    const runObservers = await this.startRunObservers();
    const bufferOutputDeltas = hasEnforcedOutputGuardrails(this.guardrailPolicies);
    try {
      const inputResult = await runInputGuardrails(this.guardrailPolicies, {
        prompt: this.promptMessage,
        history: this.chatHistory,
        inputText: textFromMessage(this.promptMessage),
        run: this.guardrailRunContext(runId)
      });
      for (const decision of inputResult.decisions) {
        await this.recordGuardrailDecision(decision, runObservers);
        yield await emit({ type: "guardrail_decision", decision });
      }
      this.promptMessage = inputResult.prompt;
      if (inputResult.blocked) {
        const output = inputResult.message ?? "The request was blocked by a guardrail.";
        const result = {
          output,
          usage: Usage.empty(),
          messages: [this.promptMessage, Message.assistant(output)],
          guardrails: [...this.guardrailDecisions]
        };
        await runObservers.end(result);
        this.runState = "completed";
        yield await emit({
          type: "final",
          runId,
          output: result.output,
          usage: result.usage,
          messages: result.messages,
          guardrails: result.guardrails
        });
        return;
      }
      newMessages = [this.promptMessage];
      this.chatHistory = await this.memoryRecorder.prepareRun(runId, newMessages);
      const pendingTurnMessages = this.memoryRecorder.pendingTurnMessages(newMessages);
      await this.runRunStartHook(newMessages);
      while (currentTurns <= this.maxTurnCount + 1) {
        const prompt = newMessages.at(-1);
        if (prompt === void 0) {
          throw new Error("PromptRequest requires at least one message");
        }
        lastPrompt = prompt;
        currentTurns += 1;
        const historyForRequest = [...this.chatHistory, ...newMessages.slice(0, -1)];
        yield await emit({
          type: "turn_start",
          turn: currentTurns,
          prompt,
          history: historyForRequest
        });
        await this.runTurnStartHook(currentTurns, prompt, historyForRequest, newMessages);
        await this.runCompletionCallHook(prompt, historyForRequest, newMessages);
        const ragText = extractRagText(prompt);
        const dynamicContext = await fetchDynamicContext(this.agent, ragText);
        const toolDefs = await fetchToolDefinitions(this.agent, ragText);
        let request = new CompletionRequestBuilder(this.agent.model, prompt).instructions(this.agent.instructions).messages(historyForRequest).documents([...this.agent.staticContext, ...dynamicContext]).tools(toolDefs).temperature(this.agent.temperature).maxTokens(this.agent.maxTokens).additionalParams(this.agent.additionalParams).toolChoice(this.agent.toolChoice).outputSchema(this.agent.outputSchema).build();
        request = await this.runCompletionRequestMiddlewares(
          request,
          currentTurns
        );
        assertCompletionRequestSupported(this.agent.model, request, { streaming: true });
        const providerRequest = this.providerTraceRequest(request, { stream: true });
        const generationObservers = await runObservers.startGeneration(
          compact({
            turn: currentTurns,
            request,
            providerRequest,
            modelInfo: {
              provider: this.agent.model.provider,
              defaultModel: this.agent.model.defaultModel,
              capabilities: this.agent.model.capabilities
            }
          })
        );
        const accumulator = new CompletionStreamAccumulator();
        const generationStartedAt = Date.now();
        let firstDeltaMs;
        const bufferResponseEvents = this.shouldBufferStreamResponseEvents();
        const emittedToolCallIds = /* @__PURE__ */ new Set();
        try {
          for await (const event of this.agent.model.streamCompletion(request)) {
            if (firstDeltaMs === void 0 && isGenerationDeltaEvent(event.type)) {
              firstDeltaMs = Date.now() - generationStartedAt;
            }
            const mapped = accumulator.accept(event);
            if (event.type === "error") {
              throw event.error;
            }
            if (mapped !== void 0) {
              await generationObservers.update?.({ turn: currentTurns, delta: mapped });
              if (mapped.type === "tool_call") {
                emittedToolCallIds.add(mapped.toolCall.id);
              }
              const shouldBuffer = bufferResponseEvents || bufferOutputDeltas && (mapped.type === "text_delta" || mapped.type === "reasoning_delta");
              if (!shouldBuffer) {
                yield await emit(addTurn(currentTurns, mapped));
              }
            }
          }
        } catch (error) {
          await generationObservers.error({ turn: currentTurns, error });
          await this.runCompletionErrorHook(prompt, error, newMessages);
          throw error;
        }
        let response = accumulator.response();
        await generationObservers.end(
          compact({
            turn: currentTurns,
            response,
            firstDeltaMs
          })
        );
        response = await this.runCompletionResponseMiddlewares(request, response, currentTurns);
        usage = Usage.add(usage, response.usage);
        await this.runCompletionResponseHook(prompt, response, newMessages);
        await this.runTurnEndHook(currentTurns, response, newMessages);
        const toolCalls = response.choice.filter(
          (item) => item.type === "tool_call"
        );
        let assistantMessage = Message.assistant(response.choice, response.messageId);
        newMessages.push(assistantMessage);
        if (toolCalls.length === 0) {
          let emittedTurnEnd = false;
          if (!bufferOutputDeltas) {
            if (bufferResponseEvents) {
              for (const event of responseStreamEvents(currentTurns, response)) {
                yield await emit(event);
              }
            }
            yield await emit({ type: "turn_end", turn: currentTurns, response });
            emittedTurnEnd = true;
          }
          if (this.steeringMessages.length > 0) {
            await this.memoryRecorder.commitMessages(
              runId,
              currentTurns,
              [assistantMessage],
              pendingTurnMessages
            );
          }
          if (await this.drainSteeringMessages(runId, currentTurns, newMessages, pendingTurnMessages)) {
            await this.memoryRecorder.commitCompletedTurn(runId, currentTurns, pendingTurnMessages);
            continue;
          }
          const guardedOutput = await this.runOutputGuardrailsForResponse(
            runId,
            usage,
            response,
            newMessages,
            runObservers
          );
          for (const decision of guardedOutput.decisions) {
            yield await emit({ type: "guardrail_decision", decision });
          }
          response = guardedOutput.response;
          assistantMessage = Message.assistant(response.choice, response.messageId);
          newMessages[newMessages.length - 1] = assistantMessage;
          await this.memoryRecorder.commitMessages(
            runId,
            currentTurns,
            [assistantMessage],
            pendingTurnMessages
          );
          if (!emittedTurnEnd && (bufferResponseEvents || bufferOutputDeltas)) {
            for (const event of responseStreamEvents(currentTurns, response)) {
              yield await emit(event);
            }
          }
          if (!emittedTurnEnd) {
            yield await emit({ type: "turn_end", turn: currentTurns, response });
          }
          const result = {
            output: guardedOutput.output,
            usage,
            messages: [...newMessages],
            trace: runObservers.trace,
            guardrails: [...this.guardrailDecisions]
          };
          await this.runRunEndHook(result, newMessages);
          await runObservers.end(result);
          await this.memoryRecorder.commitCompletedRun(
            runId,
            currentTurns,
            newMessages,
            pendingTurnMessages
          );
          this.runState = "completed";
          yield await emit({
            type: "final",
            runId,
            output: result.output,
            usage: result.usage,
            messages: result.messages,
            trace: result.trace,
            guardrails: result.guardrails
          });
          return;
        }
        if (bufferResponseEvents) {
          for (const event of responseStreamEvents(currentTurns, response)) {
            yield await emit(event);
          }
        } else {
          for (const toolCall of toolCalls) {
            if (!emittedToolCallIds.has(toolCall.id)) {
              yield await emit({ type: "tool_call", turn: currentTurns, toolCall });
            }
          }
        }
        await this.memoryRecorder.commitMessages(
          runId,
          currentTurns,
          [assistantMessage],
          pendingTurnMessages
        );
        yield await emit({ type: "turn_end", turn: currentTurns, response });
        const toolResultEvents = createAsyncQueue();
        const toolResultsPromise = this.executeToolCalls(
          runId,
          toolCalls,
          newMessages,
          (result) => {
            toolResultEvents.enqueue(result);
          },
          (event) => {
            toolResultEvents.enqueue(event);
          },
          {
            turn: currentTurns,
            runObservers,
            toolDefinitions: request.tools
          }
        );
        toolResultsPromise.then(
          () => toolResultEvents.close(),
          (error) => toolResultEvents.throw(error)
        );
        for await (const result of toolResultEvents) {
          yield await emit({ turn: currentTurns, ...result });
        }
        const toolResults = await toolResultsPromise;
        const toolMessage = Message.tool(toolResults);
        newMessages.push(toolMessage);
        await this.memoryRecorder.commitMessages(
          runId,
          currentTurns,
          [toolMessage],
          pendingTurnMessages
        );
        await this.drainSteeringMessages(runId, currentTurns, newMessages, pendingTurnMessages);
        await this.memoryRecorder.commitCompletedTurn(runId, currentTurns, pendingTurnMessages);
      }
      throw new MaxTurnsError(this.maxTurnCount, [...this.chatHistory, ...newMessages], lastPrompt);
    } catch (error) {
      const finalError = await this.runRunErrorHook(error, usage, newMessages);
      this.runState = finalError instanceof PromptCancelledError ? "cancelled" : "errored";
      await runObservers.error({ error: finalError, usage, messages: [...newMessages] });
      await this.memoryRecorder.recordError(runId, finalError, newMessages);
      yield await emit({ type: "error", error: finalError });
      throw finalError;
    }
  }
  readableStream() {
    return toReadableStream(this.stream());
  }
  async runCompletion(request, turn, runObservers) {
    assertCompletionRequestSupported(this.agent.model, request);
    const providerRequest = this.providerTraceRequest(request);
    const generationObservers = await runObservers.startGeneration(
      compact({
        turn,
        request,
        providerRequest,
        modelInfo: {
          provider: this.agent.model.provider,
          defaultModel: this.agent.model.defaultModel,
          capabilities: this.agent.model.capabilities
        }
      })
    );
    try {
      const response = await this.agent.model.completion(request);
      await generationObservers.end({ turn, response });
      return response;
    } catch (error) {
      await generationObservers.error({ turn, error });
      throw error;
    }
  }
  providerTraceRequest(request, options = {}) {
    try {
      return this.agent.model.traceRequest?.(request, options);
    } catch (error) {
      return {
        error: error instanceof Error ? error.message : String(error)
      };
    }
  }
  async executeToolCalls(runId, toolCalls, newMessages, onResult, onStreamEvent, observation) {
    const executor = new ToolCallExecutor(
      this.agent,
      this.activeHook,
      this.approvalOptions,
      {
        runId,
        sessionId: this.memoryContext?.sessionId,
        metadata: this.memoryContext?.metadata
      },
      this.concurrency,
      this.requestMiddlewares,
      (reason) => this.cancelled(newMessages, reason)
    );
    return executor.execute(toolCalls, onResult, onStreamEvent, observation);
  }
  async runOutputGuardrailsForResponse(runId, usage, response, messages, runObservers) {
    const originalOutput = textFromAssistantContent(response.choice);
    const result = await runOutputGuardrails(this.guardrailPolicies, {
      outputText: originalOutput,
      messages: [...this.chatHistory, ...messages],
      usage,
      run: this.guardrailRunContext(runId)
    });
    for (const decision of result.decisions) {
      await this.recordGuardrailDecision(decision, runObservers);
    }
    const output = result.blocked ? result.message ?? "The response was blocked by a guardrail." : result.outputText;
    if (output === originalOutput) {
      return { output, response, decisions: result.decisions };
    }
    return {
      output,
      response: {
        ...response,
        choice: [AssistantContent.text(output)]
      },
      decisions: result.decisions
    };
  }
  async recordGuardrailDecision(decision, runObservers) {
    this.guardrailDecisions.push(decision);
    await runObservers.event({
      name: "guardrail.decision",
      level: decision.action === "block" ? "WARNING" : "DEFAULT",
      attributes: guardrailDecisionAttributes(decision)
    });
  }
  guardrailRunContext(runId) {
    return compact({
      agentId: this.agent.id,
      runId,
      sessionId: this.memoryContext?.sessionId,
      metadata: this.memoryContext?.metadata
    });
  }
  async startRunObservers() {
    const failOnObserverError = this.traceOptions?.failOnObserverError === true || this.agent.observers.some((registration) => registration.failOnObserverError === true);
    return startAgentRunObservers(
      this.agent.observers,
      {
        agentName: this.agent.name,
        agentDescription: this.agent.description,
        instructions: this.agent.instructions,
        trace: this.traceOptions,
        prompt: this.promptMessage,
        history: this.chatHistory,
        maxTurns: this.maxTurnCount
      },
      failOnObserverError
    );
  }
  async recordAgentEvent(runId, event) {
    const registration = this.agent.eventStore;
    if (registration === void 0) {
      return;
    }
    if (registration.options.include === "agent_tool_events" && event.type !== "agent_tool_event") {
      return;
    }
    const turn = "turn" in event ? event.turn : void 0;
    const agentId = event.type === "agent_tool_event" ? event.agentId : this.agent.id;
    const agentName = event.type === "agent_tool_event" ? event.agentName : this.agent.name;
    await registration.store.append({
      runId,
      agentId,
      ...compact({ agentName, turn }),
      ...event.type === "agent_tool_event" ? compact({
        toolName: event.toolName,
        toolCallId: event.toolCallId,
        internalCallId: event.internalCallId
      }) : {},
      event
    });
  }
  async runCompletionCallHook(prompt, history, newMessages) {
    const action = await this.activeHook?.onCompletionCall?.({
      prompt,
      history,
      run: runControl
    });
    if (action?.type === "terminate") {
      throw this.cancelled(newMessages, action.reason);
    }
  }
  async runRunStartHook(newMessages) {
    const action = await this.activeHook?.onRunStart?.({
      prompt: this.promptMessage,
      history: this.chatHistory,
      maxTurns: this.maxTurnCount,
      run: runControl
    });
    if (action?.type === "terminate") {
      throw this.cancelled(newMessages, action.reason);
    }
  }
  async runRunEndHook(result, newMessages) {
    const action = await this.activeHook?.onRunEnd?.({
      output: result.output,
      usage: result.usage,
      messages: result.messages,
      run: runControl
    });
    if (action?.type === "terminate") {
      throw this.cancelled(newMessages, action.reason);
    }
  }
  async runRunErrorHook(error, usage, newMessages) {
    const action = await this.activeHook?.onRunError?.({
      error,
      usage,
      messages: [...this.chatHistory, ...newMessages],
      run: runControl
    });
    if (action?.type === "terminate") {
      return this.cancelled(newMessages, action.reason);
    }
    return error;
  }
  async runTurnStartHook(turn, prompt, history, newMessages) {
    const action = await this.activeHook?.onTurnStart?.({
      turn,
      prompt,
      history,
      run: runControl
    });
    if (action?.type === "terminate") {
      throw this.cancelled(newMessages, action.reason);
    }
  }
  async runTurnEndHook(turn, response, newMessages) {
    const action = await this.activeHook?.onTurnEnd?.({
      turn,
      response,
      run: runControl
    });
    if (action?.type === "terminate") {
      throw this.cancelled(newMessages, action.reason);
    }
  }
  async runCompletionRequestMiddlewares(request, turn) {
    let current = request;
    for (const middleware of this.activeMiddlewares()) {
      const replacement = await middleware.onCompletionRequest?.({
        turn,
        request: current,
        originalRequest: request
      });
      if (replacement?.request !== void 0) {
        current = replacement.request;
      }
    }
    return current;
  }
  async runCompletionResponseMiddlewares(request, response, turn) {
    let current = response;
    for (const middleware of this.activeMiddlewares()) {
      const replacement = await middleware.onCompletionResponse?.({
        turn,
        request,
        response: current,
        originalResponse: response
      });
      if (replacement?.response !== void 0) {
        current = replacement.response;
      }
    }
    return current;
  }
  async runCompletionResponseHook(prompt, response, newMessages) {
    const action = await this.activeHook?.onCompletionResponse?.({
      prompt,
      response,
      run: runControl
    });
    if (action?.type === "terminate") {
      throw this.cancelled(newMessages, action.reason);
    }
  }
  async runCompletionErrorHook(prompt, error, newMessages) {
    const action = await this.activeHook?.onCompletionError?.({
      prompt,
      error,
      run: runControl
    });
    if (action?.type === "terminate") {
      throw this.cancelled(newMessages, action.reason);
    }
  }
  activeMiddlewares() {
    return [...this.agent.middlewares, ...this.requestMiddlewares];
  }
  shouldBufferStreamResponseEvents() {
    return this.activeHook?.onCompletionResponse !== void 0 || this.activeMiddlewares().some((middleware) => middleware.onCompletionResponse !== void 0);
  }
  async drainSteeringMessages(runId, turn, newMessages, pendingTurnMessages) {
    const messages = this.steeringMessages.splice(0);
    if (messages.length === 0) {
      return false;
    }
    newMessages.push(...messages);
    await this.memoryRecorder.commitMessages(runId, turn, messages, pendingTurnMessages);
    return true;
  }
  startRun() {
    if (this.runState === "idle") {
      this.runState = "running";
      return;
    }
    if (this.runState === "running") {
      throw new Error("PromptRequest is already running.");
    }
    throw new Error("PromptRequest has already been used.");
  }
  isTerminal() {
    return this.runState === "completed" || this.runState === "errored" || this.runState === "cancelled";
  }
  cancelled(newMessages, reason) {
    return new PromptCancelledError([...this.chatHistory, ...newMessages], reason);
  }
};
function normalizePromptInput(prompt) {
  if (typeof prompt === "string") {
    return { prompt: Message.user(prompt), history: [] };
  }
  if (!Array.isArray(prompt)) {
    return { prompt, history: [] };
  }
  if (prompt.length === 0) {
    throw new TypeError("Prompt transcript must contain at least one message.");
  }
  const activePrompt = prompt.at(-1);
  if (activePrompt === void 0) {
    throw new TypeError("Prompt transcript must contain at least one message.");
  }
  return {
    prompt: activePrompt,
    history: prompt.slice(0, -1)
  };
}
function normalizeSteeringInput(input) {
  if (typeof input === "string") {
    return [Message.user(input)];
  }
  return Array.isArray(input) ? [...input] : [input];
}
function responseStreamEvents(turn, response) {
  const events = [];
  for (const item of response.choice) {
    if (item.type === "text") {
      if (item.text.length > 0) {
        events.push({ type: "text_delta", turn, delta: item.text });
      }
      continue;
    }
    if (item.type === "reasoning") {
      if (item.content === void 0) {
        if (item.text.length > 0) {
          events.push(
            compact({ type: "reasoning_delta", turn, delta: item.text, id: item.id })
          );
        }
        continue;
      }
      for (const content of item.content) {
        const delta = content.type === "encrypted" || content.type === "redacted" ? content.data : content.text;
        events.push(
          compact({
            type: "reasoning_delta",
            turn,
            delta,
            id: item.id,
            contentType: content.type,
            signature: content.type === "text" ? content.signature : void 0
          })
        );
      }
      continue;
    }
    if (item.type === "tool_call") {
      events.push({ type: "tool_call", turn, toolCall: item });
    }
  }
  return events;
}
function textFromMessage(message) {
  if (message.role === "system") {
    return message.content;
  }
  return message.content.flatMap((content) => {
    if (content.type === "text") {
      return [content.text];
    }
    if (content.type === "document" && content.source.type === "text") {
      return [content.source.text];
    }
    return [];
  }).join("\n");
}
function guardrailDecisionAttributes(decision) {
  return compact({
    policyId: decision.policyId,
    guardrailId: decision.guardrailId,
    boundary: decision.boundary,
    mode: decision.mode,
    action: decision.action,
    applied: decision.applied,
    reason: decision.reason,
    message: decision.message,
    latencyMs: decision.latencyMs
  });
}

export {
  MaxTurnsError,
  PromptCancelledError,
  ToolApprovalRequiredError,
  extractRagText,
  PromptRequest
};
