// @anvia/core@0.13.5/dist/pipeline/index.js
import {
  mapWithConcurrency
} from "../chunk-OIMLU4SF.js";

// src/pipeline/builder.ts
import { z } from "zod";

// src/pipeline/graph.ts
function initialBuilderState(metadata) {
  return {
    graph: initialGraph(metadata),
    terminalNodeId: "input",
    terminalNodeIds: ["input"],
    nextNodeIndex: 1,
    nextEdgeIndex: 1
  };
}
function initialGraph(metadata) {
  const id = normalizeId(metadata.id ?? "pipeline");
  const graph = {
    id,
    nodes: [{ id: "input", kind: "input", label: "Input" }],
    edges: []
  };
  if (metadata.name !== void 0) {
    graph.name = metadata.name;
  }
  if (metadata.description !== void 0) {
    graph.description = metadata.description;
  }
  if (metadata.metadata !== void 0) {
    graph.metadata = metadata.metadata;
  }
  return graph;
}
function appendNode(state, kind, label, options = {}) {
  const node = graphNode(kind, label, state.nextNodeIndex, {
    ...options,
    existingIds: new Set(state.graph.nodes.map((item) => item.id))
  });
  return {
    node,
    state: appendGraphNode(state, node, activeTerminalNodeIds(state), [node.id])
  };
}
function appendChildNode(state, parentId, kind, label, options = {}) {
  const node = graphNode(kind, label, state.nextNodeIndex, {
    ...options,
    existingIds: new Set(state.graph.nodes.map((item) => item.id))
  });
  return {
    node,
    state: appendGraphNode(state, node, [parentId], activeTerminalNodeIds(state))
  };
}
function activeTerminalNodeIds(state) {
  return state.terminalNodeIds.length > 0 ? state.terminalNodeIds : [state.terminalNodeId];
}
function withTerminalNodes(state, terminalNodeIds) {
  return {
    ...state,
    terminalNodeId: terminalNodeIds.at(-1) ?? state.terminalNodeId,
    terminalNodeIds
  };
}
function withOutputNode(state) {
  const graph = cloneGraph(state.graph);
  if (graph.nodes.some((node) => node.id === "output")) {
    return graph;
  }
  graph.nodes.push({ id: "output", kind: "output", label: "Output" });
  graph.edges.push(
    ...activeTerminalNodeIds(state).map((sourceId, index) => ({
      id: `edge_${state.nextEdgeIndex + index}`,
      source: sourceId,
      target: "output"
    }))
  );
  return graph;
}
function nextStageLabel(state, prefix) {
  return `${prefix} ${state.nextNodeIndex}`;
}
function cloneGraph(graph) {
  return {
    ...graph,
    nodes: graph.nodes.map((node) => ({ ...node })),
    edges: graph.edges.map((edge) => ({ ...edge }))
  };
}
function appendGraphNode(state, node, sourceIds, terminalNodeIds) {
  const edges = sourceIds.map((sourceId, index) => ({
    id: `edge_${state.nextEdgeIndex + index}`,
    source: sourceId,
    target: node.id
  }));
  const terminalNodeId = terminalNodeIds.at(-1) ?? state.terminalNodeId;
  return {
    graph: {
      ...state.graph,
      nodes: [...state.graph.nodes, node],
      edges: [...state.graph.edges, ...edges]
    },
    terminalNodeId,
    terminalNodeIds,
    nextNodeIndex: state.nextNodeIndex + 1,
    nextEdgeIndex: state.nextEdgeIndex + edges.length
  };
}
function graphNode(kind, label, index, options = {}) {
  const id = uniqueGraphNodeId(
    normalizeId(options.preferredId ?? `${kind}_${index}`),
    options.existingIds ?? /* @__PURE__ */ new Set()
  );
  const node = {
    id,
    kind,
    label
  };
  if (options.description !== void 0) {
    node.description = options.description;
  }
  if (options.metadata !== void 0) {
    node.metadata = options.metadata;
  }
  if (options.agentId !== void 0) {
    node.agentId = options.agentId;
  }
  if (options.agentName !== void 0) {
    node.agentName = options.agentName;
  }
  if (options.pipelineId !== void 0) {
    node.pipelineId = options.pipelineId;
  }
  if (options.branchKey !== void 0) {
    node.branchKey = options.branchKey;
  }
  return node;
}
function normalizeId(value) {
  const normalized = value.trim().toLowerCase().replace(/[^a-z0-9_-]+/g, "_").replace(/^_+|_+$/g, "");
  return normalized.length === 0 ? "pipeline" : normalized;
}
function uniqueGraphNodeId(baseId, existingIds) {
  let id = baseId;
  let suffix = 2;
  while (existingIds.has(id)) {
    id = `${baseId}_${suffix}`;
    suffix += 1;
  }
  return id;
}

// src/pipeline/pipeline.ts
var Pipeline = class {
  constructor(executor, pipelineGraph = initialGraph({})) {
    this.executor = executor;
    this.pipelineGraph = pipelineGraph;
    this.id = pipelineGraph.id;
    this.name = pipelineGraph.name;
    this.description = pipelineGraph.description;
    this.metadata = pipelineGraph.metadata;
  }
  executor;
  pipelineGraph;
  id;
  name;
  description;
  metadata;
  /** Run one input through the built pipeline and return the final stage output. */
  async run(input, options = {}) {
    return await this.executor(input, { observer: options.observer });
  }
  /** Run many inputs through the same pipeline with bounded concurrency. */
  async batch(inputs, options) {
    return mapWithConcurrency([...inputs], options.concurrency, (input) => this.run(input));
  }
  graph() {
    return cloneGraph(this.pipelineGraph);
  }
};

// src/pipeline/runtime.ts
async function runNode(context, node, fn) {
  const startedAt = Date.now();
  await context.observer?.onEvent({ type: "stage_started", node });
  try {
    const output = await fn();
    await context.observer?.onEvent({
      type: "stage_completed",
      node,
      durationMs: Date.now() - startedAt
    });
    return output;
  } catch (error) {
    await context.observer?.onEvent({
      type: "stage_failed",
      node,
      durationMs: Date.now() - startedAt,
      error
    });
    throw error;
  }
}

// src/pipeline/builder.ts
var PipelineBuilder = class _PipelineBuilder {
  executor;
  state;
  constructor(metadataOrExecutor, stateOrMetadata) {
    if (stateOrMetadata !== void 0 && isBuilderState(stateOrMetadata)) {
      this.executor = metadataOrExecutor;
      this.state = stateOrMetadata;
      return;
    }
    if (typeof metadataOrExecutor === "function") {
      const executor = metadataOrExecutor;
      this.executor = ((input) => executor(input));
      this.state = initialBuilderState({});
      return;
    }
    if (isZodSchema(metadataOrExecutor)) {
      const schema = metadataOrExecutor;
      const metadata = stateOrMetadata !== void 0 && isPipelineMetadata(stateOrMetadata) ? stateOrMetadata : {};
      this.executor = ((input) => schema.parse(input));
      this.state = initialBuilderState(metadata);
      return;
    }
    this.executor = identity;
    this.state = initialBuilderState(metadataOrExecutor ?? {});
  }
  /** Add a synchronous or asynchronous transform stage. */
  step(fn, metadata) {
    const next = appendNode(
      this.state,
      "step",
      metadata?.name ?? nextStageLabel(this.state, "Step"),
      {
        description: metadata?.description,
        metadata: metadata?.metadata,
        preferredId: metadata?.id
      }
    );
    return new _PipelineBuilder(
      async (input, context) => {
        const value = await this.runStep(input, context);
        const result = await runNode(context, next.node, () => fn(value));
        return result;
      },
      next.state
    );
  }
  /** Compose another pipeline operation after the current stage. */
  use(op, metadata) {
    const nested = op instanceof Pipeline ? op : void 0;
    const next = appendNode(
      this.state,
      nested === void 0 ? "step" : "pipeline",
      metadata?.name ?? nested?.name ?? nested?.id ?? nextStageLabel(this.state, nested === void 0 ? "Operation" : "Pipeline"),
      {
        description: metadata?.description ?? nested?.description,
        metadata: metadata?.metadata ?? nested?.metadata,
        preferredId: metadata?.id,
        pipelineId: nested?.id
      }
    );
    return new _PipelineBuilder(
      async (input, context) => {
        const value = await this.runStep(input, context);
        const result = await runNode(context, next.node, () => op.run(value));
        return result;
      },
      next.state
    );
  }
  /** Run named branch operations concurrently from the current value. */
  parallel(branches, metadata) {
    const parallel = appendNode(
      this.state,
      "parallel",
      metadata?.name ?? `${Object.keys(branches).length} parallel branches`,
      {
        description: metadata?.description,
        metadata: metadata?.metadata,
        preferredId: metadata?.id
      }
    );
    let nextState = parallel.state;
    const branchNodes = {};
    for (const key of Object.keys(branches)) {
      const branch = appendChildNode(nextState, parallel.node.id, "branch", key, {
        branchKey: key
      });
      nextState = branch.state;
      branchNodes[key] = branch.node;
    }
    nextState = withTerminalNodes(
      nextState,
      Object.values(branchNodes).map((node) => node.id)
    );
    return new _PipelineBuilder(async (input, context) => {
      const value = await this.runStep(input, context);
      const entries = await runNode(
        context,
        parallel.node,
        () => Promise.all(
          Object.entries(branches).map(async ([key, op]) => {
            const node = branchNodes[key];
            const output = await runNode(context, node, () => op.run(value));
            return [key, output];
          })
        )
      );
      return Object.fromEntries(entries);
    }, nextState);
  }
  /** Send the current value to an agent as text and continue with the agent output. */
  prompt(agent, metadata) {
    const next = appendNode(this.state, "agent", metadata?.name ?? agent.name ?? agent.id, {
      description: metadata?.description ?? agent.description,
      metadata: metadata?.metadata,
      preferredId: metadata?.id,
      agentId: agent.id,
      agentName: agent.name
    });
    return new _PipelineBuilder(async (input, context) => {
      const value = await this.runStep(input, context);
      return runNode(context, next.node, async () => {
        const response = await agent.prompt(String(value)).send();
        return response.output;
      });
    }, next.state);
  }
  /** Send the current value to an extractor as text and continue with typed schema data. */
  extract(extractor, metadata) {
    const next = appendNode(
      this.state,
      "extractor",
      metadata?.name ?? nextStageLabel(this.state, "Extractor"),
      {
        description: metadata?.description,
        metadata: metadata?.metadata,
        preferredId: metadata?.id
      }
    );
    return new _PipelineBuilder(async (input, context) => {
      const value = await this.runStep(input, context);
      return runNode(context, next.node, () => extractor.extract(String(value)));
    }, next.state);
  }
  /** Finish the builder and return a runnable pipeline. */
  build() {
    const graph = withOutputNode(this.state);
    return new Pipeline(
      (input, context) => this.runStep(input, context),
      graph
    );
  }
  async runStep(input, context) {
    return await this.executor(input, context);
  }
};
function identity(input) {
  return input;
}
function isZodSchema(value) {
  return value instanceof z.ZodType;
}
function isBuilderState(value) {
  if (typeof value !== "object" || value === null || Array.isArray(value)) {
    return false;
  }
  const candidate = value;
  return "graph" in candidate && "terminalNodeId" in candidate && "terminalNodeIds" in candidate && "nextNodeIndex" in candidate && "nextEdgeIndex" in candidate;
}
function isPipelineMetadata(value) {
  if (typeof value !== "object" || value === null || Array.isArray(value)) {
    return false;
  }
  const candidate = value;
  return "id" in candidate || "name" in candidate || "description" in candidate || "metadata" in candidate;
}
export {
  Pipeline,
  PipelineBuilder
};
